--- DCGO/Assets/Scripts/CardEffect/BT25/Blue/BT25_096.cs	2026-05-18 20:03:15.024447300 +0900
+++ src/HeadlessDCGO.Engine/Assets/Scripts/CardEffect/BT25/Blue/BT25_096.cs	2026-07-29 01:02:22.751907242 +0900
@@ -1,469 +1,469 @@
-using System;
+namespace HeadlessDCGO.Engine.Assets.Scripts.CardEffect.BT25.Blue;
+
 using System.Collections;
-using System.Collections.Generic;
 using System.Linq;
+using System.Threading.Tasks;
+using HeadlessDCGO.Engine.Assets.Scripts.Script;
+using HeadlessDCGO.Engine.Assets.Scripts.Script.CardEffectCommons;
+using HeadlessDCGO.Engine.Assets.Scripts.Script.CardEffects;
+using HeadlessDCGO.Engine.Headless.Services;
 
-// Mirage Beast Knight
-namespace DCGO.CardEffects.BT25
+public class BT25_096 : CEntity_Effect
 {
-    public class BT25_096 : CEntity_Effect
+    public override List<ICardEffect> CardEffects(EffectTiming timing, CardSource card)
     {
-        public override List<ICardEffect> CardEffects(EffectTiming timing, CardSource card)
-        {
-            List<ICardEffect> cardEffects = new List<ICardEffect>();
+        List<ICardEffect> cardEffects = new List<ICardEffect>();
+
+        #region Reduce Play Cost
 
-            #region Reduce Play Cost
+        bool SharedCanSelectTamerCondition(Permanent permanent)
+            => CardEffectCommons.IsPermanentExistsOnOwnerBattleAreaTamer(permanent, card)
+                && permanent.HasFaceDownDigivolutionCards;
 
-            bool SharedCanSelectTamerCondition(Permanent permanent)
-                => CardEffectCommons.IsPermanentExistsOnOwnerBattleAreaTamer(permanent, card)
-                    && permanent.HasFaceDownDigivolutionCards;
+        if (timing == EffectTiming.BeforePayCost)
+        {
+            ActivateClass activateClass = new ActivateClass();
+            activateClass.SetUpICardEffect("Reduce Play Cost -2", CanUseCondition, card);
+            activateClass.SetUpActivateClass(null, ActivateCoroutine, -1, false, EffectDescription());
+            activateClass.SetIsSkippable(true);
+            cardEffects.Add(activateClass);
 
-            if (timing == EffectTiming.BeforePayCost)
+            string EffectDescription()
             {
-                ActivateClass activateClass = new ActivateClass();
-                activateClass.SetUpICardEffect("Reduce Play Cost -2", CanUseCondition, card);
-                activateClass.SetUpActivateClass(null, ActivateCoroutine, -1, false, EffectDescription());
-                activateClass.SetIsSkippable(true);
-                cardEffects.Add(activateClass);
+                return "When this card would be used, by trashing the bottom face-down card from under any of your Tamers, reduce the use cost by 2.";
+            }
 
-                string EffectDescription()
-                {
-                    return "When this card would be used, by trashing the bottom face-down card from under any of your Tamers, reduce the use cost by 2.";
-                }
+            bool CanUseCondition(Hashtable hashtable)
+            {
+                return CardEffectCommons.CanTriggerWhenPermanentWouldPlay(hashtable, cardSource => cardSource == card)
+                    && CardEffectCommons.HasMatchConditionOwnersPermanent(card, SharedCanSelectTamerCondition);
+            }
 
-                bool CanUseCondition(Hashtable hashtable)
+            async Task ActivateCoroutine(Hashtable _hashtable)
+            {
+                if (CardEffectCommons.HasMatchConditionOwnersPermanent(card, SharedCanSelectTamerCondition))
                 {
-                    return CardEffectCommons.CanTriggerWhenPermanentWouldPlay(hashtable, cardSource => cardSource == card)
-                        && CardEffectCommons.HasMatchConditionOwnersPermanent(card, SharedCanSelectTamerCondition);
-                }
+                    Permanent? selectedPermanent = null;
 
-                IEnumerator ActivateCoroutine(Hashtable _hashtable)
-                {
-                    if (CardEffectCommons.HasMatchConditionOwnersPermanent(card, SharedCanSelectTamerCondition))
+                    #region Select Tamer to trash bottom face down digivolution card
+                    if (CardEffectCommons.HasMatchConditionPermanent(card, SharedCanSelectTamerCondition))
                     {
-                        Permanent selectedPermanent = null;
+                        SelectPermanentEffect selectPermanentEffect = GManager.instance.GetComponent<SelectPermanentEffect>();
 
-                        #region Select Tamer to trash bottom face down digivolution card
-                        if (CardEffectCommons.HasMatchConditionPermanent(SharedCanSelectTamerCondition))
-                        {
-                            SelectPermanentEffect selectPermanentEffect = GManager.instance.GetComponent<SelectPermanentEffect>();
+                        selectPermanentEffect.SetUp(
+                            selectPlayer: card.Owner,
+                            canTargetCondition: SharedCanSelectTamerCondition,
+                            canTargetCondition_ByPreSelecetedList: null,
+                            canEndSelectCondition: null,
+                            maxCount: 1,
+                            canNoSelect: true,
+                            canEndNotMax: false,
+                            selectPermanentCoroutine: SelectPermanentCoroutine,
+                            afterSelectPermanentCoroutine: null,
+                            mode: SelectPermanentEffect.Mode.Custom,
+                            cardEffect: activateClass);
 
-                            selectPermanentEffect.SetUp(
-                                selectPlayer: card.Owner,
-                                canTargetCondition: SharedCanSelectTamerCondition,
-                                canTargetCondition_ByPreSelecetedList: null,
-                                canEndSelectCondition: null,
-                                maxCount: 1,
-                                canNoSelect: true,
-                                canEndNotMax: false,
-                                selectPermanentCoroutine: SelectPermanentCoroutine,
-                                afterSelectPermanentCoroutine: null,
-                                mode: SelectPermanentEffect.Mode.Custom,
-                                cardEffect: activateClass);
+                        Task SelectPermanentCoroutine(Permanent permanent)
+                        {
+                            selectedPermanent = permanent;
+                            return Task.CompletedTask;
+                        }
 
-                            IEnumerator SelectPermanentCoroutine(Permanent permanent)
-                            {
-                                selectedPermanent = permanent;
-                                yield return null;
-                            }
+                        selectPermanentEffect.SetUpCustomMessage("Select 1 tamer to trash bottom face down digivolution card.", "The opponent is selecting 1 tamer to trash bottom face down digivolution card.");
+                        await selectPermanentEffect.Activate();
+                    }
+                    #endregion
 
-                            selectPermanentEffect.SetUpCustomMessage("Select 1 tamer to trash bottom face down digivolution card.", "The opponent is selecting 1 tamer to trash bottom face down digivolution card.");
-                            yield return ContinuousController.instance.StartCoroutine(selectPermanentEffect.Activate());
-                        }
-                        #endregion
+                    if (selectedPermanent != null)
+                    {
+                        CardSource cardToTrash = selectedPermanent.DigivolutionCards.Last(x => x.IsFlipped);
+                        await CardEffectCommons.TrashDigivolutionCardsAndProcessAccordingToResult(
+                            targetPermanent: selectedPermanent,
+                            targetDigivolutionCards: new List<CardSource>() { cardToTrash },
+                            activateClass: activateClass,
+                            successProcess: SuccessProcess,
+                            failureProcess: null);
 
-                        if (selectedPermanent != null)
+                        async Task SuccessProcess(List<CardSource> trashedCards)
                         {
-                            CardSource cardToTrash = selectedPermanent.DigivolutionCards.Last(x => x.IsFaceDown);
-                            yield return ContinuousController.instance.StartCoroutine(CardEffectCommons.TrashDigivolutionCardsAndProcessAccordingToResult(
-                                targetPermanent: selectedPermanent,
-                                targetDigivolutionCards: new List<CardSource>() { cardToTrash },
-                                activateClass: activateClass,
-                                successProcess: SuccessProcess,
-                                failureProcess: null));
-
-                            IEnumerator SuccessProcess(List<CardSource> trashedCards)
-                            {
-                                if (card.Owner.CanReduceCost(null, card)) ContinuousController.instance.PlaySE(GManager.instance.GetComponent<Effects>().BuffSE);
 
-                                ChangeCostClass changeCostClass = new ChangeCostClass();
-                                changeCostClass.SetUpICardEffect("Play Cost -2", CanUseCondition1, card);
-                                changeCostClass.SetUpChangeCostClass(changeCostFunc: ChangeCost, cardSourceCondition: CardSourceCondition, rootCondition: RootCondition, isUpDown: isUpDown, isCheckAvailability: () => false, isChangePayingCost: () => true);
-                                card.Owner.UntilCalculateFixedCostEffect.Add((_timing) => changeCostClass);
+                            ChangeCostClass changeCostClass = new ChangeCostClass();
+                            changeCostClass.SetUpICardEffect("Play Cost -2", CanUseCondition1, card);
+                            changeCostClass.SetUpChangeCostClass(changeCostFunc: ChangeCost, cardSourceCondition: CardSourceCondition, rootCondition: RootCondition, isUpDown: isUpDown, isCheckAvailability: () => false, isChangePayingCost: () => true);
+                            new Player(card.Context, card.Owner).UntilCalculateFixedCostEffect.Add((_timing) => changeCostClass);
 
-                                yield return ContinuousController.instance.StartCoroutine(CardEffectCommons.ShowReducedCost(_hashtable));
+                            await CardEffectCommons.ShowReducedCost(_hashtable);
 
-                                bool CanUseCondition1(Hashtable hashtable) => true;
+                            bool CanUseCondition1(Hashtable hashtable) => true;
 
-                                int ChangeCost(CardSource cardSource, int Cost, SelectCardEffect.Root root, List<Permanent> targetPermanents)
+                            int ChangeCost(CardSource cardSource, int Cost, SelectCardEffect.Root root, List<Permanent> targetPermanents)
+                            {
+                                if (CardSourceCondition(cardSource)
+                                && RootCondition(root)
+                                && PermanentsCondition(targetPermanents))
                                 {
-                                    if (CardSourceCondition(cardSource)
-                                    && RootCondition(root)
-                                    && PermanentsCondition(targetPermanents))
-                                    {
-                                        Cost -= 2;
-                                    }
-
-                                    return Cost;
+                                    Cost -= 2;
                                 }
 
-                                bool PermanentsCondition(List<Permanent> targetPermanents)
-                                {
-                                    return targetPermanents == null
-                                            || targetPermanents.Count((targetPermanent) => targetPermanent != null) == 0;
-                                }
+                                return Cost;
+                            }
 
-                                bool CardSourceCondition(CardSource cardSource)
-                                    => cardSource != null
-                                        && cardSource == card;
+                            bool PermanentsCondition(List<Permanent> targetPermanents)
+                            {
+                                return targetPermanents == null
+                                        || targetPermanents.Count((targetPermanent) => targetPermanent != null) == 0;
+                            }
 
-                                bool RootCondition(SelectCardEffect.Root root) => true;
+                            bool CardSourceCondition(CardSource cardSource)
+                                => cardSource != null
+                                    && cardSource == card;
 
-                                bool isUpDown() => true;
-                            }
+                            bool RootCondition(SelectCardEffect.Root root) => true;
+
+                            bool isUpDown() => true;
                         }
                     }
                 }
             }
+        }
 
-            if (timing == EffectTiming.None)
-            {
-                ChangeCostClass changeCostClass = new ChangeCostClass();
-                changeCostClass.SetUpICardEffect("Play Cost -2", CanUseCondition, card);
-                changeCostClass.SetUpChangeCostClass(changeCostFunc: ChangeCost, cardSourceCondition: CardSourceCondition, rootCondition: RootCondition, isUpDown: isUpDown, isCheckAvailability: () => true, isChangePayingCost: () => true);
-                changeCostClass.SetNotShowUI(true);
-                cardEffects.Add(changeCostClass);
+        if (timing == EffectTiming.None)
+        {
+            ChangeCostClass changeCostClass = new ChangeCostClass();
+            changeCostClass.SetUpICardEffect("Play Cost -2", CanUseCondition, card);
+            changeCostClass.SetUpChangeCostClass(changeCostFunc: ChangeCost, cardSourceCondition: CardSourceCondition, rootCondition: RootCondition, isUpDown: isUpDown, isCheckAvailability: () => true, isChangePayingCost: () => true);
+            changeCostClass.SetNotShowUI(true);
+            cardEffects.Add(changeCostClass);
 
-                bool CanUseCondition(Hashtable hashtable)
-                {
-                    return CardEffectCommons.IsExistOnHand(card)
-                        && CardEffectCommons.HasMatchConditionOwnersPermanent(card, SharedCanSelectTamerCondition);
-                }
+            bool CanUseCondition(Hashtable hashtable)
+            {
+                return CardEffectCommons.IsExistOnHand(card)
+                    && CardEffectCommons.HasMatchConditionOwnersPermanent(card, SharedCanSelectTamerCondition);
+            }
 
-                int ChangeCost(CardSource cardSource, int Cost, SelectCardEffect.Root root, List<Permanent> targetPermanents)
+            int ChangeCost(CardSource cardSource, int Cost, SelectCardEffect.Root root, List<Permanent> targetPermanents)
+            {
+                if (CardSourceCondition(cardSource))
                 {
-                    if (CardSourceCondition(cardSource))
+                    if (RootCondition(root))
                     {
-                        if (RootCondition(root))
-                        {
-                            Cost -= 2;
-                        }
+                        Cost -= 2;
                     }
-
-                    return Cost;
                 }
 
-                bool CardSourceCondition(CardSource cardSource)
-                    => cardSource != null && cardSource == card;
+                return Cost;
+            }
 
-                bool RootCondition(SelectCardEffect.Root root) => true;
+            bool CardSourceCondition(CardSource cardSource)
+                => cardSource != null && cardSource == card;
 
-                bool isUpDown() => true;
-            }
+            bool RootCondition(SelectCardEffect.Root root) => true;
 
-            #endregion
+            bool isUpDown() => true;
+        }
 
-            #region Main
+        #endregion
 
-            if (timing == EffectTiming.OptionSkill)
-            {
-                ActivateClass activateClass = new ActivateClass();
-                activateClass.SetUpICardEffect("By placing 1 [Gaogamon] and 1 [MachGaogamon] from trash as 1 [Gaomon] bottom sources, may digivolve into [MirageGaogamon] without cost or digivole requirements", CanUseCondition, card);
-                activateClass.SetUpActivateClass(null, ActivateCoroutine, -1, false, EffectDescription());
-                cardEffects.Add(activateClass);
+        #region Main
 
-                string EffectDescription()
-                {
-                    return "[Main] By placing 1 [Gaogamon] and 1 [MachGaogamon] from your trash as 1 of your [Gaomon]'s bottom digivolution cards, that Digimon may digivolve into [MirageGaogamon] in the hand, ignoring digivolution requirements and without paying the cost.";
-                }
+        if (timing == EffectTiming.OptionSkill)
+        {
+            ActivateClass activateClass = new ActivateClass();
+            activateClass.SetUpICardEffect("By placing 1 [Gaogamon] and 1 [MachGaogamon] from trash as 1 [Gaomon] bottom sources, may digivolve into [MirageGaogamon] without cost or digivole requirements", CanUseCondition, card);
+            activateClass.SetUpActivateClass(null, ActivateCoroutine, -1, false, EffectDescription());
+            cardEffects.Add(activateClass);
 
-                bool CanUseCondition(Hashtable hashtable)
-                {
-                    return CardEffectCommons.CanTriggerOptionMainEffect(hashtable, card);
-                }
+            string EffectDescription()
+            {
+                return "[Main] By placing 1 [Gaogamon] and 1 [MachGaogamon] from your trash as 1 of your [Gaomon]'s bottom digivolution cards, that Digimon may digivolve into [MirageGaogamon] in the hand, ignoring digivolution requirements and without paying the cost.";
+            }
 
-                IEnumerator ActivateCoroutine(Hashtable hashtable)
-                {
-                    #region Conditions
+            bool CanUseCondition(Hashtable hashtable)
+            {
+                return CardEffectCommons.CanTriggerOptionMainEffect(hashtable, card);
+            }
 
-                    bool IsGaomon(Permanent permanent)
-                        => CardEffectCommons.IsPermanentExistsOnOwnerBattleAreaDigimon(permanent, card)
-                            && permanent.TopCard.EqualsCardName("Gaomon");
+            async Task ActivateCoroutine(Hashtable hashtable)
+            {
+                #region Conditions
 
-                    bool IsGaogamon(CardSource cardSource)
-                        => cardSource.EqualsCardName("Gaogamon");
+                bool IsGaomon(Permanent permanent)
+                    => CardEffectCommons.IsPermanentExistsOnOwnerBattleAreaDigimon(permanent, card)
+                        && permanent.TopCard.EqualsCardName("Gaomon");
+
+                bool IsGaogamon(CardSource cardSource)
+                    => cardSource.EqualsCardName("Gaogamon");
+
+                bool IsMachGaogamon(CardSource cardSource)
+                    => cardSource.EqualsCardName("MachGaogamon");
+
+                bool isMirageGaogamon(CardSource cardSource)
+                    => cardSource.EqualsCardName("MirageGaogamon");
+
+                #endregion
+
+                if (!(CardEffectCommons.HasMatchConditionOwnersCardInTrash(card, IsGaogamon) && CardEffectCommons.HasMatchConditionOwnersCardInTrash(card, IsMachGaogamon))) return;
+                Permanent? selectedPermanent = null;
+
+                #region Select Gaomon
+                if (CardEffectCommons.MatchConditionOwnersPermanentCount(card, IsGaomon) > 1)
+                {
+                    SelectPermanentEffect selectPermanentEffect = GManager.instance.GetComponent<SelectPermanentEffect>();
+                    int maxCount = Math.Min(1, CardEffectCommons.MatchConditionOwnersPermanentCount(card, IsGaomon));
+
+                    selectPermanentEffect.SetUp(
+                        selectPlayer: card.Owner,
+                        canTargetCondition: IsGaomon,
+                        canTargetCondition_ByPreSelecetedList: null,
+                        canEndSelectCondition: null,
+                        maxCount: maxCount,
+                        canNoSelect: true,
+                        canEndNotMax: false,
+                        selectPermanentCoroutine: SelectPermanentCoroutine,
+                        afterSelectPermanentCoroutine: null,
+                        mode: SelectPermanentEffect.Mode.Custom,
+                        cardEffect: activateClass);
 
-                    bool IsMachGaogamon(CardSource cardSource)
-                        => cardSource.EqualsCardName("MachGaogamon");
+                    Task SelectPermanentCoroutine(Permanent permanent)
+                    {
+                        selectedPermanent = permanent;
+                        return Task.CompletedTask;
+                    }
 
-                    bool isMirageGaogamon(CardSource cardSource)
-                        => cardSource.EqualsCardName("MirageGaogamon");
+                    selectPermanentEffect.SetUpCustomMessage("Select 1 [Gaomon] to gain digivolution sources", "The opponent is selecting 1 [Gaomon] to gain digivolution sources.");
+                    await selectPermanentEffect.Activate();
+                }
+                if (CardEffectCommons.MatchConditionOwnersPermanentCount(card, IsGaomon) == 1)
+                    selectedPermanent = card.Owner.GetBattleAreaDigimons().Find(IsGaomon);
+                #endregion
 
-                    #endregion
+                if (selectedPermanent != null)
+                {
+                    List<CardSource> selectedTrashCards = new List<CardSource>();
 
-                    if (!(CardEffectCommons.HasMatchConditionOwnersCardInTrash(card, IsGaogamon) && CardEffectCommons.HasMatchConditionOwnersCardInTrash(card, IsMachGaogamon))) yield break;
-                    Permanent selectedPermanent = null;
+                    Task SelectCardCoroutine(CardSource cardSource)
+                    {
+                        selectedTrashCards.Add(cardSource);
+                        return Task.CompletedTask;
+                    }
 
-                    #region Select Gaomon
-                    if (CardEffectCommons.MatchConditionOwnersPermanentCount(card, IsGaomon) > 1)
+                    #region Select Gaogamon
+                    if (CardEffectCommons.HasMatchConditionOwnersCardInTrash(card, IsGaogamon))
                     {
-                        SelectPermanentEffect selectPermanentEffect = GManager.instance.GetComponent<SelectPermanentEffect>();
-                        int maxCount = Math.Min(1, CardEffectCommons.MatchConditionOwnersPermanentCount(card, IsGaomon));
+                        SelectCardEffect selectCardEffect = GManager.instance.GetComponent<SelectCardEffect>();
+                        int maxCount = Math.Min(1, CardEffectCommons.MatchConditionOwnersCardCountInTrash(card, IsGaogamon));
 
-                        selectPermanentEffect.SetUp(
-                            selectPlayer: card.Owner,
-                            canTargetCondition: IsGaomon,
+                        selectCardEffect.SetUp(
+                            canTargetCondition: IsGaogamon,
                             canTargetCondition_ByPreSelecetedList: null,
                             canEndSelectCondition: null,
+                            canNoSelect: () => true,
+                            selectCardCoroutine: SelectCardCoroutine,
+                            afterSelectCardCoroutine: null,
+                            message: "Select 1 [Gaogamon] card",
                             maxCount: maxCount,
-                            canNoSelect: true,
                             canEndNotMax: false,
-                            selectPermanentCoroutine: SelectPermanentCoroutine,
-                            afterSelectPermanentCoroutine: null,
-                            mode: SelectPermanentEffect.Mode.Custom,
+                            isShowOpponent: true,
+                            mode: SelectCardEffect.Mode.Custom,
+                            root: SelectCardEffect.Root.Trash,
+                            customRootCardList: null,
+                            canLookReverseCard: true,
+                            selectPlayer: card.Owner,
                             cardEffect: activateClass);
 
-                        IEnumerator SelectPermanentCoroutine(Permanent permanent)
-                        {
-                            selectedPermanent = permanent;
-                            yield return null;
-                        }
-
-                        selectPermanentEffect.SetUpCustomMessage("Select 1 [Gaomon] to gain digivolution sources", "The opponent is selecting 1 [Gaomon] to gain digivolution sources.");
-                        yield return ContinuousController.instance.StartCoroutine(selectPermanentEffect.Activate());
+                        selectCardEffect.SetUpCustomMessage("Select 1 [Gaogamon] card to place under [Gaomon].", "The opponent is selecting a [Gaogamon] card to place under [Gaomon].");
+                        selectCardEffect.SetUpCustomMessage_ShowCard("Selected [Gaogamon]");
+                        await selectCardEffect.Activate();
                     }
-                    if (CardEffectCommons.MatchConditionOwnersPermanentCount(card, IsGaomon) == 1)
-                        selectedPermanent = card.Owner.GetBattleAreaDigimons().Find(IsGaomon);
                     #endregion
 
-                    if (selectedPermanent != null)
+                    #region Select MachGaogamon
+                    if (CardEffectCommons.HasMatchConditionOwnersCardInTrash(card, IsMachGaogamon) && selectedTrashCards.Find(IsGaogamon) != null)
                     {
-                        List<CardSource> selectedTrashCards = new List<CardSource>();
-
-                        IEnumerator SelectCardCoroutine(CardSource cardSource)
-                        {
-                            selectedTrashCards.Add(cardSource);
-                            yield return null;
-                        }
-
-                        #region Select Gaogamon
-                        if (CardEffectCommons.HasMatchConditionOwnersCardInTrash(card, IsGaogamon))
-                        {
-                            SelectCardEffect selectCardEffect = GManager.instance.GetComponent<SelectCardEffect>();
-                            int maxCount = Math.Min(1, CardEffectCommons.MatchConditionOwnersCardCountInTrash(card, IsGaogamon));
-
-                            selectCardEffect.SetUp(
-                                canTargetCondition: IsGaogamon,
-                                canTargetCondition_ByPreSelecetedList: null,
-                                canEndSelectCondition: null,
-                                canNoSelect: () => true,
-                                selectCardCoroutine: SelectCardCoroutine,
-                                afterSelectCardCoroutine: null,
-                                message: "Select 1 [Gaogamon] card",
-                                maxCount: maxCount,
-                                canEndNotMax: false,
-                                isShowOpponent: true,
-                                mode: SelectCardEffect.Mode.Custom,
-                                root: SelectCardEffect.Root.Trash,
-                                customRootCardList: null,
-                                canLookReverseCard: true,
-                                selectPlayer: card.Owner,
-                                cardEffect: activateClass);
-
-                            selectCardEffect.SetUpCustomMessage("Select 1 [Gaogamon] card to place under [Gaomon].", "The opponent is selecting a [Gaogamon] card to place under [Gaomon].");
-                            selectCardEffect.SetUpCustomMessage_ShowCard("Selected [Gaogamon]");
-                            yield return ContinuousController.instance.StartCoroutine(selectCardEffect.Activate());
-                        }
-                        #endregion
+                        SelectCardEffect selectCardEffect = GManager.instance.GetComponent<SelectCardEffect>();
+                        int maxCount = Math.Min(1, CardEffectCommons.MatchConditionOwnersCardCountInTrash(card, IsMachGaogamon));
 
-                        #region Select MachGaogamon
-                        if (CardEffectCommons.HasMatchConditionOwnersCardInTrash(card, IsMachGaogamon) && selectedTrashCards.Find(IsGaogamon))
-                        {
-                            SelectCardEffect selectCardEffect = GManager.instance.GetComponent<SelectCardEffect>();
-                            int maxCount = Math.Min(1, CardEffectCommons.MatchConditionOwnersCardCountInTrash(card, IsMachGaogamon));
-
-                            selectCardEffect.SetUp(
-                                canTargetCondition: IsMachGaogamon,
-                                canTargetCondition_ByPreSelecetedList: null,
-                                canEndSelectCondition: null,
-                                canNoSelect: () => true,
-                                selectCardCoroutine: SelectCardCoroutine,
-                                afterSelectCardCoroutine: null,
-                                message: "Select 1 [MachGaogamon] card",
-                                maxCount: maxCount,
-                                canEndNotMax: false,
-                                isShowOpponent: true,
-                                mode: SelectCardEffect.Mode.Custom,
-                                root: SelectCardEffect.Root.Trash,
-                                customRootCardList: null,
-                                canLookReverseCard: true,
-                                selectPlayer: card.Owner,
-                                cardEffect: activateClass);
+                        selectCardEffect.SetUp(
+                            canTargetCondition: IsMachGaogamon,
+                            canTargetCondition_ByPreSelecetedList: null,
+                            canEndSelectCondition: null,
+                            canNoSelect: () => true,
+                            selectCardCoroutine: SelectCardCoroutine,
+                            afterSelectCardCoroutine: null,
+                            message: "Select 1 [MachGaogamon] card",
+                            maxCount: maxCount,
+                            canEndNotMax: false,
+                            isShowOpponent: true,
+                            mode: SelectCardEffect.Mode.Custom,
+                            root: SelectCardEffect.Root.Trash,
+                            customRootCardList: null,
+                            canLookReverseCard: true,
+                            selectPlayer: card.Owner,
+                            cardEffect: activateClass);
 
-                            selectCardEffect.SetUpCustomMessage("Select 1 [MachGaogamon] card to place under [Gaomon].", "The opponent is selecting a [MachGaogamon] card to place under [Gaomon].");
-                            selectCardEffect.SetUpCustomMessage_ShowCard("Selected [MachGaogamon]");
-                            yield return ContinuousController.instance.StartCoroutine(selectCardEffect.Activate());
-                        }
-                        #endregion
+                        selectCardEffect.SetUpCustomMessage("Select 1 [MachGaogamon] card to place under [Gaomon].", "The opponent is selecting a [MachGaogamon] card to place under [Gaomon].");
+                        selectCardEffect.SetUpCustomMessage_ShowCard("Selected [MachGaogamon]");
+                        await selectCardEffect.Activate();
+                    }
+                    #endregion
 
-                        if (selectedTrashCards.Find(IsGaogamon) && selectedTrashCards.Find(IsMachGaogamon))
-                        {
-                            yield return ContinuousController.instance.StartCoroutine(selectedPermanent.AddDigivolutionCardsBottom(
-                                addedDigivolutionCards: selectedTrashCards,
-                                cardEffect: activateClass));
-
-                            yield return ContinuousController.instance.StartCoroutine(CardEffectCommons.DigivolveIntoHandOrTrashCard(
-                                targetPermanent: selectedPermanent,
-                                cardCondition: isMirageGaogamon,
-                                payCost: false,
-                                reduceCostTuple: null,
-                                fixedCostTuple: null,
-                                ignoreDigivolutionRequirementFixedCost: 1,
-                                isHand: true,
-                                activateClass: activateClass,
-                                successProcess: null));
+                    if (selectedTrashCards.Find(IsGaogamon) != null && selectedTrashCards.Find(IsMachGaogamon) != null)
+                    {
+                        await selectedPermanent.AddDigivolutionCardsBottom(
+                            addedDigivolutionCards: selectedTrashCards,
+                            causeEffectSourceId: activateClass.EffectSourceCard?.InstanceId);
+
+                        await CardEffectCommons.DigivolveIntoHandOrTrashCard(
+                            targetPermanent: selectedPermanent,
+                            cardCondition: isMirageGaogamon,
+                            payCost: false,
+                            reduceCostTuple: null,
+                            fixedCostTuple: null,
+                            ignoreDigivolutionRequirementFixedCost: 1,
+                            isHand: true,
+                            activateClass: activateClass,
+                            successProcess: null);
 
-                        }
                     }
-
                 }
+
             }
+        }
+
+        #endregion
 
-            #endregion
+        #region Security
+        if (timing == EffectTiming.SecuritySkill)
+        {
+            ActivateClass activateClass = new ActivateClass();
+            activateClass.SetUpICardEffect($"Play 1 [Gaomon]/[Thomas H. Norstein] from hand or trash, then add this to hand", CanUseCondition, card);
+            activateClass.SetUpActivateClass(null, ActivateCoroutine, -1, false, EffectDescription());
+            activateClass.SetIsSecurityEffect(true);
+            cardEffects.Add(activateClass);
 
-            #region Security
-            if (timing == EffectTiming.SecuritySkill)
+            string EffectDescription()
             {
-                ActivateClass activateClass = new ActivateClass();
-                activateClass.SetUpICardEffect($"Play 1 [Gaomon]/[Thomas H. Norstein] from hand or trash, then add this to hand", CanUseCondition, card);
-                activateClass.SetUpActivateClass(null, ActivateCoroutine, -1, false, EffectDescription());
-                activateClass.SetIsSecurityEffect(true);
-                cardEffects.Add(activateClass);
+                return "[Security] You may play 1 [Gaomon] or [Thomas H. Norstein] from your hand or trash without paying the cost. Then, add this card to the hand.";
+            }
 
-                string EffectDescription()
-                {
-                    return "[Security] You may play 1 [Gaomon] or [Thomas H. Norstein] from your hand or trash without paying the cost. Then, add this card to the hand.";
-                }
+            bool CanSelectCardCondition(CardSource cardSource)
+            {
+                return (cardSource.EqualsCardName("Gaomon") || cardSource.EqualsCardName("Thomas H. Norstein"))
+                    && CardEffectCommons.CanPlayAsNewPermanent(cardSource, false, activateClass);
+            }
 
-                bool CanSelectCardCondition(CardSource cardSource)
-                {
-                    return (cardSource.EqualsCardName("Gaomon") || cardSource.EqualsCardName("Thomas H. Norstein"))
-                        && CardEffectCommons.CanPlayAsNewPermanent(cardSource, false, activateClass);
-                }
+            bool CanUseCondition(Hashtable hashtable) => CardEffectCommons.CanTriggerSecurityEffect(hashtable, card);
 
-                bool CanUseCondition(Hashtable hashtable) => CardEffectCommons.CanTriggerSecurityEffect(hashtable, card);
+            async Task ActivateCoroutine(Hashtable _hashtable)
+            {
+                bool canPlayFromHand = CardEffectCommons.HasMatchConditionOwnersHand(card, CanSelectCardCondition);
+                bool canPlayFromTrash = CardEffectCommons.HasMatchConditionOwnersCardInTrash(card, CanSelectCardCondition);
 
-                IEnumerator ActivateCoroutine(Hashtable _hashtable)
+                if (canPlayFromHand || canPlayFromTrash)
                 {
-                    bool canPlayFromHand = CardEffectCommons.HasMatchConditionOwnersHand(card, CanSelectCardCondition);
-                    bool canPlayFromTrash = CardEffectCommons.HasMatchConditionOwnersCardInTrash(card, CanSelectCardCondition);
 
-                    if (canPlayFromHand || canPlayFromTrash)
-                    {
+                    List<SelectionElement<int>> selectionElements = new List<SelectionElement<int>>();
 
-                        List<SelectionElement<int>> selectionElements = new List<SelectionElement<int>>();
+                    if (canPlayFromHand) selectionElements.Add(new SelectionElement<int>("Play from hand", 1, 0));
+                    if (canPlayFromTrash) selectionElements.Add(new SelectionElement<int>("Play from trash", 2, 0));
+                    selectionElements.Add(new SelectionElement<int>("Dont play a card", 3, 1));
 
-                        if (canPlayFromHand) selectionElements.Add(new SelectionElement<int>("Play from hand", 1, 0));
-                        if (canPlayFromTrash) selectionElements.Add(new SelectionElement<int>("Play from trash", 2, 0));
-                        selectionElements.Add(new SelectionElement<int>("Dont play a card", 3, 1));
 
+                    string selectPlayerMessage = "Will you play a [Gaomon] or [Thomas H. Norstein]?";
+                    string notSelectPlayerMessage = "The opponent is choosing whether to play a [Gaomon] or [Thomas H. Norstein]";
 
-                        string selectPlayerMessage = "Will you play a [Gaomon] or [Thomas H. Norstein]?";
-                        string notSelectPlayerMessage = "The opponent is choosing whether to play a [Gaomon] or [Thomas H. Norstein]";
 
+                    GManager.instance.userSelectionManager.SetIntSelection(selectionElements: selectionElements, selectPlayer: card.Owner, selectPlayerMessage: selectPlayerMessage, notSelectPlayerMessage: notSelectPlayerMessage);
+                    await GManager.instance.userSelectionManager.WaitForEndSelect();
 
-                        GManager.instance.userSelectionManager.SetIntSelection(selectionElements: selectionElements, selectPlayer: card.Owner, selectPlayerMessage: selectPlayerMessage, notSelectPlayerMessage: notSelectPlayerMessage);
-                        yield return ContinuousController.instance.StartCoroutine(GManager.instance.userSelectionManager.WaitForEndSelect());
+                    bool playFromHand = GManager.instance.userSelectionManager.SelectedIntValue == 1;
+                    bool dontPlay = GManager.instance.userSelectionManager.SelectedIntValue == 3;
 
-                        bool playFromHand = GManager.instance.userSelectionManager.SelectedIntValue == 1;
-                        bool dontPlay = GManager.instance.userSelectionManager.SelectedIntValue == 3;
+                    if (!dontPlay)
+                    {
+                        CardSource? selectedCard = null;
+                        SelectCardEffect.Root selectedRoot = SelectCardEffect.Root.None;
 
-                        if (!dontPlay)
+                        Task SelectCardCoroutine(CardSource cardSource)
                         {
-                            CardSource selectedCard = null;
-                            SelectCardEffect.Root selectedRoot = SelectCardEffect.Root.None;
-
-                            IEnumerator SelectCardCoroutine(CardSource cardSource)
-                            {
-                                selectedCard = cardSource;
-                                yield return null;
-                            }
+                            selectedCard = cardSource;
+                            return Task.CompletedTask;
+                        }
 
-                            if (playFromHand)
-                            {
-                                int maxCount = Math.Min(1, CardEffectCommons.MatchConditionOwnersCardCountInHand(card, CanSelectCardCondition));
-                                SelectHandEffect selectHandEffect = GManager.instance.GetComponent<SelectHandEffect>();
+                        if (playFromHand)
+                        {
+                            int maxCount = Math.Min(1, CardEffectCommons.MatchConditionOwnersCardCountInHand(card, CanSelectCardCondition));
+                            SelectHandEffect selectHandEffect = GManager.instance.GetComponent<SelectHandEffect>();
 
-                                selectHandEffect.SetUp(
-                                    selectPlayer: card.Owner,
-                                    canTargetCondition: CanSelectCardCondition,
-                                    canTargetCondition_ByPreSelecetedList: null,
-                                    canEndSelectCondition: null,
-                                    maxCount: maxCount,
-                                    canNoSelect: true,
-                                    canEndNotMax: false,
-                                    isShowOpponent: true,
-                                    selectCardCoroutine: SelectCardCoroutine,
-                                    afterSelectCardCoroutine: null,
-                                    mode: SelectHandEffect.Mode.Custom,
-                                    cardEffect: activateClass);
-
-                                selectHandEffect.SetUpCustomMessage("Select 1 [Gaomon] or [Thomas H. Norstein] to play.", "The opponent is selecting 1 [Gaomon] or [Thomas H. Norstein] to play.");
-                                selectHandEffect.SetUpCustomMessage_ShowCard("Selected Card");
-                                yield return ContinuousController.instance.StartCoroutine(selectHandEffect.Activate());
-                                selectedRoot = SelectCardEffect.Root.Hand;
-                            }
-                            else
-                            {
-                                int maxCount = Math.Min(1, CardEffectCommons.MatchConditionOwnersCardCountInTrash(card, CanSelectCardCondition));
-                                SelectCardEffect selectCardEffect = GManager.instance.GetComponent<SelectCardEffect>();
+                            selectHandEffect.SetUp(
+                                selectPlayer: card.Owner,
+                                canTargetCondition: CanSelectCardCondition,
+                                canTargetCondition_ByPreSelecetedList: null,
+                                canEndSelectCondition: null,
+                                maxCount: maxCount,
+                                canNoSelect: true,
+                                canEndNotMax: false,
+                                isShowOpponent: true,
+                                selectCardCoroutine: SelectCardCoroutine,
+                                afterSelectCardCoroutine: null,
+                                mode: SelectHandEffect.Mode.Custom,
+                                cardEffect: activateClass);
 
-                                selectCardEffect.SetUp(
-                                            canTargetCondition: CanSelectCardCondition,
-                                            canTargetCondition_ByPreSelecetedList: null,
-                                            canEndSelectCondition: null,
-                                            canNoSelect: () => true,
-                                            selectCardCoroutine: SelectCardCoroutine,
-                                            afterSelectCardCoroutine: null,
-                                            message: "Select 1 [Gaomon] or [Thomas H. Norstein] to play",
-                                            maxCount: maxCount,
-                                            canEndNotMax: false,
-                                            isShowOpponent: true,
-                                            mode: SelectCardEffect.Mode.Custom,
-                                            root: SelectCardEffect.Root.Trash,
-                                            customRootCardList: null,
-                                            canLookReverseCard: true,
-                                            selectPlayer: card.Owner,
-                                            cardEffect: activateClass);
-
-                                selectCardEffect.SetUpCustomMessage("Select 1 [Gaomon] or [Thomas H. Norstein] to play.", "The opponent is selecting 1 [Gaomon] or [Thomas H. Norstein] to play.");
-                                selectCardEffect.SetUpCustomMessage_ShowCard("Selected Card");
-                                yield return ContinuousController.instance.StartCoroutine(selectCardEffect.Activate());
-                                selectedRoot = SelectCardEffect.Root.Trash;
-                            }
+                            selectHandEffect.SetUpCustomMessage("Select 1 [Gaomon] or [Thomas H. Norstein] to play.", "The opponent is selecting 1 [Gaomon] or [Thomas H. Norstein] to play.");
+                            selectHandEffect.SetUpCustomMessage_ShowCard("Selected Card");
+                            await selectHandEffect.Activate();
+                            selectedRoot = SelectCardEffect.Root.Hand;
+                        }
+                        else
+                        {
+                            int maxCount = Math.Min(1, CardEffectCommons.MatchConditionOwnersCardCountInTrash(card, CanSelectCardCondition));
+                            SelectCardEffect selectCardEffect = GManager.instance.GetComponent<SelectCardEffect>();
 
-                            if (selectedCard != null && selectedRoot != SelectCardEffect.Root.None)
-                                yield return ContinuousController.instance.StartCoroutine(CardEffectCommons.PlayPermanentCards(
-                                    cardSources: new List<CardSource>() { selectedCard },
-                                    activateClass: activateClass,
-                                    payCost: false,
-                                    isTapped: false,
-                                    root: selectedRoot,
-                                    activateETB: true
-                                ));
+                            selectCardEffect.SetUp(
+                                        canTargetCondition: CanSelectCardCondition,
+                                        canTargetCondition_ByPreSelecetedList: null,
+                                        canEndSelectCondition: null,
+                                        canNoSelect: () => true,
+                                        selectCardCoroutine: SelectCardCoroutine,
+                                        afterSelectCardCoroutine: null,
+                                        message: "Select 1 [Gaomon] or [Thomas H. Norstein] to play",
+                                        maxCount: maxCount,
+                                        canEndNotMax: false,
+                                        isShowOpponent: true,
+                                        mode: SelectCardEffect.Mode.Custom,
+                                        root: SelectCardEffect.Root.Trash,
+                                        customRootCardList: null,
+                                        canLookReverseCard: true,
+                                        selectPlayer: card.Owner,
+                                        cardEffect: activateClass);
+
+                            selectCardEffect.SetUpCustomMessage("Select 1 [Gaomon] or [Thomas H. Norstein] to play.", "The opponent is selecting 1 [Gaomon] or [Thomas H. Norstein] to play.");
+                            selectCardEffect.SetUpCustomMessage_ShowCard("Selected Card");
+                            await selectCardEffect.Activate();
+                            selectedRoot = SelectCardEffect.Root.Trash;
+                        }
 
+                        if (selectedCard != null && selectedRoot != SelectCardEffect.Root.None)
+                            await CardEffectCommons.PlayPermanentCards(
+                                cardSources: new List<CardSource>() { selectedCard },
+                                activateClass: activateClass,
+                                payCost: false,
+                                isTapped: false,
+                                root: selectedRoot,
+                                activateETB: true
+                            );
 
 
-                        }
 
                     }
 
-                    yield return ContinuousController.instance.StartCoroutine(CardEffectCommons.AddThisCardToHand(card, activateClass));
                 }
-            }
-            #endregion
 
-            return cardEffects;
+                await CardEffectCommons.AddThisCardToHand(card, activateClass.EffectSourceCard);
+            }
         }
+        #endregion
+
+        return cardEffects;
     }
 }
