--- DCGO/Assets/Scripts/CardEffect/EX7/Purple/EX7_058.cs	2026-05-18 20:03:15.401854400 +0900
+++ src/HeadlessDCGO.Engine/Assets/Scripts/CardEffect/EX7/Purple/EX7_058.cs	2026-07-29 01:02:23.113943027 +0900
@@ -1,482 +1,473 @@
-using System;
+namespace HeadlessDCGO.Engine.Assets.Scripts.CardEffect.EX7.Purple;
+
 using System.Collections;
-using System.Collections.Generic;
-using System.Linq;
+using System.Threading.Tasks;
+using HeadlessDCGO.Engine.Assets.Scripts.Script;
+using HeadlessDCGO.Engine.Assets.Scripts.Script.CardEffectCommons;
+using HeadlessDCGO.Engine.Assets.Scripts.Script.CardEffects;
+using HeadlessDCGO.Engine.Headless.Services;
 
-namespace DCGO.CardEffects.EX7
+public class EX7_058 : CEntity_Effect
 {
-    public class EX7_058 : CEntity_Effect
+    public override List<ICardEffect> CardEffects(EffectTiming timing, CardSource card)
     {
-        public override List<ICardEffect> CardEffects(EffectTiming timing, CardSource card)
+        List<ICardEffect> cardEffects = new List<ICardEffect>();
+
+        #region Digivolution Condition
+        if (timing == EffectTiming.None)
         {
-            List<ICardEffect> cardEffects = new List<ICardEffect>();
+            bool PermanentCondition(Permanent targetPermanent)
+            {
+                return targetPermanent.TopCard.CardNames.Contains("LadyDevimon");
+            }
+
+            cardEffects.Add(CardEffectFactory.AddSelfDigivolutionRequirementStaticEffect(permanentCondition: PermanentCondition, digivolutionCost: 0, ignoreDigivolutionRequirement: false, card: card, condition: null));
+        }
+        #endregion
 
-            #region Digivolution Condition
-            if (timing == EffectTiming.None)
+        #region Shared On Play/When Digivolving
+        bool CanSelectPermanentCondition(Permanent permanent)
+        {
+            return CardEffectCommons.IsPermanentExistsOnOpponentBattleAreaDigimon(permanent, card);
+        }
+
+        bool CanActivateSecondEffectCondition(Hashtable hashtable)
+        {
+            if (CardEffectCommons.IsExistOnBattleAreaDigimon(card))
             {
-                bool PermanentCondition(Permanent targetPermanent)
+                if (ICardEffect.ResolvePermanentOfThisCard(card).DigivolutionCards.Count((cardSource) => cardSource.CardNames.Contains("LadyDevimon") || cardSource.CardNames.Contains("X Antibody") || cardSource.CardNames.Contains("XAntibody")) >= 1)
                 {
-                    return targetPermanent.TopCard.CardNames.Contains("LadyDevimon");
+                    return true;
                 }
-
-                cardEffects.Add(CardEffectFactory.AddSelfDigivolutionRequirementStaticEffect(permanentCondition: PermanentCondition, digivolutionCost: 0, ignoreDigivolutionRequirement: false, card: card, condition: null));
             }
-            #endregion
 
-            #region Shared On Play/When Digivolving
-            bool CanSelectPermanentCondition(Permanent permanent)
+            return false;
+        }
+        #endregion
+
+        #region On Play
+        if (timing == EffectTiming.OnEnterFieldAnyone)
+        {
+            ActivateClass activateClass = new ActivateClass();
+            activateClass.SetUpICardEffect("Give effects to opponent's Digimon and play a Token", CanUseCondition, card);
+            activateClass.SetUpActivateClass(CanActivateCondition, ActivateCoroutine, -1, false, EffectDiscription());
+            cardEffects.Add(activateClass);
+
+            string EffectDiscription()
             {
-                return CardEffectCommons.IsPermanentExistsOnOpponentBattleAreaDigimon(permanent, card);
+                return "[On Play] 1 of your opponent's Digimon gains \"[End of Attack] Delete this Digimon.\" until the end of their turn. Then, if this Digimon has [LadyDevimon]/[X Antibody] in its digivolution cards, you may play 1 [Volée & Zerdrücken] Token (Digimon/Lv.4/Purple/5000 DP/<Blocker>/<Retaliation>).";
             }
 
-            bool CanActivateSecondEffectCondition(Hashtable hashtable)
+            bool CanUseCondition(Hashtable hashtable)
             {
                 if (CardEffectCommons.IsExistOnBattleAreaDigimon(card))
-                {
-                    if (card.PermanentOfThisCard().DigivolutionCards.Count((cardSource) => cardSource.CardNames.Contains("LadyDevimon") || cardSource.CardNames.Contains("X Antibody") || cardSource.CardNames.Contains("XAntibody")) >= 1)
-                    {
-                        return true;
-                    }
-                }
+                    return CardEffectCommons.CanTriggerOnPlay(hashtable, card);
 
                 return false;
             }
-            #endregion
 
-            #region On Play
-            if (timing == EffectTiming.OnEnterFieldAnyone)
+            bool CanActivateCondition(Hashtable hashtable)
             {
-                ActivateClass activateClass = new ActivateClass();
-                activateClass.SetUpICardEffect("Give effects to opponent's Digimon and play a Token", CanUseCondition, card);
-                activateClass.SetUpActivateClass(CanActivateCondition, ActivateCoroutine, -1, false, EffectDiscription());
-                cardEffects.Add(activateClass);
+                return CardEffectCommons.IsExistOnBattleAreaDigimon(card);
+            }
 
-                string EffectDiscription()
+            async Task ActivateCoroutine(Hashtable hashtable)
+            {
+                if (CardEffectCommons.HasMatchConditionPermanent(card, CanSelectPermanentCondition))
                 {
-                    return "[On Play] 1 of your opponent's Digimon gains \"[End of Attack] Delete this Digimon.\" until the end of their turn. Then, if this Digimon has [LadyDevimon]/[X Antibody] in its digivolution cards, you may play 1 [Volée & Zerdrücken] Token (Digimon/Lv.4/Purple/5000 DP/<Blocker>/<Retaliation>).";
-                }
+                    int maxCount = Math.Min(1, CardEffectCommons.MatchConditionPermanentCount(card, CanSelectPermanentCondition));
 
-                bool CanUseCondition(Hashtable hashtable)
-                {
-                    if (CardEffectCommons.IsExistOnBattleAreaDigimon(card))
-                        return CardEffectCommons.CanTriggerOnPlay(hashtable, card);
+                    SelectPermanentEffect selectPermanentEffect = GManager.instance.GetComponent<SelectPermanentEffect>();
 
-                    return false;
-                }
+                    selectPermanentEffect.SetUp(
+                        selectPlayer: card.Owner,
+                        canTargetCondition: CanSelectPermanentCondition,
+                        canTargetCondition_ByPreSelecetedList: null,
+                        canEndSelectCondition: null,
+                        maxCount: maxCount,
+                        canNoSelect: false,
+                        canEndNotMax: false,
+                        selectPermanentCoroutine: SelectPermanentCoroutine,
+                        afterSelectPermanentCoroutine: null,
+                        mode: SelectPermanentEffect.Mode.Custom,
+                        cardEffect: activateClass);
+
+                    selectPermanentEffect.SetUpCustomMessage(
+                        "Select 1 Digimon that will get [End of Attack] Delete this Digimon.",
+                        "The opponent is selecting 1 Digimon that will get [End of Attack] Delete this Digimon.");
 
-                bool CanActivateCondition(Hashtable hashtable)
-                {
-                    return CardEffectCommons.IsExistOnBattleAreaDigimon(card);
-                }
+                    await selectPermanentEffect.Activate();
 
-                IEnumerator ActivateCoroutine(Hashtable hashtable)
-                {
-                    if (CardEffectCommons.HasMatchConditionPermanent(CanSelectPermanentCondition))
+                    Task SelectPermanentCoroutine(Permanent permanent)
                     {
-                        int maxCount = Math.Min(1, CardEffectCommons.MatchConditionPermanentCount(CanSelectPermanentCondition));
-
-                        SelectPermanentEffect selectPermanentEffect = GManager.instance.GetComponent<SelectPermanentEffect>();
-
-                        selectPermanentEffect.SetUp(
-                            selectPlayer: card.Owner,
-                            canTargetCondition: CanSelectPermanentCondition,
-                            canTargetCondition_ByPreSelecetedList: null,
-                            canEndSelectCondition: null,
-                            maxCount: maxCount,
-                            canNoSelect: false,
-                            canEndNotMax: false,
-                            selectPermanentCoroutine: SelectPermanentCoroutine,
-                            afterSelectPermanentCoroutine: null,
-                            mode: SelectPermanentEffect.Mode.Custom,
-                            cardEffect: activateClass);
-
-                        selectPermanentEffect.SetUpCustomMessage(
-                            "Select 1 Digimon that will get [End of Attack] Delete this Digimon.",
-                            "The opponent is selecting 1 Digimon that will get [End of Attack] Delete this Digimon.");
-
-                        yield return ContinuousController.instance.StartCoroutine(selectPermanentEffect.Activate());
-
-                        IEnumerator SelectPermanentCoroutine(Permanent permanent)
+                        if (permanent != null)
                         {
-                            if (permanent != null)
-                            {
-                                yield return ContinuousController.instance.StartCoroutine(GManager.instance.GetComponent<Effects>().CreateDebuffEffect(permanent));
-                                ActivateClass activateClass1 = new ActivateClass();
-                                activateClass1.SetUpICardEffect("Delete this Digimon", CanUseCondition1, permanent.TopCard);
-                                activateClass1.SetUpActivateClass(CanActivateCondition1, ActivateCoroutine1, -1, false, EffectDiscription1());
-                                activateClass1.SetEffectSourcePermanent(permanent);
-                                permanent.UntilOwnerTurnEndEffects.Add(GetCardEffect);
-
-                                if (!permanent.TopCard.CanNotBeAffected(activateClass))
-                                {
-                                    yield return ContinuousController.instance.StartCoroutine(GManager.instance.GetComponent<Effects>().CreateDebuffEffect(permanent));
-                                }
-
-                                string EffectDiscription1()
-                                {
-                                    return "[End of Attack] Delete this Digimon.";
-                                }
-
-                                bool CanUseCondition1(Hashtable hashtable1)
-                                {
-                                    if (CardEffectCommons.IsOpponentTurn(card))
-                                    {
-                                        if (CardEffectCommons.IsPermanentExistsOnOwnerBattleArea(permanent, permanent.TopCard))
-                                        {
-                                            if (CardEffectCommons.CanTriggerOnEndAttack(hashtable1, permanent.TopCard))
-                                            {
-                                                return true;
-                                            }
-                                        }
-                                    }                                    
+                            ActivateClass activateClass1 = new ActivateClass();
+                            activateClass1.SetUpICardEffect("Delete this Digimon", CanUseCondition1, permanent.TopCard);
+                            activateClass1.SetUpActivateClass(CanActivateCondition1, ActivateCoroutine1, -1, false, EffectDiscription1());
+                            activateClass1.SetEffectSourcePermanent(permanent);
+                            permanent.UntilOwnerTurnEndEffects.Add(GetCardEffect);
 
-                                    return false;
-                                }
+                            string EffectDiscription1()
+                            {
+                                return "[End of Attack] Delete this Digimon.";
+                            }
 
-                                bool CanActivateCondition1(Hashtable hashtable1)
+                            bool CanUseCondition1(Hashtable hashtable1)
+                            {
+                                if (CardEffectCommons.IsOpponentTurn(card))
                                 {
-                                    if (CardEffectCommons.IsPermanentExistsOnBattleArea(permanent))
+                                    if (CardEffectCommons.IsPermanentExistsOnOwnerBattleArea(permanent, permanent.TopCard))
                                     {
-                                        if (!permanent.TopCard.CanNotBeAffected(activateClass))
+                                        if (CardEffectCommons.CanTriggerOnEndAttack(hashtable1, permanent.TopCard))
                                         {
                                             return true;
                                         }
                                     }
-
-                                    return false;
                                 }
 
-                                IEnumerator ActivateCoroutine1(Hashtable _hashtable1)
+                                return false;
+                            }
+
+                            bool CanActivateCondition1(Hashtable hashtable1)
+                            {
+                                if (CardEffectCommons.IsPermanentExistsOnBattleArea(permanent))
                                 {
-                                    if (CardEffectCommons.IsPermanentExistsOnBattleArea(permanent))
+                                    if (!permanent.TopCard.CanNotBeAffected(activateClass))
                                     {
-                                        yield return ContinuousController.instance.StartCoroutine(new DestroyPermanentsClass(
-                                        new List<Permanent>() { permanent },
-                                        CardEffectCommons.CardEffectHashtable(activateClass1)).Destroy());
+                                        return true;
                                     }
                                 }
 
-                                ICardEffect GetCardEffect(EffectTiming _timing)
+                                return false;
+                            }
+
+                            async Task ActivateCoroutine1(Hashtable _hashtable1)
+                            {
+                                if (CardEffectCommons.IsPermanentExistsOnBattleArea(permanent))
                                 {
-                                    if (_timing == EffectTiming.OnEndAttack)
-                                    {
-                                        return activateClass1;
-                                    }
+                                    await new DestroyPermanentsClass(
+                                    new List<Permanent>() { permanent },
+                                    CardEffectCommons.CardEffectHashtable(activateClass1)).Destroy();
+                                }
+                            }
 
-                                    return null;
+                            ICardEffect GetCardEffect(EffectTiming _timing)
+                            {
+                                if (_timing == EffectTiming.OnEndAttack)
+                                {
+                                    return activateClass1;
                                 }
 
+                                return null;
                             }
                         }
+
+                        return Task.CompletedTask;
                     }
+                }
 
-                    if (CanActivateSecondEffectCondition(hashtable))
+                if (CanActivateSecondEffectCondition(hashtable))
+                {
+                    List<SelectionElement<bool>> selectionElements = new List<SelectionElement<bool>>()
                     {
-                        List<SelectionElement<bool>> selectionElements = new List<SelectionElement<bool>>()
-                        {
-                            new SelectionElement<bool>(message: $"Play Token", value : true, spriteIndex: 0),
-                            new SelectionElement<bool>(message: $"Not Play Token", value : false, spriteIndex: 1),
-                        };
+                        new SelectionElement<bool>(message: $"Play Token", value : true, spriteIndex: 0),
+                        new SelectionElement<bool>(message: $"Not Play Token", value : false, spriteIndex: 1),
+                    };
 
-                        string selectPlayerMessage = "Will you play a token?";
-                        string notSelectPlayerMessage = "The opponent is choosing to play a token.";
+                    string selectPlayerMessage = "Will you play a token?";
+                    string notSelectPlayerMessage = "The opponent is choosing to play a token.";
 
-                        GManager.instance.userSelectionManager.SetBoolSelection(selectionElements: selectionElements, selectPlayer: card.Owner, selectPlayerMessage: selectPlayerMessage, notSelectPlayerMessage: notSelectPlayerMessage);
+                    GManager.instance.userSelectionManager.SetBoolSelection(selectionElements: selectionElements, selectPlayer: card.Owner, selectPlayerMessage: selectPlayerMessage, notSelectPlayerMessage: notSelectPlayerMessage);
 
-                        yield return ContinuousController.instance.StartCoroutine(GManager.instance.userSelectionManager.WaitForEndSelect());
+                    await GManager.instance.userSelectionManager.WaitForEndSelect();
 
-                        bool canPlay = GManager.instance.userSelectionManager.SelectedBoolValue;
+                    bool canPlay = GManager.instance.userSelectionManager.SelectedBoolValue;
 
-                        if (canPlay)
-                            yield return ContinuousController.instance.StartCoroutine(CardEffectCommons.PlayVoleeZerdrucken(activateClass));
-                    }
+                    if (canPlay)
+                        await CardEffectCommons.PlayVoleeZerdrucken(activateClass);
                 }
             }
-            #endregion
+        }
+        #endregion
 
-            #region When Digivolving
-            if (timing == EffectTiming.OnEnterFieldAnyone)
-            {
-                ActivateClass activateClass = new ActivateClass();
-                activateClass.SetUpICardEffect("Give effects to opponent's Digimon and play a Token", CanUseCondition, card);
-                activateClass.SetUpActivateClass(CanActivateCondition, ActivateCoroutine, -1, false, EffectDiscription());
-                cardEffects.Add(activateClass);
+        #region When Digivolving
+        if (timing == EffectTiming.WhenDigivolving)
+        {
+            ActivateClass activateClass = new ActivateClass();
+            activateClass.SetUpICardEffect("Give effects to opponent's Digimon and play a Token", CanUseCondition, card);
+            activateClass.SetUpActivateClass(CanActivateCondition, ActivateCoroutine, -1, false, EffectDiscription());
+            cardEffects.Add(activateClass);
 
-                string EffectDiscription()
-                {
-                    return "[When Digivolving] 1 of your opponent's Digimon gains \" Delete this Digimon.\" until the end of their turn. Then, if this Digimon has [LadyDevimon]/[X Antibody] in its digivolution cards, you may play 1 [Volée & Zerdrücken] Token (Digimon/Lv.4/Purple/5000 DP/<Blocker>/<Retaliation>).";
-                }
+            string EffectDiscription()
+            {
+                return "[When Digivolving] 1 of your opponent's Digimon gains \" Delete this Digimon.\" until the end of their turn. Then, if this Digimon has [LadyDevimon]/[X Antibody] in its digivolution cards, you may play 1 [Volée & Zerdrücken] Token (Digimon/Lv.4/Purple/5000 DP/<Blocker>/<Retaliation>).";
+            }
 
-                bool CanUseCondition(Hashtable hashtable)
-                {
-                    if(CardEffectCommons.IsExistOnBattleAreaDigimon(card))
-                        return CardEffectCommons.CanTriggerWhenDigivolving(hashtable, card);
+            bool CanUseCondition(Hashtable hashtable)
+            {
+                if (CardEffectCommons.IsExistOnBattleAreaDigimon(card))
+                    return CardEffectCommons.CanTriggerWhenDigivolving(hashtable, card);
 
-                    return false;
-                }
+                return false;
+            }
 
-                bool CanActivateCondition(Hashtable hashtable)
-                {
-                    return CardEffectCommons.IsExistOnBattleAreaDigimon(card);
-                }
+            bool CanActivateCondition(Hashtable hashtable)
+            {
+                return CardEffectCommons.IsExistOnBattleAreaDigimon(card);
+            }
 
-                IEnumerator ActivateCoroutine(Hashtable hashtable)
+            async Task ActivateCoroutine(Hashtable hashtable)
+            {
+                if (CardEffectCommons.HasMatchConditionPermanent(card, CanSelectPermanentCondition))
                 {
-                    if (CardEffectCommons.HasMatchConditionPermanent(CanSelectPermanentCondition))
-                    {
-                        int maxCount = Math.Min(1, CardEffectCommons.MatchConditionPermanentCount(CanSelectPermanentCondition));
+                    int maxCount = Math.Min(1, CardEffectCommons.MatchConditionPermanentCount(card, CanSelectPermanentCondition));
 
-                        SelectPermanentEffect selectPermanentEffect = GManager.instance.GetComponent<SelectPermanentEffect>();
+                    SelectPermanentEffect selectPermanentEffect = GManager.instance.GetComponent<SelectPermanentEffect>();
 
-                        selectPermanentEffect.SetUp(
-                            selectPlayer: card.Owner,
-                            canTargetCondition: CanSelectPermanentCondition,
-                            canTargetCondition_ByPreSelecetedList: null,
-                            canEndSelectCondition: null,
-                            maxCount: maxCount,
-                            canNoSelect: false,
-                            canEndNotMax: false,
-                            selectPermanentCoroutine: SelectPermanentCoroutine,
-                            afterSelectPermanentCoroutine: null,
-                            mode: SelectPermanentEffect.Mode.Custom,
-                            cardEffect: activateClass);
-
-                        selectPermanentEffect.SetUpCustomMessage(
-                            "Select 1 Digimon that will get [End of Attack] Delete this Digimon.",
-                            "The opponent is selecting 1 Digimon that will get [End of Attack] Delete this Digimon.");
+                    selectPermanentEffect.SetUp(
+                        selectPlayer: card.Owner,
+                        canTargetCondition: CanSelectPermanentCondition,
+                        canTargetCondition_ByPreSelecetedList: null,
+                        canEndSelectCondition: null,
+                        maxCount: maxCount,
+                        canNoSelect: false,
+                        canEndNotMax: false,
+                        selectPermanentCoroutine: SelectPermanentCoroutine,
+                        afterSelectPermanentCoroutine: null,
+                        mode: SelectPermanentEffect.Mode.Custom,
+                        cardEffect: activateClass);
+
+                    selectPermanentEffect.SetUpCustomMessage(
+                        "Select 1 Digimon that will get [End of Attack] Delete this Digimon.",
+                        "The opponent is selecting 1 Digimon that will get [End of Attack] Delete this Digimon.");
 
-                        yield return ContinuousController.instance.StartCoroutine(selectPermanentEffect.Activate());
+                    await selectPermanentEffect.Activate();
 
-                        IEnumerator SelectPermanentCoroutine(Permanent permanent)
+                    Task SelectPermanentCoroutine(Permanent permanent)
+                    {
+                        if (permanent != null)
                         {
-                            if (permanent != null)
-                            {
-                                yield return ContinuousController.instance.StartCoroutine(GManager.instance.GetComponent<Effects>().CreateDebuffEffect(permanent));
-                                ActivateClass activateClass1 = new ActivateClass();
-                                activateClass1.SetUpICardEffect("Delete this Digimon", CanUseCondition1, permanent.TopCard);
-                                activateClass1.SetUpActivateClass(CanActivateCondition1, ActivateCoroutine1, -1, false, EffectDiscription1());
-                                activateClass1.SetEffectSourcePermanent(permanent);
-                                permanent.UntilOwnerTurnEndEffects.Add(GetCardEffect);
-
-                                if (!permanent.TopCard.CanNotBeAffected(activateClass))
-                                {
-                                    yield return ContinuousController.instance.StartCoroutine(GManager.instance.GetComponent<Effects>().CreateDebuffEffect(permanent));
-                                }
+                            ActivateClass activateClass1 = new ActivateClass();
+                            activateClass1.SetUpICardEffect("Delete this Digimon", CanUseCondition1, permanent.TopCard);
+                            activateClass1.SetUpActivateClass(CanActivateCondition1, ActivateCoroutine1, -1, false, EffectDiscription1());
+                            activateClass1.SetEffectSourcePermanent(permanent);
+                            permanent.UntilOwnerTurnEndEffects.Add(GetCardEffect);
 
-                                string EffectDiscription1()
-                                {
-                                    return "[End of Attack] Delete this Digimon.";
-                                }
+                            string EffectDiscription1()
+                            {
+                                return "[End of Attack] Delete this Digimon.";
+                            }
 
-                                bool CanUseCondition1(Hashtable hashtable1)
+                            bool CanUseCondition1(Hashtable hashtable1)
+                            {
+                                if (CardEffectCommons.IsOpponentTurn(card))
                                 {
-                                    if (CardEffectCommons.IsOpponentTurn(card))
+                                    if (CardEffectCommons.IsPermanentExistsOnOwnerBattleArea(permanent, permanent.TopCard))
                                     {
-                                        if (CardEffectCommons.IsPermanentExistsOnOwnerBattleArea(permanent, permanent.TopCard))
+                                        if (CardEffectCommons.CanTriggerOnEndAttack(hashtable1, permanent.TopCard))
                                         {
-                                            if (CardEffectCommons.CanTriggerOnEndAttack(hashtable1, permanent.TopCard))
-                                            {
-                                                return true;
-                                            }
+                                            return true;
                                         }
                                     }
-
-                                    return false;
                                 }
 
-                                bool CanActivateCondition1(Hashtable hashtable1)
+                                return false;
+                            }
+
+                            bool CanActivateCondition1(Hashtable hashtable1)
+                            {
+                                if (CardEffectCommons.IsPermanentExistsOnBattleArea(permanent))
                                 {
-                                    if (CardEffectCommons.IsPermanentExistsOnBattleArea(permanent))
+                                    if (!permanent.TopCard.CanNotBeAffected(activateClass))
                                     {
-                                        if (!permanent.TopCard.CanNotBeAffected(activateClass))
-                                        {
-                                            return true;
-                                        }
+                                        return true;
                                     }
-
-                                    return false;
                                 }
 
-                                IEnumerator ActivateCoroutine1(Hashtable _hashtable1)
+                                return false;
+                            }
+
+                            async Task ActivateCoroutine1(Hashtable _hashtable1)
+                            {
+                                if (CardEffectCommons.IsPermanentExistsOnBattleArea(permanent))
                                 {
-                                    if (CardEffectCommons.IsPermanentExistsOnBattleArea(permanent))
-                                    {
-                                        yield return ContinuousController.instance.StartCoroutine(new DestroyPermanentsClass(
-                                        new List<Permanent>() { permanent },
-                                        CardEffectCommons.CardEffectHashtable(activateClass1)).Destroy());
-                                    }
+                                    await new DestroyPermanentsClass(
+                                    new List<Permanent>() { permanent },
+                                    CardEffectCommons.CardEffectHashtable(activateClass1)).Destroy();
                                 }
+                            }
 
-                                ICardEffect GetCardEffect(EffectTiming _timing)
+                            ICardEffect GetCardEffect(EffectTiming _timing)
+                            {
+                                if (_timing == EffectTiming.OnEndAttack)
                                 {
-                                    if (_timing == EffectTiming.OnEndAttack)
-                                    {
-                                        return activateClass1;
-                                    }
-
-                                    return null;
+                                    return activateClass1;
                                 }
 
+                                return null;
                             }
                         }
+
+                        return Task.CompletedTask;
                     }
+                }
 
-                    if (CanActivateSecondEffectCondition(hashtable))
+                if (CanActivateSecondEffectCondition(hashtable))
+                {
+                    List<SelectionElement<bool>> selectionElements = new List<SelectionElement<bool>>()
                     {
-                        List<SelectionElement<bool>> selectionElements = new List<SelectionElement<bool>>()
-                        {
-                            new SelectionElement<bool>(message: $"Play Token", value : true, spriteIndex: 0),
-                            new SelectionElement<bool>(message: $"Not Play Token", value : false, spriteIndex: 1),
-                        };
+                        new SelectionElement<bool>(message: $"Play Token", value : true, spriteIndex: 0),
+                        new SelectionElement<bool>(message: $"Not Play Token", value : false, spriteIndex: 1),
+                    };
 
-                        string selectPlayerMessage = "Will you play a token?";
-                        string notSelectPlayerMessage = "The opponent is choosing to play a token.";
+                    string selectPlayerMessage = "Will you play a token?";
+                    string notSelectPlayerMessage = "The opponent is choosing to play a token.";
 
-                        GManager.instance.userSelectionManager.SetBoolSelection(selectionElements: selectionElements, selectPlayer: card.Owner, selectPlayerMessage: selectPlayerMessage, notSelectPlayerMessage: notSelectPlayerMessage);
+                    GManager.instance.userSelectionManager.SetBoolSelection(selectionElements: selectionElements, selectPlayer: card.Owner, selectPlayerMessage: selectPlayerMessage, notSelectPlayerMessage: notSelectPlayerMessage);
 
-                        yield return ContinuousController.instance.StartCoroutine(GManager.instance.userSelectionManager.WaitForEndSelect());
+                    await GManager.instance.userSelectionManager.WaitForEndSelect();
 
-                        bool canPlay = GManager.instance.userSelectionManager.SelectedBoolValue;
+                    bool canPlay = GManager.instance.userSelectionManager.SelectedBoolValue;
 
-                        if(canPlay)
-                            yield return ContinuousController.instance.StartCoroutine(CardEffectCommons.PlayVoleeZerdrucken(activateClass));
-                    }
+                    if (canPlay)
+                        await CardEffectCommons.PlayVoleeZerdrucken(activateClass);
                 }
             }
-            #endregion
+        }
+        #endregion
 
-            #region Inherit
-            if (timing == EffectTiming.OnDestroyedAnyone)
-            {
-                ActivateClass activateClass = new ActivateClass();
-                activateClass.SetUpICardEffect("Play 1 level 4 or lower purple Digimon from your trash", CanUseCondition, card);
-                activateClass.SetUpActivateClass(CanActivateCondition, ActivateCoroutine, 1, true, EffectDiscription());
-                activateClass.SetHashString("PlayLevel4_EX7_058");
-                activateClass.SetIsInheritedEffect(true);
-                cardEffects.Add(activateClass);
+        #region Inherit
+        if (timing == EffectTiming.OnDestroyedAnyone)
+        {
+            ActivateClass activateClass = new ActivateClass();
+            activateClass.SetUpICardEffect("Play 1 level 4 or lower purple Digimon from your trash", CanUseCondition, card);
+            activateClass.SetUpActivateClass(CanActivateCondition, ActivateCoroutine, 1, true, EffectDiscription());
+            activateClass.SetHashString("PlayLevel4_EX7_058");
+            activateClass.SetIsInheritedEffect(true);
+            cardEffects.Add(activateClass);
 
-                string EffectDiscription()
-                {
-                    return "[Opponent's Turn] [Once per turn]  When an opponent's Digimon is deleted, you may play 1 purple level 4 or lower Digimon card from your trash without paying the cost.";
-                }
+            string EffectDiscription()
+            {
+                return "[Opponent's Turn] [Once per turn]  When an opponent's Digimon is deleted, you may play 1 purple level 4 or lower Digimon card from your trash without paying the cost.";
+            }
 
-                bool CanSelectCardInTrash(CardSource cardSource)
+            bool CanSelectCardInTrash(CardSource cardSource)
+            {
+                if (CardEffectCommons.CanPlayAsNewPermanent(cardSource: cardSource, payCost: false, cardEffect: activateClass))
                 {
-                    if (CardEffectCommons.CanPlayAsNewPermanent(cardSource: cardSource, payCost: false, cardEffect: activateClass))
+                    if (cardSource.HasCardColor("Purple"))
                     {
-                        if (cardSource.HasCardColor(CardColor.Purple))
+                        if (cardSource.IsDigimon)
                         {
-                            if (cardSource.IsDigimon)
+                            if (cardSource.Level <= 4)
                             {
-                                if (cardSource.Level <= 4)
+                                if (cardSource.HasLevel)
                                 {
-                                    if (cardSource.HasLevel)
-                                    {
-                                        return true;
-                                    }
+                                    return true;
                                 }
                             }
                         }
                     }
-
-                    return false;
                 }
 
-                bool CanUseCondition(Hashtable hashtable)
+                return false;
+            }
+
+            bool CanUseCondition(Hashtable hashtable)
+            {
+                if (CardEffectCommons.IsExistOnBattleAreaDigimon(card))
                 {
-                    if (CardEffectCommons.IsExistOnBattleAreaDigimon(card))
+                    if (CardEffectCommons.IsOpponentTurn(card))
                     {
-                        if (CardEffectCommons.IsOpponentTurn(card))
+                        if (CardEffectCommons.CanTriggerOnPermanentDeleted(hashtable, PermanentCondition))
                         {
-                            if (CardEffectCommons.CanTriggerOnPermanentDeleted(hashtable, PermanentCondition))
-                            {
-                                return true;
-                            }
+                            return true;
                         }
                     }
-
-                    return false;
                 }
 
-                bool PermanentCondition(Permanent permanent)
+                return false;
+            }
+
+            bool PermanentCondition(Permanent permanent)
+            {
+                if (CardEffectCommons.IsPermanentExistsOnOpponentBattleAreaDigimon(permanent, card))
                 {
-                    if (CardEffectCommons.IsPermanentExistsOnOpponentBattleAreaDigimon(permanent, card))
+                    if (permanent.IsDigimon)
                     {
-                        if (permanent.IsDigimon)
-                        {
-                            return true;
-                        }
+                        return true;
                     }
-
-                    return false;
                 }
 
-                bool CanActivateCondition(Hashtable hashtable)
+                return false;
+            }
+
+            bool CanActivateCondition(Hashtable hashtable)
+            {
+                if (CardEffectCommons.IsExistOnBattleAreaDigimon(card))
                 {
-                    if (CardEffectCommons.IsExistOnBattleAreaDigimon(card))
+                    if (CardEffectCommons.HasMatchConditionOwnersCardInTrash(card, (cardSource) => CanSelectCardInTrash(cardSource)))
                     {
-                        if (CardEffectCommons.HasMatchConditionOwnersCardInTrash(card, (cardSource) => CanSelectCardInTrash(cardSource)))
-                        {
-                            return true;
-                        }
+                        return true;
                     }
-
-                    return false;
                 }
 
-                IEnumerator ActivateCoroutine(Hashtable hashtable)
-                {
-                    if (CardEffectCommons.HasMatchConditionOwnersCardInTrash(card, CanSelectCardInTrash))
-                    {
-                        int maxCount = Math.Min(1, card.Owner.TrashCards.Count(CanSelectCardInTrash));
+                return false;
+            }
 
-                        List<CardSource> selectedCards = new List<CardSource>();
+            async Task ActivateCoroutine(Hashtable hashtable)
+            {
+                if (CardEffectCommons.HasMatchConditionOwnersCardInTrash(card, CanSelectCardInTrash))
+                {
+                    int maxCount = Math.Min(1, new Player(card.Context, card.Owner).TrashCards.Count(CanSelectCardInTrash));
 
-                        SelectCardEffect selectCardEffect = GManager.instance.GetComponent<SelectCardEffect>();
+                    List<CardSource> selectedCards = new List<CardSource>();
 
-                        selectCardEffect.SetUp(
-                                    canTargetCondition: CanSelectCardInTrash,
-                                    canTargetCondition_ByPreSelecetedList: null,
-                                    canEndSelectCondition: null,
-                                    canNoSelect: () => true,
-                                    selectCardCoroutine: SelectCardCoroutine,
-                                    afterSelectCardCoroutine: null,
-                                    message: "Select 1 card to play.",
-                                    maxCount: maxCount,
-                                    canEndNotMax: false,
-                                    isShowOpponent: true,
-                                    mode: SelectCardEffect.Mode.Custom,
-                                    root: SelectCardEffect.Root.Trash,
-                                    customRootCardList: null,
-                                    canLookReverseCard: true,
-                                    selectPlayer: card.Owner,
-                                    cardEffect: activateClass);
+                    SelectCardEffect selectCardEffect = GManager.instance.GetComponent<SelectCardEffect>();
 
-                        selectCardEffect.SetUpCustomMessage("Select 1 card to play.", "The opponent is selecting 1 card to play.");
-                        selectCardEffect.SetUpCustomMessage_ShowCard("Played Card");
+                    selectCardEffect.SetUp(
+                                canTargetCondition: CanSelectCardInTrash,
+                                canTargetCondition_ByPreSelecetedList: null,
+                                canEndSelectCondition: null,
+                                canNoSelect: () => true,
+                                selectCardCoroutine: SelectCardCoroutine,
+                                afterSelectCardCoroutine: null,
+                                message: "Select 1 card to play.",
+                                maxCount: maxCount,
+                                canEndNotMax: false,
+                                isShowOpponent: true,
+                                mode: SelectCardEffect.Mode.Custom,
+                                root: SelectCardEffect.Root.Trash,
+                                customRootCardList: null,
+                                canLookReverseCard: true,
+                                selectPlayer: card.Owner,
+                                cardEffect: activateClass);
 
-                        yield return StartCoroutine(selectCardEffect.Activate());
+                    selectCardEffect.SetUpCustomMessage("Select 1 card to play.", "The opponent is selecting 1 card to play.");
+                    selectCardEffect.SetUpCustomMessage_ShowCard("Played Card");
 
-                        IEnumerator SelectCardCoroutine(CardSource cardSource)
-                        {
-                            selectedCards.Add(cardSource);
+                    await selectCardEffect.Activate();
 
-                            yield return null;
-                        }
+                    Task SelectCardCoroutine(CardSource cardSource)
+                    {
+                        selectedCards.Add(cardSource);
 
-                        yield return ContinuousController.instance.StartCoroutine(CardEffectCommons.PlayPermanentCards(
-                            cardSources: selectedCards,
-                            activateClass: activateClass,
-                            payCost: false,
-                            isTapped: false,
-                            root: SelectCardEffect.Root.Trash,
-                            activateETB: true));
+                        return Task.CompletedTask;
                     }
+
+                    await CardEffectCommons.PlayPermanentCards(
+                        cardSources: selectedCards,
+                        activateClass: activateClass,
+                        payCost: false,
+                        isTapped: false,
+                        root: SelectCardEffect.Root.Trash,
+                        activateETB: true);
                 }
             }
-            #endregion
-
-            return cardEffects;
         }
+        #endregion
+
+        return cardEffects;
     }
-}
\ 파일 끝 개행 문자 없음
+}
