--- DCGO/Assets/Scripts/CardEffect/BT21/Purple/BT21_077.cs	2026-05-18 20:03:14.911461600 +0900
+++ src/HeadlessDCGO.Engine/Assets/Scripts/CardEffect/BT21/Purple/BT21_077.cs	2026-07-29 01:02:22.516884001 +0900
@@ -1,419 +1,421 @@
+namespace HeadlessDCGO.Engine.Assets.Scripts.CardEffect.BT21.Purple;
+
 using System.Collections;
-using System.Collections.Generic;
 using System.Linq;
+using System.Threading.Tasks;
+using HeadlessDCGO.Engine.Assets.Scripts.Script;
+using HeadlessDCGO.Engine.Assets.Scripts.Script.CardEffectCommons;
+using HeadlessDCGO.Engine.Assets.Scripts.Script.CardEffects;
+using HeadlessDCGO.Engine.Headless.Services;
 
-//BT21_077 Regulusmon
-namespace DCGO.CardEffects.BT21
+public class BT21_077 : CEntity_Effect
 {
-    public class BT21_077 : CEntity_Effect
+    public override List<ICardEffect> CardEffects(EffectTiming timing, CardSource card)
     {
-        public override List<ICardEffect> CardEffects(EffectTiming timing, CardSource card)
-        {
-            List<ICardEffect> cardEffects = new List<ICardEffect>();
+        List<ICardEffect> cardEffects = new List<ICardEffect>();
 
-            #region Digivolution condition
+        #region Digivolution condition
 
-            if (timing == EffectTiming.None)
+        if (timing == EffectTiming.None)
+        {
+            bool Condition(Permanent permanent)
             {
-                bool Condition(Permanent permanent)
-                {
-                    return permanent.Level == 4 && permanent.TopCard.HasText("Gammamon");
-                }
-
-                cardEffects.Add(CardEffectFactory.AddSelfDigivolutionRequirementStaticEffect(Condition, 3, false, card, null));
+                return permanent.Level == 4 && permanent.TopCard.HasText("Gammamon");
             }
 
-            #endregion
+            cardEffects.Add(CardEffectFactory.AddSelfDigivolutionRequirementStaticEffect(Condition, 3, false, card, null));
+        }
 
-            #region On Play/When Digivolving shared
+        #endregion
 
-            bool CanTargetPermanent(Permanent permanent)
-            {
-                return permanent.IsDigimon && CardEffectCommons.IsPermanentExistsOnOpponentBattleAreaDigimon(permanent, card);
-            }
+        #region On Play/When Digivolving shared
+
+        bool CanTargetPermanent(Permanent permanent)
+        {
+            return permanent.IsDigimon && CardEffectCommons.IsPermanentExistsOnOpponentBattleAreaDigimon(permanent, card);
+        }
+
+        bool CanActivateConditionShared(Hashtable hashtable)
+        {
+            return CardEffectCommons.IsExistOnBattleAreaDigimon(card);
+        }
+
+        bool CanTrashCard(CardSource cardSource)
+        {
+            return cardSource.HasText("Gammamon");
+        }
+
+        #endregion
 
-            bool CanActivateConditionShared(Hashtable hashtable)
+        #region On Play
+
+        if (timing == EffectTiming.OnEnterFieldAnyone)
+        {
+            ActivateClass activateClass = new ActivateClass();
+            activateClass.SetUpICardEffect("Force attack and give collision", CanUseCondition, card);
+            activateClass.SetUpActivateClass(CanActivateConditionShared, ActivateCoroutine, -1, false, EffectDescription());
+            cardEffects.Add(activateClass);
+
+            string EffectDescription()
             {
-                return CardEffectCommons.IsExistOnBattleAreaDigimon(card);
+                return "[On Play] By trashing 1 card with [Gammamon] in its text from your hand, give 1 of your opponent's Digimon <Collision> and \"[Start of Your Main Phase] This Digimon attacks.\" until their turn ends.";
             }
 
-            bool CanTrashCard(CardSource card)
+            bool CanUseCondition(Hashtable hashtable)
             {
-                return card.HasText("Gammamon");
+                return CardEffectCommons.CanTriggerOnPlay(hashtable, card);
             }
 
-            #endregion
+            async Task ActivateCoroutine(Hashtable hashtable)
+            {
+                if (new Player(card.Context, card.Owner).HandCards.Count(CanTrashCard) >= 1)
+                {
+                    bool discarded = false;
 
-            #region On Play
+                    int discardCount = 1;
 
-            if (timing == EffectTiming.OnEnterFieldAnyone)
-            {
-                ActivateClass activateClass = new ActivateClass();
-                activateClass.SetUpICardEffect("Force attack and give collision", CanUseCondition, card);
-                activateClass.SetUpActivateClass(CanActivateConditionShared, ActivateCoroutine, -1, false, EffectDescription());
-                cardEffects.Add(activateClass);
+                    SelectHandEffect selectHandEffect = GManager.instance.GetComponent<SelectHandEffect>();
 
-                string EffectDescription()
-                {
-                    return "[On Play] By trashing 1 card with [Gammamon] in its text from your hand, give 1 of your opponent's Digimon <Collision> and \"[Start of Your Main Phase] This Digimon attacks.\" until their turn ends.";
-                }
+                    selectHandEffect.SetUp(
+                        selectPlayer: card.Owner,
+                        canTargetCondition: CanTrashCard,
+                        canTargetCondition_ByPreSelecetedList: null,
+                        canEndSelectCondition: null,
+                        maxCount: discardCount,
+                        canNoSelect: true,
+                        canEndNotMax: false,
+                        isShowOpponent: true,
+                        selectCardCoroutine: null,
+                        afterSelectCardCoroutine: AfterSelectCardCoroutine,
+                        mode: SelectHandEffect.Mode.Discard,
+                        cardEffect: activateClass);
 
-                bool CanUseCondition(Hashtable hashtable)
-                {
-                    return CardEffectCommons.CanTriggerOnPlay(hashtable, card);
-                }
+                    await selectHandEffect.Activate();
 
-                IEnumerator ActivateCoroutine(Hashtable hashtable)
-                {
-                    if (card.Owner.HandCards.Count(CanTrashCard) >= 1)
+                    async Task AfterSelectCardCoroutine(List<CardSource> cardSources)
                     {
-                        bool discarded = false;
+                        if (cardSources.Count >= 1)
+                        {
+                            discarded = true;
 
-                        int discardCount = 1;
+                            await Task.CompletedTask;
+                        }
+                    }
 
-                        SelectHandEffect selectHandEffect = GManager.instance.GetComponent<SelectHandEffect>();
+                    if (discarded)
+                    {
+                        if (CardEffectCommons.HasMatchConditionPermanent(card, CanTargetPermanent))
+                        {
+                            Permanent? selectedPermanent = null;
 
-                        selectHandEffect.SetUp(
-                            selectPlayer: card.Owner,
-                            canTargetCondition: CanTrashCard,
-                            canTargetCondition_ByPreSelecetedList: null,
-                            canEndSelectCondition: null,
-                            maxCount: discardCount,
-                            canNoSelect: true,
-                            canEndNotMax: false,
-                            isShowOpponent: true,
-                            selectCardCoroutine: null,
-                            afterSelectCardCoroutine: AfterSelectCardCoroutine,
-                            mode: SelectHandEffect.Mode.Discard,
-                            cardEffect: activateClass);
+                            SelectPermanentEffect selectPermanentEffect = GManager.instance.GetComponent<SelectPermanentEffect>();
 
-                        yield return StartCoroutine(selectHandEffect.Activate());
+                            selectPermanentEffect.SetUp(
+                                selectPlayer: card.Owner,
+                                canTargetCondition: CanTargetPermanent,
+                                canTargetCondition_ByPreSelecetedList: null,
+                                canEndSelectCondition: null,
+                                maxCount: 1,
+                                canNoSelect: false,
+                                canEndNotMax: false,
+                                selectPermanentCoroutine: SelectPermanentCoroutine,
+                                afterSelectPermanentCoroutine: null,
+                                mode: SelectPermanentEffect.Mode.Custom,
+                                cardEffect: activateClass);
+
+                            selectPermanentEffect.SetUpCustomMessage("Select 1 Digimon that will get effects.",
+                                "The opponent is selecting 1 Digimon that will get effects.");
+                            await selectPermanentEffect.Activate();
 
-                        IEnumerator AfterSelectCardCoroutine(List<CardSource> cardSources)
-                        {
-                            if (cardSources.Count >= 1)
+                            async Task SelectPermanentCoroutine(Permanent permanent)
                             {
-                                discarded = true;
-
-                                yield return null;
+                                selectedPermanent = permanent;
+                                await Task.CompletedTask;
                             }
-                        }
 
-                        if (discarded)
-                        {
-                            if (CardEffectCommons.HasMatchConditionPermanent(CanTargetPermanent))
+                            if (selectedPermanent != null)
                             {
-                                Permanent selectedPermanent = null;
+                                await CardEffectCommons.GainCollision(
+                                    targetPermanent: selectedPermanent,
+                                    effectDuration: EffectDuration.UntilOwnerTurnEnd,
+                                    activateClass: activateClass);
 
-                                SelectPermanentEffect selectPermanentEffect = GManager.instance.GetComponent<SelectPermanentEffect>();
-
-                                selectPermanentEffect.SetUp(
-                                    selectPlayer: card.Owner,
-                                    canTargetCondition: CanTargetPermanent,
-                                    canTargetCondition_ByPreSelecetedList: null,
-                                    canEndSelectCondition: null,
-                                    maxCount: 1,
-                                    canNoSelect: false,
-                                    canEndNotMax: false,
-                                    selectPermanentCoroutine: SelectPermanentCoroutine,
-                                    afterSelectPermanentCoroutine: null,
-                                    mode: SelectPermanentEffect.Mode.Custom,
+                                await CardEffectCommons.StartOfMainAttack(
+                                    targetPermanent: selectedPermanent,
                                     cardEffect: activateClass);
-
-                                selectPermanentEffect.SetUpCustomMessage("Select 1 Digimon that will get effects.",
-                                    "The opponent is selecting 1 Digimon that will get effects.");
-                                yield return ContinuousController.instance.StartCoroutine(selectPermanentEffect.Activate());
-
-                                IEnumerator SelectPermanentCoroutine(Permanent permanent)
-                                {
-                                    selectedPermanent = permanent;
-                                    yield return null;
-                                }
-
-                                if (selectedPermanent != null)
-                                {
-                                    yield return ContinuousController.instance.StartCoroutine(CardEffectCommons.GainCollision(
-                                        targetPermanent: selectedPermanent,
-                                        effectDuration: EffectDuration.UntilOwnerTurnEnd,
-                                        activateClass: activateClass));
-
-                                    yield return ContinuousController.instance.StartCoroutine(CardEffectCommons.StartOfMainAttack(
-                                        targetPermanent: selectedPermanent,
-                                        cardEffect: activateClass));
-                                }
                             }
                         }
                     }
                 }
             }
+        }
 
-            #endregion
+        #endregion
 
-            #region When Digivolving
+        #region When Digivolving
 
-            if (timing == EffectTiming.OnEnterFieldAnyone)
-            {
-                ActivateClass activateClass = new ActivateClass();
-                activateClass.SetUpICardEffect("Force attack and give collision", CanUseCondition, card);
-                activateClass.SetUpActivateClass(CanActivateConditionShared, ActivateCoroutine, -1, false, EffectDescription());
-                cardEffects.Add(activateClass);
+        if (timing == EffectTiming.WhenDigivolving)
+        {
+            ActivateClass activateClass = new ActivateClass();
+            activateClass.SetUpICardEffect("Force attack and give collision", CanUseCondition, card);
+            activateClass.SetUpActivateClass(CanActivateConditionShared, ActivateCoroutine, -1, false, EffectDescription());
+            cardEffects.Add(activateClass);
 
-                string EffectDescription()
-                {
-                    return "[When Digivolving] By trashing 1 card with [Gammamon] in its text from your hand, give 1 of your opponent's Digimon <Collision> and \"[Start of Your Main Phase] This Digimon attacks.\" until their turn ends.";
-                }
+            string EffectDescription()
+            {
+                return "[When Digivolving] By trashing 1 card with [Gammamon] in its text from your hand, give 1 of your opponent's Digimon <Collision> and \"[Start of Your Main Phase] This Digimon attacks.\" until their turn ends.";
+            }
 
-                bool CanUseCondition(Hashtable hashtable)
-                {
-                    return CardEffectCommons.CanTriggerWhenDigivolving(hashtable, card);
-                }
+            bool CanUseCondition(Hashtable hashtable)
+            {
+                return CardEffectCommons.CanTriggerWhenDigivolving(hashtable, card);
+            }
 
-                IEnumerator ActivateCoroutine(Hashtable hashtable)
+            async Task ActivateCoroutine(Hashtable hashtable)
+            {
+                if (new Player(card.Context, card.Owner).HandCards.Count(CanTrashCard) >= 1)
                 {
-                    if (card.Owner.HandCards.Count(CanTrashCard) >= 1)
-                    {
-                        bool discarded = false;
+                    bool discarded = false;
 
-                        int discardCount = 1;
+                    int discardCount = 1;
 
-                        SelectHandEffect selectHandEffect = GManager.instance.GetComponent<SelectHandEffect>();
+                    SelectHandEffect selectHandEffect = GManager.instance.GetComponent<SelectHandEffect>();
 
-                        selectHandEffect.SetUp(
-                            selectPlayer: card.Owner,
-                            canTargetCondition: CanTrashCard,
-                            canTargetCondition_ByPreSelecetedList: null,
-                            canEndSelectCondition: null,
-                            maxCount: discardCount,
-                            canNoSelect: true,
-                            canEndNotMax: false,
-                            isShowOpponent: true,
-                            selectCardCoroutine: null,
-                            afterSelectCardCoroutine: AfterSelectCardCoroutine,
-                            mode: SelectHandEffect.Mode.Discard,
-                            cardEffect: activateClass);
+                    selectHandEffect.SetUp(
+                        selectPlayer: card.Owner,
+                        canTargetCondition: CanTrashCard,
+                        canTargetCondition_ByPreSelecetedList: null,
+                        canEndSelectCondition: null,
+                        maxCount: discardCount,
+                        canNoSelect: true,
+                        canEndNotMax: false,
+                        isShowOpponent: true,
+                        selectCardCoroutine: null,
+                        afterSelectCardCoroutine: AfterSelectCardCoroutine,
+                        mode: SelectHandEffect.Mode.Discard,
+                        cardEffect: activateClass);
 
-                        yield return StartCoroutine(selectHandEffect.Activate());
+                    await selectHandEffect.Activate();
 
-                        IEnumerator AfterSelectCardCoroutine(List<CardSource> cardSources)
+                    async Task AfterSelectCardCoroutine(List<CardSource> cardSources)
+                    {
+                        if (cardSources.Count >= 1)
                         {
-                            if (cardSources.Count >= 1)
-                            {
-                                discarded = true;
+                            discarded = true;
 
-                                yield return null;
-                            }
+                            await Task.CompletedTask;
                         }
+                    }
 
-                        if (discarded)
+                    if (discarded)
+                    {
+                        if (CardEffectCommons.HasMatchConditionPermanent(card, CanTargetPermanent))
                         {
-                            if (CardEffectCommons.HasMatchConditionPermanent(CanTargetPermanent))
+                            Permanent? selectedPermanent = null;
+
+                            SelectPermanentEffect selectPermanentEffect = GManager.instance.GetComponent<SelectPermanentEffect>();
+
+                            selectPermanentEffect.SetUp(
+                                selectPlayer: card.Owner,
+                                canTargetCondition: CanTargetPermanent,
+                                canTargetCondition_ByPreSelecetedList: null,
+                                canEndSelectCondition: null,
+                                maxCount: 1,
+                                canNoSelect: false,
+                                canEndNotMax: false,
+                                selectPermanentCoroutine: SelectPermanentCoroutine,
+                                afterSelectPermanentCoroutine: null,
+                                mode: SelectPermanentEffect.Mode.Custom,
+                                cardEffect: activateClass);
+
+                            selectPermanentEffect.SetUpCustomMessage("Select 1 Digimon that will get effects.",
+                                "The opponent is selecting 1 Digimon that will get effects.");
+                            await selectPermanentEffect.Activate();
+
+                            async Task SelectPermanentCoroutine(Permanent permanent)
                             {
-                                Permanent selectedPermanent = null;
+                                selectedPermanent = permanent;
+                                await Task.CompletedTask;
+                            }
 
-                                SelectPermanentEffect selectPermanentEffect = GManager.instance.GetComponent<SelectPermanentEffect>();
+                            if (selectedPermanent != null)
+                            {
+                                await CardEffectCommons.GainCollision(
+                                    targetPermanent: selectedPermanent,
+                                    effectDuration: EffectDuration.UntilOwnerTurnEnd,
+                                    activateClass: activateClass);
 
-                                selectPermanentEffect.SetUp(
-                                    selectPlayer: card.Owner,
-                                    canTargetCondition: CanTargetPermanent,
-                                    canTargetCondition_ByPreSelecetedList: null,
-                                    canEndSelectCondition: null,
-                                    maxCount: 1,
-                                    canNoSelect: false,
-                                    canEndNotMax: false,
-                                    selectPermanentCoroutine: SelectPermanentCoroutine,
-                                    afterSelectPermanentCoroutine: null,
-                                    mode: SelectPermanentEffect.Mode.Custom,
+                                await CardEffectCommons.StartOfMainAttack(
+                                    targetPermanent: selectedPermanent,
                                     cardEffect: activateClass);
-
-                                selectPermanentEffect.SetUpCustomMessage("Select 1 Digimon that will get effects.",
-                                    "The opponent is selecting 1 Digimon that will get effects.");
-                                yield return ContinuousController.instance.StartCoroutine(selectPermanentEffect.Activate());
-
-                                IEnumerator SelectPermanentCoroutine(Permanent permanent)
-                                {
-                                    selectedPermanent = permanent;
-                                    yield return null;
-                                }
-
-                                if (selectedPermanent != null)
-                                {
-                                    yield return ContinuousController.instance.StartCoroutine(CardEffectCommons.GainCollision(
-                                        targetPermanent: selectedPermanent,
-                                        effectDuration: EffectDuration.UntilOwnerTurnEnd,
-                                        activateClass: activateClass));
-
-                                    yield return ContinuousController.instance.StartCoroutine(CardEffectCommons.StartOfMainAttack(
-                                        targetPermanent: selectedPermanent,
-                                        cardEffect: activateClass));
-                                }
                             }
                         }
                     }
                 }
             }
+        }
 
-            #endregion
-
-            #region On deletion regular
-
-            if (timing == EffectTiming.OnDestroyedAnyone)
-            {
-                ActivateClass activateClass = new ActivateClass();
-                activateClass.SetUpICardEffect("Play Cannonweissmon or level 4 or lower gammamon in text digimon", CanUseCondition, card);
-                activateClass.SetUpActivateClass(CanActivateCondition, ActivateCoroutine, -1, false, EffectDescription());
-                cardEffects.Add(activateClass);
-
-                string EffectDescription()
-                {
-                    return "[On Deletion] You may play 1 [Canoweissmon] or 1 level 4 or lower Digimon card with [Gammamon] in its text from your trash without paying the cost.";
-                }
-
-                bool CanUseCondition(Hashtable hashtable)
-                {
-                    return CardEffectCommons.CanTriggerOnDeletion(hashtable, card);
-                }
-
-                bool CanPlay(CardSource cardSource)
-                {
-                    return ((cardSource.HasText("Gammamon") && cardSource.HasLevel && cardSource.Level <= 4 && cardSource.IsDigimon) || cardSource.EqualsCardName("Canoweissmon")) &&
-                           CardEffectCommons.CanPlayAsNewPermanent(cardSource: cardSource, payCost: false, cardEffect: activateClass);
-                }
-
-                bool CanActivateCondition(Hashtable hashtable)
-                {
-                    return CardEffectCommons.CanActivateOnDeletion(hashtable, card) && CardEffectCommons.HasMatchConditionOwnersCardInTrash(card, CanPlay);
-                }
-
-                IEnumerator ActivateCoroutine(Hashtable hashtable)
-                {
-                    List<CardSource> selectedCards = new List<CardSource>();
-
-                    SelectCardEffect selectCardEffect = GManager.instance.GetComponent<SelectCardEffect>();
-
-                    selectCardEffect.SetUp(
-                        canTargetCondition: CanPlay,
-                        canTargetCondition_ByPreSelecetedList: null,
-                        canEndSelectCondition: null,
-                        canNoSelect: () => true,
-                        selectCardCoroutine: SelectCardCoroutine,
-                        afterSelectCardCoroutine: null,
-                        message: "Select 1 [Canoweissmon] or level 4 or lower [Gammamon] in text Digimon to play.",
-                        maxCount: 1,
-                        canEndNotMax: false,
-                        isShowOpponent: true,
-                        mode: SelectCardEffect.Mode.Custom,
-                        root: SelectCardEffect.Root.Trash,
-                        customRootCardList: null,
-                        canLookReverseCard: true,
-                        selectPlayer: card.Owner,
-                        cardEffect: activateClass);
-
-                    selectCardEffect.SetUpCustomMessage("Select 1 [Cannonweissmon] or level 4 or lower [Gammamon] in text Digimon to play.",
-                        "The opponent is selecting 1 [Cannonweissmon] or level 4 or lower [Gammamon] in text Digimon to play.");
-                    selectCardEffect.SetUpCustomMessage_ShowCard("Played Card");
+        #endregion
 
-                    yield return StartCoroutine(selectCardEffect.Activate());
+        #region On deletion regular
 
-                    IEnumerator SelectCardCoroutine(CardSource cardSource)
-                    {
-                        selectedCards.Add(cardSource);
+        if (timing == EffectTiming.OnDestroyedAnyone)
+        {
+            ActivateClass activateClass = new ActivateClass();
+            activateClass.SetUpICardEffect("Play Cannonweissmon or level 4 or lower gammamon in text digimon", CanUseCondition, card);
+            activateClass.SetUpActivateClass(CanActivateCondition, ActivateCoroutine, -1, false, EffectDescription());
+            cardEffects.Add(activateClass);
 
-                        yield return null;
-                    }
+            string EffectDescription()
+            {
+                return "[On Deletion] You may play 1 [Canoweissmon] or 1 level 4 or lower Digimon card with [Gammamon] in its text from your trash without paying the cost.";
+            }
 
-                    yield return ContinuousController.instance.StartCoroutine(CardEffectCommons.PlayPermanentCards(
-                        cardSources: selectedCards,
-                        activateClass: activateClass,
-                        payCost: false,
-                        isTapped: false,
-                        root: SelectCardEffect.Root.Trash,
-                        activateETB: true));
-                }
+            bool CanUseCondition(Hashtable hashtable)
+            {
+                return CardEffectCommons.CanTriggerOnDeletion(hashtable, card);
             }
 
-            #endregion
+            bool CanPlay(CardSource cardSource)
+            {
+                return ((cardSource.HasText("Gammamon") && cardSource.HasLevel && cardSource.Level <= 4 && cardSource.IsDigimon) || cardSource.EqualsCardName("Canoweissmon")) &&
+                       CardEffectCommons.CanPlayAsNewPermanent(cardSource: cardSource, payCost: false, cardEffect: activateClass);
+            }
 
-            #region On deletion inherit
+            bool CanActivateCondition(Hashtable hashtable)
+            {
+                return CardEffectCommons.CanActivateOnDeletion(hashtable, card) && CardEffectCommons.HasMatchConditionOwnersCardInTrash(card, CanPlay);
+            }
 
-            if (timing == EffectTiming.OnDestroyedAnyone)
+            async Task ActivateCoroutine(Hashtable hashtable)
             {
-                ActivateClass activateClass = new ActivateClass();
-                activateClass.SetUpICardEffect("Play Cannonweissmon or level 4 or lower gammamon in text digimon", CanUseCondition, card);
-                activateClass.SetUpActivateClass(CanActivateCondition, ActivateCoroutine, -1, false, EffectDescription());
-                activateClass.SetIsInheritedEffect(true);
-                cardEffects.Add(activateClass);
+                List<CardSource> selectedCards = new List<CardSource>();
 
-                string EffectDescription()
-                {
-                    return "[On Deletion] You may play 1 level 4 or lower Digimon card with [Gammamon] in its text from your trash without paying the cost.";
-                }
+                SelectCardEffect selectCardEffect = GManager.instance.GetComponent<SelectCardEffect>();
 
-                bool CanUseCondition(Hashtable hashtable)
-                {
-                    return CardEffectCommons.CanTriggerOnDeletion(hashtable, card);
-                }
+                selectCardEffect.SetUp(
+                    canTargetCondition: CanPlay,
+                    canTargetCondition_ByPreSelecetedList: null,
+                    canEndSelectCondition: null,
+                    canNoSelect: () => true,
+                    selectCardCoroutine: SelectCardCoroutine,
+                    afterSelectCardCoroutine: null,
+                    message: "Select 1 [Canoweissmon] or level 4 or lower [Gammamon] in text Digimon to play.",
+                    maxCount: 1,
+                    canEndNotMax: false,
+                    isShowOpponent: true,
+                    mode: SelectCardEffect.Mode.Custom,
+                    root: SelectCardEffect.Root.Trash,
+                    customRootCardList: null,
+                    canLookReverseCard: true,
+                    selectPlayer: card.Owner,
+                    cardEffect: activateClass);
+
+                selectCardEffect.SetUpCustomMessage("Select 1 [Cannonweissmon] or level 4 or lower [Gammamon] in text Digimon to play.",
+                    "The opponent is selecting 1 [Cannonweissmon] or level 4 or lower [Gammamon] in text Digimon to play.");
+                selectCardEffect.SetUpCustomMessage_ShowCard("Played Card");
+
+                await selectCardEffect.Activate();
+
+                async Task SelectCardCoroutine(CardSource cardSource)
+                {
+                    selectedCards.Add(cardSource);
+
+                    await Task.CompletedTask;
+                }
+
+                await CardEffectCommons.PlayPermanentCards(
+                    cardSources: selectedCards,
+                    activateClass: activateClass,
+                    payCost: false,
+                    isTapped: false,
+                    root: SelectCardEffect.Root.Trash,
+                    activateETB: true);
+            }
+        }
 
-                bool CanPlay(CardSource cardSource)
-                {
-                    return cardSource.HasText("Gammamon") && cardSource.HasLevel && cardSource.Level <= 4 && cardSource.IsDigimon &&
-                           CardEffectCommons.CanPlayAsNewPermanent(cardSource: cardSource, payCost: false, cardEffect: activateClass);
-                }
+        #endregion
 
-                bool CanActivateCondition(Hashtable hashtable)
-                {
-                    return CardEffectCommons.CanActivateOnDeletion( hashtable, card) && CardEffectCommons.HasMatchConditionOwnersCardInTrash(card, CanPlay);
-                }
+        #region On deletion inherit
 
-                IEnumerator ActivateCoroutine(Hashtable hashtable)
-                {
-                    List<CardSource> selectedCards = new List<CardSource>();
+        if (timing == EffectTiming.OnDestroyedAnyone)
+        {
+            ActivateClass activateClass = new ActivateClass();
+            activateClass.SetUpICardEffect("Play Cannonweissmon or level 4 or lower gammamon in text digimon", CanUseCondition, card);
+            activateClass.SetUpActivateClass(CanActivateCondition, ActivateCoroutine, -1, false, EffectDescription());
+            activateClass.SetIsInheritedEffect(true);
+            cardEffects.Add(activateClass);
 
-                    SelectCardEffect selectCardEffect = GManager.instance.GetComponent<SelectCardEffect>();
+            string EffectDescription()
+            {
+                return "[On Deletion] You may play 1 level 4 or lower Digimon card with [Gammamon] in its text from your trash without paying the cost.";
+            }
 
-                    selectCardEffect.SetUp(
-                        canTargetCondition: CanPlay,
-                        canTargetCondition_ByPreSelecetedList: null,
-                        canEndSelectCondition: null,
-                        canNoSelect: () => true,
-                        selectCardCoroutine: SelectCardCoroutine,
-                        afterSelectCardCoroutine: null,
-                        message: "Select 1 level 4 or lower [Gammamon] in text Digimon to play.",
-                        maxCount: 1,
-                        canEndNotMax: false,
-                        isShowOpponent: true,
-                        mode: SelectCardEffect.Mode.Custom,
-                        root: SelectCardEffect.Root.Trash,
-                        customRootCardList: null,
-                        canLookReverseCard: true,
-                        selectPlayer: card.Owner,
-                        cardEffect: activateClass);
+            bool CanUseCondition(Hashtable hashtable)
+            {
+                return CardEffectCommons.CanTriggerOnDeletion(hashtable, card);
+            }
 
-                    selectCardEffect.SetUpCustomMessage("Select 1 level 4 or lower [Gammamon] in text Digimon to play.",
-                        "The opponent is selecting 1 level 4 or lower [Gammamon] in text Digimon to play.");
-                    selectCardEffect.SetUpCustomMessage_ShowCard("Played Card");
+            bool CanPlay(CardSource cardSource)
+            {
+                return cardSource.HasText("Gammamon") && cardSource.HasLevel && cardSource.Level <= 4 && cardSource.IsDigimon &&
+                       CardEffectCommons.CanPlayAsNewPermanent(cardSource: cardSource, payCost: false, cardEffect: activateClass);
+            }
 
-                    yield return StartCoroutine(selectCardEffect.Activate());
+            bool CanActivateCondition(Hashtable hashtable)
+            {
+                return CardEffectCommons.CanActivateOnDeletion(hashtable, card) && CardEffectCommons.HasMatchConditionOwnersCardInTrash(card, CanPlay);
+            }
 
-                    IEnumerator SelectCardCoroutine(CardSource cardSource)
-                    {
-                        selectedCards.Add(cardSource);
+            async Task ActivateCoroutine(Hashtable hashtable)
+            {
+                List<CardSource> selectedCards = new List<CardSource>();
 
-                        yield return null;
-                    }
+                SelectCardEffect selectCardEffect = GManager.instance.GetComponent<SelectCardEffect>();
 
-                    yield return ContinuousController.instance.StartCoroutine(CardEffectCommons.PlayPermanentCards(
-                        cardSources: selectedCards,
-                        activateClass: activateClass,
-                        payCost: false,
-                        isTapped: false,
-                        root: SelectCardEffect.Root.Trash,
-                        activateETB: true));
-                }
+                selectCardEffect.SetUp(
+                    canTargetCondition: CanPlay,
+                    canTargetCondition_ByPreSelecetedList: null,
+                    canEndSelectCondition: null,
+                    canNoSelect: () => true,
+                    selectCardCoroutine: SelectCardCoroutine,
+                    afterSelectCardCoroutine: null,
+                    message: "Select 1 level 4 or lower [Gammamon] in text Digimon to play.",
+                    maxCount: 1,
+                    canEndNotMax: false,
+                    isShowOpponent: true,
+                    mode: SelectCardEffect.Mode.Custom,
+                    root: SelectCardEffect.Root.Trash,
+                    customRootCardList: null,
+                    canLookReverseCard: true,
+                    selectPlayer: card.Owner,
+                    cardEffect: activateClass);
+
+                selectCardEffect.SetUpCustomMessage("Select 1 level 4 or lower [Gammamon] in text Digimon to play.",
+                    "The opponent is selecting 1 level 4 or lower [Gammamon] in text Digimon to play.");
+                selectCardEffect.SetUpCustomMessage_ShowCard("Played Card");
+
+                await selectCardEffect.Activate();
+
+                async Task SelectCardCoroutine(CardSource cardSource)
+                {
+                    selectedCards.Add(cardSource);
+
+                    await Task.CompletedTask;
+                }
+
+                await CardEffectCommons.PlayPermanentCards(
+                    cardSources: selectedCards,
+                    activateClass: activateClass,
+                    payCost: false,
+                    isTapped: false,
+                    root: SelectCardEffect.Root.Trash,
+                    activateETB: true);
             }
+        }
 
-            #endregion
+        #endregion
 
-            return cardEffects;
-        }
+        return cardEffects;
     }
 }
