--- DCGO/Assets/Scripts/CardEffect/EX9/White/EX9_074.cs	2026-05-18 20:03:15.449654400 +0900
+++ src/HeadlessDCGO.Engine/Assets/Scripts/CardEffect/EX9/White/EX9_074.cs	2026-07-29 01:02:23.219953494 +0900
@@ -1,539 +1,531 @@
+namespace HeadlessDCGO.Engine.Assets.Scripts.CardEffect.EX9.White;
+
 using System;
 using System.Collections;
 using System.Collections.Generic;
 using System.Linq;
-using UnityEngine;
+using System.Threading.Tasks;
+using HeadlessDCGO.Engine.Assets.Scripts.Script;
+using HeadlessDCGO.Engine.Assets.Scripts.Script.CardEffectCommons;
+using HeadlessDCGO.Engine.Assets.Scripts.Script.CardEffects;
+using HeadlessDCGO.Engine.Headless.Services;
 
-// Kimeramon
-namespace DCGO.CardEffects.EX9
+public class EX9_074 : CEntity_Effect
 {
-    public class EX9_074 : CEntity_Effect
+    public override List<ICardEffect> CardEffects(EffectTiming timing, CardSource card)
     {
-        public override List<ICardEffect> CardEffects(EffectTiming timing, CardSource card)
+        List<ICardEffect> cardEffects = new List<ICardEffect>();
+
+        #region Assembly
+        if (timing == EffectTiming.None)
         {
-            List<ICardEffect> cardEffects = new List<ICardEffect>();
+            AddAssemblyConditionClass addAssemblyConditionClass = new AddAssemblyConditionClass();
+            addAssemblyConditionClass.SetUpICardEffect($"Assembly", CanUseCondition, card);
+            addAssemblyConditionClass.SetUpAddAssemblyConditionClass(getAssemblyCondition: GetAssembly);
+            addAssemblyConditionClass.SetNotShowUI(true);
+            cardEffects.Add(addAssemblyConditionClass);
 
-            #region Assembly
-            if (timing == EffectTiming.None)
+            bool CanUseCondition(Hashtable hashtable)
             {
-                AddAssemblyConditionClass addAssemblyConditionClass = new AddAssemblyConditionClass();
-                addAssemblyConditionClass.SetUpICardEffect($"Assembly", CanUseCondition, card);
-                addAssemblyConditionClass.SetUpAddAssemblyConditionClass(getAssemblyCondition: GetAssembly);
-                addAssemblyConditionClass.SetNotShowUI(true);
-                cardEffects.Add(addAssemblyConditionClass);
+                return true;
+            }
 
-                bool CanUseCondition(Hashtable hashtable)
+            AssemblyCondition GetAssembly(CardSource cardSource)
+            {
+                if (cardSource == card)
                 {
-                    return true;
-                }
+                    AssemblyConditionElement element = new AssemblyConditionElement(CanSelectCardCondition);
 
-                AssemblyCondition GetAssembly(CardSource cardSource)
-                {
-                    if (cardSource == card)
+                    bool CanSelectCardCondition(CardSource cardSource)
                     {
-                        AssemblyConditionElement element = new AssemblyConditionElement(CanSelectCardCondition);
-
-                        bool CanSelectCardCondition(CardSource cardSource)
+                        if (cardSource != null)
                         {
-                            if (cardSource != null)
+                            if (cardSource.Owner == card.Owner)
                             {
-                                if (cardSource.Owner == card.Owner)
+                                if (cardSource.IsDigimon)
                                 {
-                                    if (cardSource.IsDigimon)
+                                    if (cardSource.IsLevel(4) || cardSource.Level_Assembly.Contains(4))
                                     {
-                                        if (cardSource.IsLevel4 || cardSource.Level_Assembly.Contains(4))
+                                        if (cardSource.HasDMTraits)
                                         {
-                                            if (cardSource.HasDMTraits)
-                                            {
-                                                return true;
-                                            }
+                                            return true;
                                         }
                                     }
                                 }
                             }
-
-                            return false;
                         }
 
-                        bool CanTargetCondition_ByPreSelecetedList(List<CardSource> cardSources, CardSource cardSource)
-                        {
-                            List<string> cardNames = new List<string>();
+                        return false;
+                    }
 
-                            foreach (CardSource cardSource1 in cardSources)
+                    bool CanTargetCondition_ByPreSelecetedList(List<CardSource> cardSources, CardSource cardSource)
+                    {
+                        List<string> cardNames = new List<string>();
+
+                        foreach (CardSource cardSource1 in cardSources)
+                        {
+                            foreach (string cardName in cardSource1.CardNames)
                             {
-                                foreach (string cardName in cardSource1.CardNames)
+                                if (!cardNames.Contains(cardName))
                                 {
-                                    if (!cardNames.Contains(cardName))
-                                    {
-                                        cardNames.Add(cardName);
-                                    }
+                                    cardNames.Add(cardName);
                                 }
                             }
-
-                            if (cardSource.CardNames.Count((cardName) => cardNames.Contains(cardName)) >= 1)
-                            {
-                                return false;
-                            }
-
-                            return true;
                         }
 
-                        AssemblyCondition assemblyCondition = new AssemblyCondition(
-                            element:element,
-                            CanTargetCondition_ByPreSelecetedList: CanTargetCondition_ByPreSelecetedList,
-                            selectMessage: "7 level 4 [DM] trait Digimon cards w/different names",
-                            elementCount: 7,
-                            reduceCost: 7);
+                        if (cardSource.CardNames.Count((cardName) => cardNames.Contains(cardName)) >= 1)
+                        {
+                            return false;
+                        }
 
-                        return assemblyCondition;
+                        return true;
                     }
 
-                    return null;
+                    AssemblyCondition assemblyCondition = new AssemblyCondition(
+                        element: element,
+                        CanTargetCondition_ByPreSelecetedList: CanTargetCondition_ByPreSelecetedList,
+                        selectMessage: "7 level 4 [DM] trait Digimon cards w/different names",
+                        elementCount: 7,
+                        reduceCost: 7);
+
+                    return assemblyCondition;
                 }
+
+                return null;
             }
-            #endregion
+        }
+        #endregion
 
-            #region Sec +1
+        #region Sec +1
+        if (timing == EffectTiming.None)
+        {
+            cardEffects.Add(CardEffectFactory.ChangeSelfSAttackStaticEffect(
+            changeValue: 1,
+            isInheritedEffect: false,
+            card: card,
+            condition: null));
+        }
+        #endregion
+
+        #region Rush
+        if (timing == EffectTiming.None)
+        {
+            cardEffects.Add(CardEffectFactory.RushSelfStaticEffect(isInheritedEffect: false, card: card, condition: null));
+        }
+        #endregion
+
+        #region On Play
+        if (timing == EffectTiming.OnEnterFieldAnyone)
+        {
+            ActivateClass activateClass = new ActivateClass();
+            activateClass.SetUpICardEffect("Place digimon from trash as top source card, delete digimon", CanUseCondition, card);
+            activateClass.SetUpActivateClass(CanActivateCondition, ActivateCoroutine, -1, false, EffectDiscription());
+            cardEffects.Add(activateClass);
 
-            if (timing == EffectTiming.None)
+            string EffectDiscription()
             {
-                cardEffects.Add(CardEffectFactory.ChangeSelfSAttackStaticEffect(
-                changeValue: 1,
-                isInheritedEffect: false,
-                card: card,
-                condition: null));
+                return "[On Play] You may place 1 level 4 or lower [DM] trait Digimon card from your trash as this Digimon's top digivolution card. Then, delete 1 of your opponent's Digimon with the same color as any of this Digimon's digivolution cards. If this Digimon has 6 or more colors in its digivolution cards, instead delete 1 of each of your opponent's Digimon with different colors.";
             }
 
-            #endregion
-
-            #region Rush
+            bool CanUseCondition(Hashtable hashtable)
+            {
+                return CardEffectCommons.CanTriggerOnPlay(hashtable, card);
+            }
 
-            if (timing == EffectTiming.None)
+            bool CanActivateCondition(Hashtable hashtable)
             {
-                cardEffects.Add(CardEffectFactory.RushSelfStaticEffect(isInheritedEffect: false, card: card, condition: null));
+                return CardEffectCommons.IsExistOnBattleArea(card);
             }
 
-            #endregion
+            bool CanSelectCardCondition(CardSource cardSource)
+            {
+                return cardSource.IsDigimon && cardSource.HasLevel && cardSource.Level <= 4 && cardSource.HasDMTraits;
+            }
 
-            #region On Play
+            bool CanSelectPermanentCondition(Permanent permanent, List<CardColor> cardColors)
+            {
+                return CardEffectCommons.IsPermanentExistsOnOpponentBattleAreaDigimon(permanent, card) && CardSource.ToCardColorList(permanent.TopCard.CardColors).Exists(color => cardColors.Contains(color));
+            }
 
-            if (timing == EffectTiming.OnEnterFieldAnyone)
+            bool CanEndSelectCondition(List<Permanent> permanents)
             {
-                ActivateClass activateClass = new ActivateClass();
-                activateClass.SetUpICardEffect("Place digimon from trash as top source card, delete digimon", CanUseCondition, card);
-                activateClass.SetUpActivateClass(CanActivateCondition, ActivateCoroutine, -1, false, EffectDiscription());
-                cardEffects.Add(activateClass);
+                if (permanents.Count <= 0)
+                    return false;
 
-                string EffectDiscription()
-                {
-                    return "[On Play] You may place 1 level 4 or lower [DM] trait Digimon card from your trash as this Digimon's top digivolution card. Then, delete 1 of your opponent's Digimon with the same color as any of this Digimon's digivolution cards. If this Digimon has 6 or more colors in its digivolution cards, instead delete 1 of each of your opponent's Digimon with different colors.";
-                }
+                return true;
+            }
 
-                bool CanUseCondition(Hashtable hashtable)
+            async Task ActivateCoroutine(Hashtable hashtable)
+            {
+                if (CardEffectCommons.HasMatchConditionOwnersCardInTrash(card, CanSelectCardCondition))
                 {
-                    return CardEffectCommons.CanTriggerOnPlay(hashtable, card);
-                }
+                    CardSource? selectedCard = null;
+                    int maxCount = Math.Min(1, CardEffectCommons.MatchConditionOwnersCardCountInTrash(card, CanSelectCardCondition));
+                    SelectCardEffect selectCardEffect = GManager.instance.GetComponent<SelectCardEffect>();
 
-                bool CanActivateCondition(Hashtable hashtable)
-                {
-                    return CardEffectCommons.IsExistOnBattleArea(card);
-                }
+                    Task SelectCardCoroutine(CardSource cardSource)
+                    {
+                        selectedCard = cardSource;
+                        return Task.CompletedTask;
+                    }
 
-                bool CanSelectCardCondition(CardSource cardSource)
-                {
-                    return cardSource.IsDigimon && cardSource.HasLevel && cardSource.Level <= 4 && cardSource.HasDMTraits;
-                }
+                    selectCardEffect.SetUp(
+                                canTargetCondition: CanSelectCardCondition,
+                                canTargetCondition_ByPreSelecetedList: null,
+                                canEndSelectCondition: null,
+                                canNoSelect: () => true,
+                                selectCardCoroutine: SelectCardCoroutine,
+                                afterSelectCardCoroutine: null,
+                                message: "Select 1 [DM] digimon to add as top source",
+                                maxCount: maxCount,
+                                canEndNotMax: true,
+                                isShowOpponent: true,
+                                mode: SelectCardEffect.Mode.Custom,
+                                root: SelectCardEffect.Root.Trash,
+                                customRootCardList: null,
+                                canLookReverseCard: true,
+                                selectPlayer: card.Owner,
+                                cardEffect: activateClass);
 
-                bool CanSelectPermanentCondition(Permanent permanent, List<CardColor> cardColors)
-                {
-                    return CardEffectCommons.IsPermanentExistsOnOpponentBattleAreaDigimon(permanent, card) && permanent.TopCard.CardColors.Exists(color => cardColors.Contains(color));
-                }
+                    selectCardEffect.SetUpCustomMessage("Select 1 [DM] digimon to add as top source", "The opponent is selecting 1 digimon to add as top source");
+                    await selectCardEffect.Activate();
 
-                bool CanEndSelectCondition(List<Permanent> permanents)
-                {
-                    if (permanents.Count <= 0)
-                        return false;
+                    if (selectedCard != null)
+                    {
+                        await CardObjectController.AddExecutingCard(selectedCard);
+                        await ICardEffect.ResolvePermanentOfThisCard(card).AddDigivolutionCardsTop(new List<CardSource>() { selectedCard }, card.InstanceId);
+                    }
+                }
 
-                    return true;
+                List<CardColor> colours = ICardEffect.ResolvePermanentOfThisCard(card).DigivolutionCards
+                            .Where(x => !x.IsFlipped)
+                            .SelectMany(e => CardSource.ToCardColorList(e.CardColors))
+                            .Distinct()
+                            .ToList();
+
+                if (colours.Count <= 5 && CardEffectCommons.HasMatchConditionOpponentsPermanent(card, permanent => CanSelectPermanentCondition(permanent, colours)))
+                {
+                    int maxCount1 = Math.Min(1, CardEffectCommons.MatchConditionPermanentCount(card, permanent => CanSelectPermanentCondition(permanent, colours)));
+                    SelectPermanentEffect selectPermanentEffect = GManager.instance.GetComponent<SelectPermanentEffect>();
+
+                    selectPermanentEffect.SetUp(
+                        selectPlayer: card.Owner,
+                        canTargetCondition: p => CanSelectPermanentCondition(p, colours),
+                        canTargetCondition_ByPreSelecetedList: null,
+                        canEndSelectCondition: null,
+                        maxCount: maxCount1,
+                        canNoSelect: false,
+                        canEndNotMax: false,
+                        selectPermanentCoroutine: null,
+                        afterSelectPermanentCoroutine: null,
+                        mode: SelectPermanentEffect.Mode.Destroy,
+                        cardEffect: activateClass);
+
+                    await selectPermanentEffect.Activate();
                 }
 
-                IEnumerator ActivateCoroutine(Hashtable hashtable)
+                if (colours.Count >= 6 && CardEffectCommons.HasMatchConditionOpponentsPermanent(card, permanent => CardEffectCommons.IsPermanentExistsOnOpponentBattleAreaDigimon(permanent, card)))
                 {
-                    if (CardEffectCommons.HasMatchConditionOwnersCardInTrash(card, CanSelectCardCondition))
-                    {
-                        CardSource selectedCard = null;
-                        int maxCount = Math.Min(1, CardEffectCommons.MatchConditionOwnersCardCountInTrash(card, CanSelectCardCondition));
-                        SelectCardEffect selectCardEffect = GManager.instance.GetComponent<SelectCardEffect>();
+                    List<Permanent> permanentToDelete = new List<Permanent>();
+                    List<CardColor> selectableColors = new List<CardColor>();
 
-                        selectCardEffect.SetUp(
-                                    canTargetCondition: CanSelectCardCondition,
-                                    canTargetCondition_ByPreSelecetedList: null,
-                                    canEndSelectCondition: null,
-                                    canNoSelect: () => true,
-                                    selectCardCoroutine: SelectCardCoroutine,
-                                    afterSelectCardCoroutine: null,
-                                    message: "Select 1 [DM] digimon to add as top source",
-                                    maxCount: maxCount,
-                                    canEndNotMax: true,
-                                    isShowOpponent: true,
-                                    mode: SelectCardEffect.Mode.Custom,
-                                    root: SelectCardEffect.Root.Trash,
-                                    customRootCardList: null,
-                                    canLookReverseCard: true,
-                                    selectPlayer: card.Owner,
-                                    cardEffect: activateClass);
+                    foreach (string cardColor in DataBase.CardColorNameDictionary.Values)
+                    {
 
-                        selectCardEffect.SetUpCustomMessage("Select 1 [DM] digimon to add as top source", "The opponent is selecting 1 digimon to add as top source");
-                        yield return StartCoroutine(selectCardEffect.Activate());
+                        CardColor compareColor = DictionaryUtility.GetCardColor(cardColor.Trim(), DataBase.CardColorNameDictionary);
 
-                        IEnumerator SelectCardCoroutine(CardSource cardSource)
+                        bool CanSelectOpponentDigimon(Permanent permanent)
                         {
-                            selectedCard = cardSource;
-                            yield return null;
+                            return CardEffectCommons.IsPermanentExistsOnOpponentBattleAreaDigimon(permanent, card) &&
+                                   CardSource.ToCardColorList(permanent.TopCard.CardColors).Contains(compareColor);
                         }
 
-                        if (selectedCard != null)
+                        if (CardEffectCommons.HasMatchConditionPermanent(card, CanSelectOpponentDigimon))
                         {
-                            yield return ContinuousController.instance.StartCoroutine(CardObjectController.AddExecutingCard(selectedCard));
-                            yield return ContinuousController.instance.StartCoroutine(card.PermanentOfThisCard().AddDigivolutionCardsTop(new List<CardSource>() { selectedCard }, activateClass));
+                            selectableColors.Add(compareColor);
                         }
                     }
 
-                    List<CardColor> colours = card.PermanentOfThisCard().DigivolutionCards
-                                .Filter(x => !x.IsFlipped)
-                                .SelectMany(e => e.CardColors)
-                                .Distinct()
-                                .ToList();
-
-                    if (colours.Count <= 5 && CardEffectCommons.HasMatchConditionOpponentsPermanent(card, permanent => CanSelectPermanentCondition(permanent, colours)))
+                    foreach (CardColor deletableColor in selectableColors)
                     {
-                        int maxCount1 = Math.Min(1, CardEffectCommons.MatchConditionPermanentCount(permanent => CanSelectPermanentCondition(permanent, colours)));
-                        SelectPermanentEffect selectPermanentEffect = GManager.instance.GetComponent<SelectPermanentEffect>();
-
-                        selectPermanentEffect.SetUp(
-                            selectPlayer: card.Owner,
-                            canTargetCondition: permanent => CanSelectPermanentCondition(permanent, colours),
-                            canTargetCondition_ByPreSelecetedList: null,
-                            canEndSelectCondition: null,
-                            maxCount: maxCount1,
-                            canNoSelect: false,
-                            canEndNotMax: false,
-                            selectPermanentCoroutine: null,
-                            afterSelectPermanentCoroutine: null,
-                            mode: SelectPermanentEffect.Mode.Destroy,
-                            cardEffect: activateClass);
 
-                        yield return ContinuousController.instance.StartCoroutine(selectPermanentEffect.Activate());
-                    }
-
-                    if (colours.Count >= 6 && CardEffectCommons.HasMatchConditionOpponentsPermanent(card, permanent => CardEffectCommons.IsPermanentExistsOnOpponentBattleAreaDigimon(permanent, card)))
-                    {
-                        List<Permanent> permanentToDelete = new List<Permanent>();
-                        List<CardColor> selectableColors = new List<CardColor>();
+                        bool CanSelectOpponentDigimon(Permanent permanent)
+                        {
+                            return CardEffectCommons.IsPermanentExistsOnOpponentBattleAreaDigimon(permanent, card) &&
+                                   CardSource.ToCardColorList(permanent.TopCard.CardColors).Contains(deletableColor) &&
+                                   CanTargetCondition_ByPreSelecetedList(permanentToDelete, permanent);
+                        }
 
-                        foreach (string cardColor in DataBase.CardColorNameDictionary.Values)
+                        if (CardEffectCommons.HasMatchConditionPermanent(card, CanSelectOpponentDigimon))
                         {
+                            SelectPermanentEffect selectPermanentEffect = GManager.instance.GetComponent<SelectPermanentEffect>();
 
-                            CardColor compareColor = DictionaryUtility.GetCardColor(cardColor.Trim(), DataBase.CardColorNameDictionary);
+                            selectPermanentEffect.SetUp(
+                                selectPlayer: card.Owner,
+                                canTargetCondition: CanSelectOpponentDigimon,
+                                canTargetCondition_ByPreSelecetedList: CanTargetCondition_ByPreSelecetedList,
+                                canEndSelectCondition: CanEndSelectCondition,
+                                maxCount: 1,
+                                canNoSelect: false,
+                                canEndNotMax: true,
+                                selectPermanentCoroutine: DigimonToDelete,
+                                afterSelectPermanentCoroutine: null,
+                                mode: SelectPermanentEffect.Mode.Custom,
+                                cardEffect: activateClass);
 
-                            bool CanSelectOpponentDigimon(Permanent permanent)
-                            {
-                                return CardEffectCommons.IsPermanentExistsOnOpponentBattleAreaDigimon(permanent, card) &&
-                                       permanent.TopCard.CardColors.Contains(compareColor);
-                            }
+                            selectPermanentEffect.SetUpCustomMessage($"Select 1 {deletableColor.ToString()} Digimon to delete", $"Opponent is selecting 1 {deletableColor.ToString()} Digimon to delete");
 
-                            if (CardEffectCommons.HasMatchConditionPermanent(CanSelectOpponentDigimon))
-                            {
-                                selectableColors.Add(compareColor);
-                            }
+                            await selectPermanentEffect.Activate();
                         }
+                    }
 
-                        foreach(CardColor deletableColor in selectableColors)
-                        {
-
-                            bool CanSelectOpponentDigimon(Permanent permanent)
-                            {
-                                return CardEffectCommons.IsPermanentExistsOnOpponentBattleAreaDigimon(permanent, card) &&
-                                       permanent.TopCard.CardColors.Contains(deletableColor) &&
-                                       CanTargetCondition_ByPreSelecetedList(permanentToDelete, permanent);
-                            }
+                    bool CanTargetCondition_ByPreSelecetedList(List<Permanent> permanents, Permanent permanent)
+                    {
+                        if (permanentToDelete.Contains(permanent))
+                            return false;
 
-                            if (CardEffectCommons.HasMatchConditionPermanent(CanSelectOpponentDigimon))
-                            {
-                                SelectPermanentEffect selectPermanentEffect = GManager.instance.GetComponent<SelectPermanentEffect>();
+                        return true;
+                    }
 
-                                selectPermanentEffect.SetUp(
-                                    selectPlayer: card.Owner,
-                                    canTargetCondition: CanSelectOpponentDigimon,
-                                    canTargetCondition_ByPreSelecetedList: CanTargetCondition_ByPreSelecetedList,
-                                    canEndSelectCondition: CanEndSelectCondition,
-                                    maxCount: 1,
-                                    canNoSelect: false,
-                                    canEndNotMax: true,
-                                    selectPermanentCoroutine: DigimonToDelete,
-                                    afterSelectPermanentCoroutine: null,
-                                    mode: SelectPermanentEffect.Mode.Custom,
-                                    cardEffect: activateClass);
+                    Task DigimonToDelete(Permanent permanent)
+                    {
+                        if (permanent != null)
+                            permanentToDelete.Add(permanent);
 
-                                selectPermanentEffect.SetUpCustomMessage($"Select 1 {deletableColor.ToString()} Digimon to delete", $"Opponent is selecting 1 {deletableColor.ToString()} Digimon to delete");
+                        return Task.CompletedTask;
+                    }
 
-                                yield return ContinuousController.instance.StartCoroutine(selectPermanentEffect.Activate());
-                            }
-                        }
+                    if (permanentToDelete.Count > 0)
+                    {
+                        await new DestroyPermanentsClass(permanentToDelete, hashtable).Destroy();
+                    }
+                }
+            }
+        }
+        #endregion
 
-                        bool CanTargetCondition_ByPreSelecetedList(List<Permanent> permanents, Permanent permanent)
-                        {
-                            if (permanentToDelete.Contains(permanent)) 
-                                return false;
-                            
-                            return true;
-                        }
+        #region When Digivolving
+        if (timing == EffectTiming.OnEnterFieldAnyone)
+        {
+            ActivateClass activateClass = new ActivateClass();
+            activateClass.SetUpICardEffect("Place digimon from trash as top source card, delete digimon", CanUseCondition, card);
+            activateClass.SetUpActivateClass(CanActivateCondition, ActivateCoroutine, -1, false, EffectDiscription());
+            cardEffects.Add(activateClass);
 
-                        IEnumerator DigimonToDelete (Permanent permanent)
-                        {
-                            if(permanent != null)
-                                permanentToDelete.Add(permanent);
+            string EffectDiscription()
+            {
+                return "[When Digivolving] You may place 1 level 4 or lower [DM] trait Digimon card from your trash as this Digimon's top digivolution card. Then, delete 1 of your opponent's Digimon with the same color as any of this Digimon's digivolution cards. If this Digimon has 6 or more colors in its digivolution cards, instead delete 1 of each of your opponent's Digimon with different colors.";
+            }
 
-                            yield return null;
-                        }
+            bool CanUseCondition(Hashtable hashtable)
+            {
+                return CardEffectCommons.CanTriggerWhenDigivolving(hashtable, card);
+            }
 
-                        if(permanentToDelete.Count > 0)
-                        {
-                            yield return ContinuousController.instance.StartCoroutine(new DestroyPermanentsClass(permanentToDelete, hashtable).Destroy());
-                        }
-                    }
-                }
+            bool CanActivateCondition(Hashtable hashtable)
+            {
+                return CardEffectCommons.IsExistOnBattleArea(card);
             }
 
-            #endregion
+            bool CanSelectCardCondition(CardSource cardSource)
+            {
+                return cardSource.IsDigimon && cardSource.HasLevel && cardSource.Level <= 4 && cardSource.HasDMTraits;
+            }
 
-            #region When Digivolving
+            bool CanSelectPermanentCondition(Permanent permanent, List<CardColor> cardColors)
+            {
+                return CardEffectCommons.IsPermanentExistsOnOpponentBattleAreaDigimon(permanent, card) && CardSource.ToCardColorList(permanent.TopCard.CardColors).Exists(color => cardColors.Contains(color));
+            }
 
-            if (timing == EffectTiming.OnEnterFieldAnyone)
+            bool CanEndSelectCondition(List<Permanent> permanents)
             {
-                ActivateClass activateClass = new ActivateClass();
-                activateClass.SetUpICardEffect("Place digimon from trash as top source card, delete digimon", CanUseCondition, card);
-                activateClass.SetUpActivateClass(CanActivateCondition, ActivateCoroutine, -1, false, EffectDiscription());
-                cardEffects.Add(activateClass);
+                if (permanents.Count <= 0)
+                    return false;
 
-                string EffectDiscription()
-                {
-                    return "[When Digivolving] You may place 1 level 4 or lower [DM] trait Digimon card from your trash as this Digimon's top digivolution card. Then, delete 1 of your opponent's Digimon with the same color as any of this Digimon's digivolution cards. If this Digimon has 6 or more colors in its digivolution cards, instead delete 1 of each of your opponent's Digimon with different colors.";
-                }
+                return true;
+            }
 
-                bool CanUseCondition(Hashtable hashtable)
+            async Task ActivateCoroutine(Hashtable hashtable)
+            {
+                if (CardEffectCommons.HasMatchConditionOwnersCardInTrash(card, CanSelectCardCondition))
                 {
-                    return CardEffectCommons.CanTriggerWhenDigivolving(hashtable, card);
-                }
+                    CardSource? selectedCard = null;
+                    int maxCount = Math.Min(1, CardEffectCommons.MatchConditionOwnersCardCountInTrash(card, CanSelectCardCondition));
+                    SelectCardEffect selectCardEffect = GManager.instance.GetComponent<SelectCardEffect>();
 
-                bool CanActivateCondition(Hashtable hashtable)
-                {
-                    return CardEffectCommons.IsExistOnBattleArea(card);
-                }
+                    Task SelectCardCoroutine(CardSource cardSource)
+                    {
+                        selectedCard = cardSource;
+                        return Task.CompletedTask;
+                    }
 
-                bool CanSelectCardCondition(CardSource cardSource)
-                {
-                    return cardSource.IsDigimon && cardSource.HasLevel && cardSource.Level <= 4 && cardSource.HasDMTraits;
-                }
+                    selectCardEffect.SetUp(
+                                canTargetCondition: CanSelectCardCondition,
+                                canTargetCondition_ByPreSelecetedList: null,
+                                canEndSelectCondition: null,
+                                canNoSelect: () => true,
+                                selectCardCoroutine: SelectCardCoroutine,
+                                afterSelectCardCoroutine: null,
+                                message: "Select 1 [DM] digimon to add as top source",
+                                maxCount: maxCount,
+                                canEndNotMax: true,
+                                isShowOpponent: true,
+                                mode: SelectCardEffect.Mode.Custom,
+                                root: SelectCardEffect.Root.Trash,
+                                customRootCardList: null,
+                                canLookReverseCard: true,
+                                selectPlayer: card.Owner,
+                                cardEffect: activateClass);
 
-                bool CanSelectPermanentCondition(Permanent permanent, List<CardColor> cardColors)
-                {
-                    return CardEffectCommons.IsPermanentExistsOnOpponentBattleAreaDigimon(permanent, card) && permanent.TopCard.CardColors.Exists(color => cardColors.Contains(color));
-                }
+                    selectCardEffect.SetUpCustomMessage("Select 1 [DM] digimon to add as top source", "The opponent is selecting 1 digimon to add as top source");
+                    await selectCardEffect.Activate();
 
-                bool CanEndSelectCondition(List<Permanent> permanents)
-                {
-                    if (permanents.Count <= 0)
-                        return false;
+                    if (selectedCard != null)
+                    {
+                        await CardObjectController.AddExecutingCard(selectedCard);
+                        await ICardEffect.ResolvePermanentOfThisCard(card).AddDigivolutionCardsTop(new List<CardSource>() { selectedCard }, card.InstanceId);
+                    }
+                }
 
-                    return true;
+                List<CardColor> colours = ICardEffect.ResolvePermanentOfThisCard(card).DigivolutionCards
+                            .Where(x => !x.IsFlipped)
+                            .SelectMany(e => CardSource.ToCardColorList(e.CardColors))
+                            .Distinct()
+                            .ToList();
+
+                if (colours.Count <= 5 && CardEffectCommons.HasMatchConditionOpponentsPermanent(card, permanent => CanSelectPermanentCondition(permanent, colours)))
+                {
+                    int maxCount1 = Math.Min(1, CardEffectCommons.MatchConditionPermanentCount(card, permanent => CanSelectPermanentCondition(permanent, colours)));
+                    SelectPermanentEffect selectPermanentEffect = GManager.instance.GetComponent<SelectPermanentEffect>();
+
+                    selectPermanentEffect.SetUp(
+                        selectPlayer: card.Owner,
+                        canTargetCondition: p => CanSelectPermanentCondition(p, colours),
+                        canTargetCondition_ByPreSelecetedList: null,
+                        canEndSelectCondition: null,
+                        maxCount: maxCount1,
+                        canNoSelect: false,
+                        canEndNotMax: false,
+                        selectPermanentCoroutine: null,
+                        afterSelectPermanentCoroutine: null,
+                        mode: SelectPermanentEffect.Mode.Destroy,
+                        cardEffect: activateClass);
+
+                    await selectPermanentEffect.Activate();
                 }
 
-                IEnumerator ActivateCoroutine(Hashtable hashtable)
+                if (colours.Count >= 6 && CardEffectCommons.HasMatchConditionOpponentsPermanent(card, permanent => CardEffectCommons.IsPermanentExistsOnOpponentBattleAreaDigimon(permanent, card)))
                 {
-                    if (CardEffectCommons.HasMatchConditionOwnersCardInTrash(card, CanSelectCardCondition))
-                    {
-                        CardSource selectedCard = null;
-                        int maxCount = Math.Min(1, CardEffectCommons.MatchConditionOwnersCardCountInTrash(card, CanSelectCardCondition));
-                        SelectCardEffect selectCardEffect = GManager.instance.GetComponent<SelectCardEffect>();
+                    List<Permanent> permanentToDelete = new List<Permanent>();
+                    List<CardColor> selectableColors = new List<CardColor>();
 
-                        selectCardEffect.SetUp(
-                                    canTargetCondition: CanSelectCardCondition,
-                                    canTargetCondition_ByPreSelecetedList: null,
-                                    canEndSelectCondition: null,
-                                    canNoSelect: () => true,
-                                    selectCardCoroutine: SelectCardCoroutine,
-                                    afterSelectCardCoroutine: null,
-                                    message: "Select 1 [DM] digimon to add as top source",
-                                    maxCount: maxCount,
-                                    canEndNotMax: true,
-                                    isShowOpponent: true,
-                                    mode: SelectCardEffect.Mode.Custom,
-                                    root: SelectCardEffect.Root.Trash,
-                                    customRootCardList: null,
-                                    canLookReverseCard: true,
-                                    selectPlayer: card.Owner,
-                                    cardEffect: activateClass);
+                    foreach (string cardColor in DataBase.CardColorNameDictionary.Values)
+                    {
 
-                        selectCardEffect.SetUpCustomMessage("Select 1 [DM] digimon to add as top source", "The opponent is selecting 1 digimon to add as top source");
-                        yield return StartCoroutine(selectCardEffect.Activate());
+                        CardColor compareColor = DictionaryUtility.GetCardColor(cardColor.Trim(), DataBase.CardColorNameDictionary);
 
-                        IEnumerator SelectCardCoroutine(CardSource cardSource)
+                        bool CanSelectOpponentDigimon(Permanent permanent)
                         {
-                            selectedCard = cardSource;
-                            yield return null;
+                            return CardEffectCommons.IsPermanentExistsOnOpponentBattleAreaDigimon(permanent, card) &&
+                                   CardSource.ToCardColorList(permanent.TopCard.CardColors).Contains(compareColor);
                         }
 
-                        if (selectedCard != null)
+                        if (CardEffectCommons.HasMatchConditionPermanent(card, CanSelectOpponentDigimon))
                         {
-                            yield return ContinuousController.instance.StartCoroutine(CardObjectController.AddExecutingCard(selectedCard));
-                            yield return ContinuousController.instance.StartCoroutine(card.PermanentOfThisCard().AddDigivolutionCardsTop(new List<CardSource>() { selectedCard }, activateClass));
+                            selectableColors.Add(compareColor);
                         }
                     }
 
-                    List<CardColor> colours = card.PermanentOfThisCard().DigivolutionCards
-                                .Filter(x => !x.IsFlipped)
-                                .SelectMany(e => e.CardColors)
-                                .Distinct()
-                                .ToList();
-
-                    if (colours.Count <= 5 && CardEffectCommons.HasMatchConditionOpponentsPermanent(card, permanent => CanSelectPermanentCondition(permanent, colours)))
+                    foreach (CardColor deletableColor in selectableColors)
                     {
-                        int maxCount1 = Math.Min(1, CardEffectCommons.MatchConditionPermanentCount(permanent => CanSelectPermanentCondition(permanent, colours)));
-                        SelectPermanentEffect selectPermanentEffect = GManager.instance.GetComponent<SelectPermanentEffect>();
-
-                        selectPermanentEffect.SetUp(
-                            selectPlayer: card.Owner,
-                            canTargetCondition: permanent => CanSelectPermanentCondition(permanent, colours),
-                            canTargetCondition_ByPreSelecetedList: null,
-                            canEndSelectCondition: null,
-                            maxCount: maxCount1,
-                            canNoSelect: false,
-                            canEndNotMax: false,
-                            selectPermanentCoroutine: null,
-                            afterSelectPermanentCoroutine: null,
-                            mode: SelectPermanentEffect.Mode.Destroy,
-                            cardEffect: activateClass);
-
-                        yield return ContinuousController.instance.StartCoroutine(selectPermanentEffect.Activate());
-                    }
 
-                    if (colours.Count >= 6 && CardEffectCommons.HasMatchConditionOpponentsPermanent(card, permanent => CardEffectCommons.IsPermanentExistsOnOpponentBattleAreaDigimon(permanent, card)))
-                    {
-                        List<Permanent> permanentToDelete = new List<Permanent>();
-                        List<CardColor> selectableColors = new List<CardColor>();
-
-                        foreach (string cardColor in DataBase.CardColorNameDictionary.Values)
+                        bool CanSelectOpponentDigimon(Permanent permanent)
                         {
-
-                            CardColor compareColor = DictionaryUtility.GetCardColor(cardColor.Trim(), DataBase.CardColorNameDictionary);
-
-                            bool CanSelectOpponentDigimon(Permanent permanent)
-                            {
-                                return CardEffectCommons.IsPermanentExistsOnOpponentBattleAreaDigimon(permanent, card) &&
-                                       permanent.TopCard.CardColors.Contains(compareColor);
-                            }
-
-                            if (CardEffectCommons.HasMatchConditionPermanent(CanSelectOpponentDigimon))
-                            {
-                                selectableColors.Add(compareColor);
-                            }
+                            return CardEffectCommons.IsPermanentExistsOnOpponentBattleAreaDigimon(permanent, card) &&
+                                   CardSource.ToCardColorList(permanent.TopCard.CardColors).Contains(deletableColor) &&
+                                   CanTargetCondition_ByPreSelecetedList(permanentToDelete, permanent);
                         }
 
-                        foreach (CardColor deletableColor in selectableColors)
+                        if (CardEffectCommons.HasMatchConditionPermanent(card, CanSelectOpponentDigimon))
                         {
+                            SelectPermanentEffect selectPermanentEffect = GManager.instance.GetComponent<SelectPermanentEffect>();
 
-                            bool CanSelectOpponentDigimon(Permanent permanent)
-                            {
-                                return CardEffectCommons.IsPermanentExistsOnOpponentBattleAreaDigimon(permanent, card) &&
-                                       permanent.TopCard.CardColors.Contains(deletableColor) &&
-                                       CanTargetCondition_ByPreSelecetedList(permanentToDelete, permanent);
-                            }
-
-                            if (CardEffectCommons.HasMatchConditionPermanent(CanSelectOpponentDigimon))
-                            {
-                                SelectPermanentEffect selectPermanentEffect = GManager.instance.GetComponent<SelectPermanentEffect>();
-
-                                selectPermanentEffect.SetUp(
-                                    selectPlayer: card.Owner,
-                                    canTargetCondition: CanSelectOpponentDigimon,
-                                    canTargetCondition_ByPreSelecetedList: CanTargetCondition_ByPreSelecetedList,
-                                    canEndSelectCondition: CanEndSelectCondition,
-                                    maxCount: 1,
-                                    canNoSelect: false,
-                                    canEndNotMax: true,
-                                    selectPermanentCoroutine: DigimonToDelete,
-                                    afterSelectPermanentCoroutine: null,
-                                    mode: SelectPermanentEffect.Mode.Custom,
-                                    cardEffect: activateClass);
+                            selectPermanentEffect.SetUp(
+                                selectPlayer: card.Owner,
+                                canTargetCondition: CanSelectOpponentDigimon,
+                                canTargetCondition_ByPreSelecetedList: CanTargetCondition_ByPreSelecetedList,
+                                canEndSelectCondition: CanEndSelectCondition,
+                                maxCount: 1,
+                                canNoSelect: false,
+                                canEndNotMax: true,
+                                selectPermanentCoroutine: DigimonToDelete,
+                                afterSelectPermanentCoroutine: null,
+                                mode: SelectPermanentEffect.Mode.Custom,
+                                cardEffect: activateClass);
 
-                                selectPermanentEffect.SetUpCustomMessage($"Select 1 {deletableColor.ToString()} Digimon to delete", $"Opponent is selecting 1 {deletableColor.ToString()} Digimon to delete");
+                            selectPermanentEffect.SetUpCustomMessage($"Select 1 {deletableColor.ToString()} Digimon to delete", $"Opponent is selecting 1 {deletableColor.ToString()} Digimon to delete");
 
-                                yield return ContinuousController.instance.StartCoroutine(selectPermanentEffect.Activate());
-                            }
+                            await selectPermanentEffect.Activate();
                         }
+                    }
 
-                        bool CanTargetCondition_ByPreSelecetedList(List<Permanent> permanents, Permanent permanent)
-                        {
-                            if (permanentToDelete.Contains(permanent))
-                                return false;
+                    bool CanTargetCondition_ByPreSelecetedList(List<Permanent> permanents, Permanent permanent)
+                    {
+                        if (permanentToDelete.Contains(permanent))
+                            return false;
 
-                            return true;
-                        }
+                        return true;
+                    }
 
-                        IEnumerator DigimonToDelete(Permanent permanent)
-                        {
-                            if (permanent != null)
-                                permanentToDelete.Add(permanent);
+                    Task DigimonToDelete(Permanent permanent)
+                    {
+                        if (permanent != null)
+                            permanentToDelete.Add(permanent);
 
-                            yield return null;
-                        }
+                        return Task.CompletedTask;
+                    }
 
-                        if (permanentToDelete.Count > 0)
-                        {
-                            Hashtable _hashtable = new Hashtable();
-                            hashtable.Add("CardEffect", activateClass);
-                            yield return ContinuousController.instance.StartCoroutine(new DestroyPermanentsClass(permanentToDelete, _hashtable).Destroy());
-                        }
+                    if (permanentToDelete.Count > 0)
+                    {
+                        Hashtable _hashtable = new Hashtable();
+                        hashtable.Add("CardEffect", activateClass);
+                        await new DestroyPermanentsClass(permanentToDelete, _hashtable).Destroy();
                     }
                 }
             }
+        }
+        #endregion
 
-            #endregion
-
-            #region All Turns
-
-            if (timing == EffectTiming.None)
+        #region All Turns
+        if (timing == EffectTiming.None)
+        {
+            int count()
             {
-                int count()
+                if (CardEffectCommons.IsExistOnBattleArea(card))
                 {
-                    if (CardEffectCommons.IsExistOnBattleArea(card))
-                    {
-                        return card.PermanentOfThisCard().DigivolutionCardsColors.Count;
-                    }
-
-                    return 0;
+                    return ICardEffect.ResolvePermanentOfThisCard(card).DigivolutionCardsColors.Count;
                 }
 
-                bool Condition()
+                return 0;
+            }
+
+            bool Condition()
+            {
+                if (CardEffectCommons.IsExistOnBattleArea(card))
                 {
-                    if (CardEffectCommons.IsExistOnBattleArea(card))
+                    if (count() >= 1)
                     {
-                        if (count() >= 1)
-                        {
-                            return true;
-                        }
+                        return true;
                     }
-
-                    return false;
                 }
 
-                cardEffects.Add(CardEffectFactory.ChangeSelfDPStaticEffect<Func<int>>(changeValue: () => 1000 * count(), isInheritedEffect: false, card: card, condition: Condition));
+                return false;
             }
 
-            #endregion
-
-            return cardEffects;
+            cardEffects.Add(CardEffectFactory.ChangeSelfDPStaticEffect<Func<int>>(changeValue: () => 1000 * count(), isInheritedEffect: false, card: card, condition: Condition));
         }
+        #endregion
+
+        return cardEffects;
     }
 }
