--- DCGO/Assets/Scripts/CardEffect/EX8/Purple/EX8_059.cs	2026-05-18 20:03:15.424234400 +0900
+++ src/HeadlessDCGO.Engine/Assets/Scripts/CardEffect/EX8/Purple/EX8_059.cs	2026-07-29 01:02:23.175949149 +0900
@@ -1,455 +1,447 @@
+namespace HeadlessDCGO.Engine.Assets.Scripts.CardEffect.EX8.Purple;
+
 using System;
 using System.Collections;
 using System.Collections.Generic;
+using System.Threading.Tasks;
+using HeadlessDCGO.Engine.Assets.Scripts.Script;
+using HeadlessDCGO.Engine.Assets.Scripts.Script.CardEffectCommons;
+using HeadlessDCGO.Engine.Assets.Scripts.Script.CardEffects;
+using HeadlessDCGO.Engine.Headless.Services;
 
-namespace DCGO.CardEffects.EX8
+public class EX8_059 : CEntity_Effect
 {
-    public class EX8_059 : CEntity_Effect
+    public override List<ICardEffect> CardEffects(EffectTiming timing, CardSource card)
     {
-        public override List<ICardEffect> CardEffects(EffectTiming timing, CardSource card)
+        List<ICardEffect> cardEffects = new List<ICardEffect>();
+
+        #region Alternate Digivolution
+        if (timing == EffectTiming.None)
         {
-            List<ICardEffect> cardEffects = new List<ICardEffect>();
+            bool PermanentCondition(Permanent targetPermanent)
+            {
+                return targetPermanent.TopCard.HasLevel && targetPermanent.TopCard.IsLevel(3) && targetPermanent.TopCard.EqualsTraits("NSo");
+            }
 
-            #region Alternate Digivolution
+            cardEffects.Add(CardEffectFactory.AddSelfDigivolutionRequirementStaticEffect(
+                permanentCondition: PermanentCondition,
+                digivolutionCost: 2,
+                ignoreDigivolutionRequirement: false,
+                card: card,
+                condition: null));
+        }
+        #endregion
 
-            if (timing == EffectTiming.None)
-            {
-                bool PermanentCondition(Permanent targetPermanent)
-                {
-                    return targetPermanent.TopCard.IsLevel3 && targetPermanent.TopCard.EqualsTraits("NSo");
-                }
+        #region Shared On Play/When Digivolving
+        bool CanSelectPermanentCondition(Permanent permanent)
+        {
+            return CardEffectCommons.IsPermanentExistsOnOpponentBattleAreaDigimon(permanent, card);
+        }
 
-                cardEffects.Add(CardEffectFactory.AddSelfDigivolutionRequirementStaticEffect(
-                    permanentCondition: PermanentCondition,
-                    digivolutionCost: 2,
-                    ignoreDigivolutionRequirement: false,
-                    card: card,
-                    condition: null)
-                );
-            }
+        #endregion
 
-            #endregion
+        #region On Play
+        if (timing == EffectTiming.OnEnterFieldAnyone)
+        {
+            ActivateClass activateClass = new ActivateClass();
+            activateClass.SetUpICardEffect("Give effects to opponent's Digimon", CanUseCondition, card);
+            activateClass.SetUpActivateClass(CanActivateCondition, ActivateCoroutine, -1, false, EffectDescription());
+            cardEffects.Add(activateClass);
 
-            #region Shared On Play/When Digivolving
-            bool CanSelectPermanentCondition(Permanent permanent)
+            string EffectDescription()
             {
-                return CardEffectCommons.IsPermanentExistsOnOpponentBattleAreaDigimon(permanent, card);
+                return "[On Play] By trashing 1 card in your hand, 1 of your opponent's Digimon gains \"[On Deletion] Trash 1 card in your hand.\" until the end of their turn.";
             }
-            #endregion
 
-            #region On Play
-            if (timing == EffectTiming.OnEnterFieldAnyone)
+            bool CanUseCondition(Hashtable hashtable)
             {
-                ActivateClass activateClass = new ActivateClass();
-                activateClass.SetUpICardEffect("Give effects to opponent's Digimon", CanUseCondition, card);
-                activateClass.SetUpActivateClass(CanActivateCondition, ActivateCoroutine, -1, false, EffectDescription());
-                cardEffects.Add(activateClass);
-
-                string EffectDescription()
-                {
-                    return "[On Play] By trashing 1 card in your hand, 1 of your opponent's Digimon gains \"[On Deletion] Trash 1 card in your hand.\" until the end of their turn.";
-                }
-
-                bool CanUseCondition(Hashtable hashtable)
-                {
-                    if (CardEffectCommons.IsExistOnBattleAreaDigimon(card))
-                        return CardEffectCommons.CanTriggerOnPlay(hashtable, card);
+                if (CardEffectCommons.IsExistOnBattleAreaDigimon(card))
+                    return CardEffectCommons.CanTriggerOnPlay(hashtable, card);
 
-                    return false;
-                }
+                return false;
+            }
 
-                bool CanActivateCondition(Hashtable hashtable)
-                {
-                    return CardEffectCommons.IsExistOnBattleAreaDigimon(card);
-                }
+            bool CanActivateCondition(Hashtable hashtable)
+            {
+                return CardEffectCommons.IsExistOnBattleAreaDigimon(card);
+            }
 
-                IEnumerator ActivateCoroutine(Hashtable hashtable)
+            async Task ActivateCoroutine(Hashtable hashtable)
+            {
+                if (new Player(card.Context, card.Owner).HandCards.Count >= 1)
                 {
-                    if (card.Owner.HandCards.Count >= 1)
-                    {
-                        bool discarded = false;
+                    bool discarded = false;
 
-                        int discardCount = 1;
+                    int discardCount = 1;
 
-                        SelectHandEffect selectHandEffect = GManager.instance.GetComponent<SelectHandEffect>();
+                    SelectHandEffect selectHandEffect = GManager.instance.GetComponent<SelectHandEffect>();
 
-                        selectHandEffect.SetUp(
-                            selectPlayer: card.Owner,
-                            canTargetCondition: cardSource => true,
-                            canTargetCondition_ByPreSelecetedList: null,
-                            canEndSelectCondition: null,
-                            maxCount: discardCount,
-                            canNoSelect: true,
-                            canEndNotMax: false,
-                            isShowOpponent: true,
-                            selectCardCoroutine: null,
-                            afterSelectCardCoroutine: AfterSelectCardCoroutine,
-                            mode: SelectHandEffect.Mode.Discard,
-                            cardEffect: activateClass);
+                    selectHandEffect.SetUp(
+                        selectPlayer: card.Owner,
+                        canTargetCondition: cardSource => true,
+                        canTargetCondition_ByPreSelecetedList: null,
+                        canEndSelectCondition: null,
+                        maxCount: discardCount,
+                        canNoSelect: true,
+                        canEndNotMax: false,
+                        isShowOpponent: true,
+                        selectCardCoroutine: null,
+                        afterSelectCardCoroutine: AfterSelectCardCoroutine,
+                        mode: SelectHandEffect.Mode.Discard,
+                        cardEffect: activateClass);
 
-                        yield return StartCoroutine(selectHandEffect.Activate());
+                    await selectHandEffect.Activate();
 
-                        IEnumerator AfterSelectCardCoroutine(List<CardSource> cardSources)
+                    async Task AfterSelectCardCoroutine(List<CardSource> cardSources)
+                    {
+                        if (cardSources.Count == 1)
                         {
-                            if (cardSources.Count == 1)
-                            {
-                                discarded = true;
+                            discarded = true;
 
-                                yield return null;
-                            }
+                            await Task.CompletedTask;
                         }
+                    }
 
-                        if (discarded)
+                    if (discarded)
+                    {
+                        if (CardEffectCommons.HasMatchConditionPermanent(card, CanSelectPermanentCondition))
                         {
-                            if (CardEffectCommons.HasMatchConditionPermanent(CanSelectPermanentCondition))
-                            {
-                                Permanent selectedPermanent = null;
+                            Permanent selectedPermanent = null;
 
-                                int maxCount = Math.Min(1,
-                                    CardEffectCommons.MatchConditionPermanentCount(CanSelectPermanentCondition));
+                            int maxCount = Math.Min(1,
+                                CardEffectCommons.MatchConditionPermanentCount(card, CanSelectPermanentCondition));
 
-                                SelectPermanentEffect selectPermanentEffect =
-                                    GManager.instance.GetComponent<SelectPermanentEffect>();
+                            SelectPermanentEffect selectPermanentEffect =
+                                GManager.instance.GetComponent<SelectPermanentEffect>();
 
-                                selectPermanentEffect.SetUp(
-                                    selectPlayer: card.Owner,
-                                    canTargetCondition: CanSelectPermanentCondition,
-                                    canTargetCondition_ByPreSelecetedList: null,
-                                    canEndSelectCondition: null,
-                                    maxCount: maxCount,
-                                    canNoSelect: false,
-                                    canEndNotMax: false,
-                                    selectPermanentCoroutine: SelectPermanentCoroutine,
-                                    afterSelectPermanentCoroutine: null,
-                                    mode: SelectPermanentEffect.Mode.Custom,
-                                    cardEffect: activateClass);
-
-                                selectPermanentEffect.SetUpCustomMessage(
-                                    "Select 1 Digimon that will get [On Deletion] Trash 1 card in your hand.",
-                                    "The opponent is selecting 1 Digimon that will get [On Deletion] Trash 1 card in your hand.");
+                            selectPermanentEffect.SetUp(
+                                selectPlayer: card.Owner,
+                                canTargetCondition: CanSelectPermanentCondition,
+                                canTargetCondition_ByPreSelecetedList: null,
+                                canEndSelectCondition: null,
+                                maxCount: maxCount,
+                                canNoSelect: false,
+                                canEndNotMax: false,
+                                selectPermanentCoroutine: SelectPermanentCoroutine,
+                                afterSelectPermanentCoroutine: null,
+                                mode: SelectPermanentEffect.Mode.Custom,
+                                cardEffect: activateClass);
+
+                            selectPermanentEffect.SetUpCustomMessage(
+                                "Select 1 Digimon that will get [On Deletion] Trash 1 card in your hand.",
+                                "The opponent is selecting 1 Digimon that will get [On Deletion] Trash 1 card in your hand.");
 
-                                yield return ContinuousController.instance.StartCoroutine(selectPermanentEffect.Activate());
+                            await selectPermanentEffect.Activate();
 
-                                IEnumerator SelectPermanentCoroutine(Permanent permanent)
-                                {
-                                    if(permanent != null)
-                                        selectedPermanent = permanent;
+                            Task SelectPermanentCoroutine(Permanent permanent)
+                            {
+                                if (permanent != null)
+                                    selectedPermanent = permanent;
 
-                                    yield return null;
-                                }
+                                return Task.CompletedTask;
+                            }
 
-                                if (selectedPermanent != null)
-                                {
-                                    CardSource _topCard = selectedPermanent.TopCard;
+                            if (selectedPermanent != null)
+                            {
+                                CardSource _topCard = selectedPermanent.TopCard;
 
-                                    ActivateClass activateClass1 = new ActivateClass();
-                                    activateClass1.SetUpICardEffect("Trash 1 card", CanUseCondition1, selectedPermanent.TopCard);
-                                    activateClass1.SetUpActivateClass(CanActivateCondition1, ActivateCoroutine1, -1, false, EffectDiscription1());
-                                    activateClass1.SetEffectSourcePermanent(selectedPermanent);
-                                    CardEffectCommons.AddEffectToPermanent(targetPermanent: selectedPermanent, effectDuration: EffectDuration.UntilOpponentTurnEnd, card: card, cardEffect: activateClass1, timing: EffectTiming.OnDestroyedAnyone);
+                                ActivateClass activateClass1 = new ActivateClass();
+                                activateClass1.SetUpICardEffect("Trash 1 card", CanUseCondition1, selectedPermanent.TopCard);
+                                activateClass1.SetUpActivateClass(CanActivateCondition1, ActivateCoroutine1, -1, false, EffectDiscription1());
+                                activateClass1.SetEffectSourcePermanent(selectedPermanent);
+                                CardEffectCommons.AddEffectToPermanent(targetPermanent: selectedPermanent, effectDuration: EffectDuration.UntilOpponentTurnEnd, card: card, cardEffect: activateClass1, timing: EffectTiming.OnDestroyedAnyone);
 
-                                    if (!selectedPermanent.TopCard.CanNotBeAffected(activateClass))
-                                    {
-                                        yield return ContinuousController.instance.StartCoroutine(GManager.instance.GetComponent<Effects>().CreateDebuffEffect(selectedPermanent));
-                                    }
 
-                                    string EffectDiscription1()
-                                    {
-                                        return "[On Deletion] Trash 1 card in your hand.";
-                                    }
+                                string EffectDiscription1()
+                                {
+                                    return "[On Deletion] Trash 1 card in your hand.";
+                                }
 
-                                    bool CanUseCondition1(Hashtable hashtable1)
+                                bool CanUseCondition1(Hashtable hashtable1)
+                                {
+                                    if (CardEffectCommons.IsPermanentExistsOnBattleArea(selectedPermanent))
                                     {
-                                        if (CardEffectCommons.IsPermanentExistsOnBattleArea(selectedPermanent))
+                                        if (CardEffectCommons.CanTriggerOnPermanentDeleted(hashtable1, (permanent) => permanent.InstanceId == selectedPermanent.InstanceId))
                                         {
-                                            if (CardEffectCommons.CanTriggerOnPermanentDeleted(hashtable1, (permanent) => permanent == selectedPermanent))
+                                            if (!selectedPermanent.TopCard.CanNotBeAffected(activateClass))
                                             {
-                                                if (!selectedPermanent.TopCard.CanNotBeAffected(activateClass))
-                                                {
-                                                    return true;
-                                                }
+                                                return true;
                                             }
                                         }
-
-                                        return false;
                                     }
 
-                                    bool CanActivateCondition1(Hashtable hashtable1)
-                                    {
-                                        if (CardEffectCommons.IsTopCardInTrashOnDeletion(hashtable1))
-                                        {
-                                            return _topCard.Owner.HandCards.Count > 0;
-                                        }
+                                    return false;
+                                }
 
-                                        return false;
+                                bool CanActivateCondition1(Hashtable hashtable1)
+                                {
+                                    if (CardEffectCommons.IsTopCardInTrashOnDeletion(hashtable1))
+                                    {
+                                        return new Player(_topCard.Context, _topCard.Owner).HandCards.Count > 0;
                                     }
 
-                                    IEnumerator ActivateCoroutine1(Hashtable _hashtable1)
-                                    {
-                                        SelectHandEffect selectHandEffect = GManager.instance.GetComponent<SelectHandEffect>();
+                                    return false;
+                                }
+
+                                async Task ActivateCoroutine1(Hashtable _hashtable1)
+                                {
+                                    SelectHandEffect selectHandEffect = GManager.instance.GetComponent<SelectHandEffect>();
 
-                                        selectHandEffect.SetUp(
-                                            selectPlayer: _topCard.Owner,
-                                            canTargetCondition: (cardSource) => true,
-                                            canTargetCondition_ByPreSelecetedList: null,
-                                            canEndSelectCondition: null,
-                                            maxCount: 1,
-                                            canNoSelect: false,
-                                            canEndNotMax: false,
-                                            isShowOpponent: true,
-                                            selectCardCoroutine: null,
-                                            afterSelectCardCoroutine: null,
-                                            mode: SelectHandEffect.Mode.Discard,
-                                            cardEffect: activateClass);
+                                    selectHandEffect.SetUp(
+                                        selectPlayer: _topCard.Owner,
+                                        canTargetCondition: (cardSource) => true,
+                                        canTargetCondition_ByPreSelecetedList: null,
+                                        canEndSelectCondition: null,
+                                        maxCount: 1,
+                                        canNoSelect: false,
+                                        canEndNotMax: false,
+                                        isShowOpponent: true,
+                                        selectCardCoroutine: null,
+                                        afterSelectCardCoroutine: null,
+                                        mode: SelectHandEffect.Mode.Discard,
+                                        cardEffect: activateClass);
 
-                                        yield return StartCoroutine(selectHandEffect.Activate());
-                                    }
+                                    await selectHandEffect.Activate();
                                 }
                             }
                         }
                     }
                 }
             }
-            #endregion
+        }
+        #endregion
 
-            #region When Digivolving
-            if (timing == EffectTiming.OnEnterFieldAnyone)
-            {
-                ActivateClass activateClass = new ActivateClass();
-                activateClass.SetUpICardEffect("Give effects to opponent's Digimon", CanUseCondition, card);
-                activateClass.SetUpActivateClass(CanActivateCondition, ActivateCoroutine, -1, false, EffectDescription());
-                cardEffects.Add(activateClass);
+        #region When Digivolving
+        if (timing == EffectTiming.WhenDigivolving)
+        {
+            ActivateClass activateClass = new ActivateClass();
+            activateClass.SetUpICardEffect("Give effects to opponent's Digimon", CanUseCondition, card);
+            activateClass.SetUpActivateClass(CanActivateCondition, ActivateCoroutine, -1, false, EffectDescription());
+            cardEffects.Add(activateClass);
 
-                string EffectDescription()
-                {
-                    return "[When Digivolving] By trashing 1 card in your hand, 1 of your opponent's Digimon gains \"[On Deletion] Trash 1 card in your hand.\" until the end of their turn.";
-                }
+            string EffectDescription()
+            {
+                return "[When Digivolving] By trashing 1 card in your hand, 1 of your opponent's Digimon gains \"[On Deletion] Trash 1 card in your hand.\" until the end of their turn.";
+            }
 
-                bool CanUseCondition(Hashtable hashtable)
-                {
-                    if (CardEffectCommons.IsExistOnBattleAreaDigimon(card))
-                        return CardEffectCommons.CanTriggerWhenDigivolving(hashtable, card);
+            bool CanUseCondition(Hashtable hashtable)
+            {
+                if (CardEffectCommons.IsExistOnBattleAreaDigimon(card))
+                    return CardEffectCommons.CanTriggerWhenDigivolving(hashtable, card);
 
-                    return false;
-                }
+                return false;
+            }
 
-                bool CanActivateCondition(Hashtable hashtable)
-                {
-                    return CardEffectCommons.IsExistOnBattleAreaDigimon(card);
-                }
+            bool CanActivateCondition(Hashtable hashtable)
+            {
+                return CardEffectCommons.IsExistOnBattleAreaDigimon(card);
+            }
 
-                IEnumerator ActivateCoroutine(Hashtable hashtable)
+            async Task ActivateCoroutine(Hashtable hashtable)
+            {
+                if (new Player(card.Context, card.Owner).HandCards.Count >= 1)
                 {
-                    if (card.Owner.HandCards.Count >= 1)
-                    {
-                        bool discarded = false;
+                    bool discarded = false;
 
-                        int discardCount = 1;
+                    int discardCount = 1;
 
-                        SelectHandEffect selectHandEffect = GManager.instance.GetComponent<SelectHandEffect>();
+                    SelectHandEffect selectHandEffect = GManager.instance.GetComponent<SelectHandEffect>();
 
-                        selectHandEffect.SetUp(
-                            selectPlayer: card.Owner,
-                            canTargetCondition: cardSource => true,
-                            canTargetCondition_ByPreSelecetedList: null,
-                            canEndSelectCondition: null,
-                            maxCount: discardCount,
-                            canNoSelect: true,
-                            canEndNotMax: false,
-                            isShowOpponent: true,
-                            selectCardCoroutine: null,
-                            afterSelectCardCoroutine: AfterSelectCardCoroutine,
-                            mode: SelectHandEffect.Mode.Discard,
-                            cardEffect: activateClass);
+                    selectHandEffect.SetUp(
+                        selectPlayer: card.Owner,
+                        canTargetCondition: cardSource => true,
+                        canTargetCondition_ByPreSelecetedList: null,
+                        canEndSelectCondition: null,
+                        maxCount: discardCount,
+                        canNoSelect: true,
+                        canEndNotMax: false,
+                        isShowOpponent: true,
+                        selectCardCoroutine: null,
+                        afterSelectCardCoroutine: AfterSelectCardCoroutine,
+                        mode: SelectHandEffect.Mode.Discard,
+                        cardEffect: activateClass);
 
-                        yield return StartCoroutine(selectHandEffect.Activate());
+                    await selectHandEffect.Activate();
 
-                        IEnumerator AfterSelectCardCoroutine(List<CardSource> cardSources)
+                    async Task AfterSelectCardCoroutine(List<CardSource> cardSources)
+                    {
+                        if (cardSources.Count == 1)
                         {
-                            if (cardSources.Count == 1)
-                            {
-                                discarded = true;
+                            discarded = true;
 
-                                yield return null;
-                            }
+                            await Task.CompletedTask;
                         }
+                    }
 
-                        if (discarded)
+                    if (discarded)
+                    {
+                        if (CardEffectCommons.HasMatchConditionPermanent(card, CanSelectPermanentCondition))
                         {
-                            if (CardEffectCommons.HasMatchConditionPermanent(CanSelectPermanentCondition))
-                            {
-                                Permanent selectedPermanent = null;
+                            Permanent selectedPermanent = null;
 
-                                int maxCount = Math.Min(1,
-                                    CardEffectCommons.MatchConditionPermanentCount(CanSelectPermanentCondition));
+                            int maxCount = Math.Min(1,
+                                CardEffectCommons.MatchConditionPermanentCount(card, CanSelectPermanentCondition));
 
-                                SelectPermanentEffect selectPermanentEffect =
-                                    GManager.instance.GetComponent<SelectPermanentEffect>();
+                            SelectPermanentEffect selectPermanentEffect =
+                                GManager.instance.GetComponent<SelectPermanentEffect>();
 
-                                selectPermanentEffect.SetUp(
-                                    selectPlayer: card.Owner,
-                                    canTargetCondition: CanSelectPermanentCondition,
-                                    canTargetCondition_ByPreSelecetedList: null,
-                                    canEndSelectCondition: null,
-                                    maxCount: maxCount,
-                                    canNoSelect: false,
-                                    canEndNotMax: false,
-                                    selectPermanentCoroutine: SelectPermanentCoroutine,
-                                    afterSelectPermanentCoroutine: null,
-                                    mode: SelectPermanentEffect.Mode.Custom,
-                                    cardEffect: activateClass);
-
-                                selectPermanentEffect.SetUpCustomMessage(
-                                    "Select 1 Digimon that will get [On Deletion] Trash 1 card in your hand.",
-                                    "The opponent is selecting 1 Digimon that will get [On Deletion] Trash 1 card in your hand.");
+                            selectPermanentEffect.SetUp(
+                                selectPlayer: card.Owner,
+                                canTargetCondition: CanSelectPermanentCondition,
+                                canTargetCondition_ByPreSelecetedList: null,
+                                canEndSelectCondition: null,
+                                maxCount: maxCount,
+                                canNoSelect: false,
+                                canEndNotMax: false,
+                                selectPermanentCoroutine: SelectPermanentCoroutine,
+                                afterSelectPermanentCoroutine: null,
+                                mode: SelectPermanentEffect.Mode.Custom,
+                                cardEffect: activateClass);
+
+                            selectPermanentEffect.SetUpCustomMessage(
+                                "Select 1 Digimon that will get [On Deletion] Trash 1 card in your hand.",
+                                "The opponent is selecting 1 Digimon that will get [On Deletion] Trash 1 card in your hand.");
 
-                                yield return ContinuousController.instance.StartCoroutine(selectPermanentEffect.Activate());
+                            await selectPermanentEffect.Activate();
 
-                                IEnumerator SelectPermanentCoroutine(Permanent permanent)
-                                {
-                                    if (permanent != null)
-                                        selectedPermanent = permanent;
+                            Task SelectPermanentCoroutine(Permanent permanent)
+                            {
+                                if (permanent != null)
+                                    selectedPermanent = permanent;
 
-                                    yield return null;
-                                }
+                                return Task.CompletedTask;
+                            }
 
-                                if (selectedPermanent != null)
-                                {
-                                    CardSource _topCard = selectedPermanent.TopCard;
+                            if (selectedPermanent != null)
+                            {
+                                CardSource _topCard = selectedPermanent.TopCard;
 
-                                    ActivateClass activateClass1 = new ActivateClass();
-                                    activateClass1.SetUpICardEffect("Trash 1 card", CanUseCondition1, selectedPermanent.TopCard);
-                                    activateClass1.SetUpActivateClass(CanActivateCondition1, ActivateCoroutine1, -1, false, EffectDiscription1());
-                                    activateClass1.SetEffectSourcePermanent(selectedPermanent);
-                                    CardEffectCommons.AddEffectToPermanent(targetPermanent: selectedPermanent, effectDuration: EffectDuration.UntilOpponentTurnEnd, card: card, cardEffect: activateClass1, timing: EffectTiming.OnDestroyedAnyone);
+                                ActivateClass activateClass1 = new ActivateClass();
+                                activateClass1.SetUpICardEffect("Trash 1 card", CanUseCondition1, selectedPermanent.TopCard);
+                                activateClass1.SetUpActivateClass(CanActivateCondition1, ActivateCoroutine1, -1, false, EffectDiscription1());
+                                activateClass1.SetEffectSourcePermanent(selectedPermanent);
+                                CardEffectCommons.AddEffectToPermanent(targetPermanent: selectedPermanent, effectDuration: EffectDuration.UntilOpponentTurnEnd, card: card, cardEffect: activateClass1, timing: EffectTiming.OnDestroyedAnyone);
 
-                                    if (!selectedPermanent.TopCard.CanNotBeAffected(activateClass))
-                                    {
-                                        yield return ContinuousController.instance.StartCoroutine(GManager.instance.GetComponent<Effects>().CreateDebuffEffect(selectedPermanent));
-                                    }
 
-                                    string EffectDiscription1()
-                                    {
-                                        return "[On Deletion] Trash 1 card in your hand.";
-                                    }
+                                string EffectDiscription1()
+                                {
+                                    return "[On Deletion] Trash 1 card in your hand.";
+                                }
 
-                                    bool CanUseCondition1(Hashtable hashtable1)
+                                bool CanUseCondition1(Hashtable hashtable1)
+                                {
+                                    if (CardEffectCommons.IsPermanentExistsOnBattleArea(selectedPermanent))
                                     {
-                                        if (CardEffectCommons.IsPermanentExistsOnBattleArea(selectedPermanent))
+                                        if (CardEffectCommons.CanTriggerOnPermanentDeleted(hashtable1, (permanent) => permanent.InstanceId == selectedPermanent.InstanceId))
                                         {
-                                            if (CardEffectCommons.CanTriggerOnPermanentDeleted(hashtable1, (permanent) => permanent == selectedPermanent))
+                                            if (!selectedPermanent.TopCard.CanNotBeAffected(activateClass))
                                             {
-                                                if (!selectedPermanent.TopCard.CanNotBeAffected(activateClass))
-                                                {
-                                                    return true;
-                                                }
+                                                return true;
                                             }
                                         }
-
-                                        return false;
                                     }
 
-                                    bool CanActivateCondition1(Hashtable hashtable1)
-                                    {
-                                        if (CardEffectCommons.IsTopCardInTrashOnDeletion(hashtable1))
-                                        {
-                                            return _topCard.Owner.HandCards.Count > 0;
-                                        }
+                                    return false;
+                                }
 
-                                        return false;
+                                bool CanActivateCondition1(Hashtable hashtable1)
+                                {
+                                    if (CardEffectCommons.IsTopCardInTrashOnDeletion(hashtable1))
+                                    {
+                                        return new Player(_topCard.Context, _topCard.Owner).HandCards.Count > 0;
                                     }
 
-                                    IEnumerator ActivateCoroutine1(Hashtable _hashtable1)
-                                    {
-                                        SelectHandEffect selectHandEffect = GManager.instance.GetComponent<SelectHandEffect>();
+                                    return false;
+                                }
 
-                                        selectHandEffect.SetUp(
-                                            selectPlayer: _topCard.Owner,
-                                            canTargetCondition: (cardSource) => true,
-                                            canTargetCondition_ByPreSelecetedList: null,
-                                            canEndSelectCondition: null,
-                                            maxCount: 1,
-                                            canNoSelect: false,
-                                            canEndNotMax: false,
-                                            isShowOpponent: true,
-                                            selectCardCoroutine: null,
-                                            afterSelectCardCoroutine: null,
-                                            mode: SelectHandEffect.Mode.Discard,
-                                            cardEffect: activateClass);
+                                async Task ActivateCoroutine1(Hashtable _hashtable1)
+                                {
+                                    SelectHandEffect selectHandEffect = GManager.instance.GetComponent<SelectHandEffect>();
 
-                                        yield return StartCoroutine(selectHandEffect.Activate());
-                                    }
+                                    selectHandEffect.SetUp(
+                                        selectPlayer: _topCard.Owner,
+                                        canTargetCondition: (cardSource) => true,
+                                        canTargetCondition_ByPreSelecetedList: null,
+                                        canEndSelectCondition: null,
+                                        maxCount: 1,
+                                        canNoSelect: false,
+                                        canEndNotMax: false,
+                                        isShowOpponent: true,
+                                        selectCardCoroutine: null,
+                                        afterSelectCardCoroutine: null,
+                                        mode: SelectHandEffect.Mode.Discard,
+                                        cardEffect: activateClass);
+
+                                    await selectHandEffect.Activate();
                                 }
                             }
                         }
                     }
                 }
             }
-            #endregion
-            
-            #region ESS - When Attacking
-            
-            if (timing == EffectTiming.OnAllyAttack)
-            {
-                ActivateClass activateClass = new ActivateClass();
-                activateClass.SetUpICardEffect("Draw 1 and trash 1 card from hand", CanUseCondition, card);
-                activateClass.SetUpActivateClass(CanActivateCondition, ActivateCoroutine, 1, false, EffectDiscription());
-                activateClass.SetIsInheritedEffect(true);
-                activateClass.SetHashString("Draw_EX8_059");
-                cardEffects.Add(activateClass);
+        }
+        #endregion
 
-                string EffectDiscription()
-                {
-                    return "[When Attacking] [Once Per Turn] <Draw 1> and trash 1 card in your hand.";
-                }
+        #region ESS - When Attacking
+        if (timing == EffectTiming.OnAllyAttack)
+        {
+            ActivateClass activateClass = new ActivateClass();
+            activateClass.SetUpICardEffect("Draw 1 and trash 1 card from hand", CanUseCondition, card);
+            activateClass.SetUpActivateClass(CanActivateCondition, ActivateCoroutine, 1, false, EffectDiscription());
+            activateClass.SetIsInheritedEffect(true);
+            activateClass.SetHashString("Draw_EX8_059");
+            cardEffects.Add(activateClass);
+
+            string EffectDiscription()
+            {
+                return "[When Attacking] [Once Per Turn] <Draw 1> and trash 1 card in your hand.";
+            }
 
-                bool CanUseCondition(Hashtable hashtable)
+            bool CanUseCondition(Hashtable hashtable)
+            {
+                return CardEffectCommons.CanTriggerOnAttack(hashtable, card);
+            }
+
+            bool CanActivateCondition(Hashtable hashtable)
+            {
+                if (CardEffectCommons.IsExistOnBattleArea(card))
                 {
-                    return CardEffectCommons.CanTriggerOnAttack(hashtable, card);
+                    return true;
                 }
 
-                bool CanActivateCondition(Hashtable hashtable)
-                {
-                    if (CardEffectCommons.IsExistOnBattleArea(card))
-                    {
-                        return true;
-                    }
+                return false;
+            }
 
-                    return false;
-                }
+            async Task ActivateCoroutine(Hashtable _hashtable)
+            {
+                await new DrawClass(card.Context, card.Owner, 1, activateClass).Draw();
 
-                IEnumerator ActivateCoroutine(Hashtable _hashtable)
+                if (new Player(card.Context, card.Owner).HandCards.Count >= 1)
                 {
-                    yield return ContinuousController.instance.StartCoroutine(new DrawClass(card.Owner, 1, activateClass).Draw());
-
-                    if (card.Owner.HandCards.Count >= 1)
-                    {
-                        int discardCount = 1;
+                    int discardCount = 1;
 
-                        SelectHandEffect selectHandEffect = GManager.instance.GetComponent<SelectHandEffect>();
+                    SelectHandEffect selectHandEffect = GManager.instance.GetComponent<SelectHandEffect>();
 
-                        selectHandEffect.SetUp(
-                            selectPlayer: card.Owner,
-                            canTargetCondition: (cardSource) => true,
-                            canTargetCondition_ByPreSelecetedList: null,
-                            canEndSelectCondition: null,
-                            maxCount: discardCount,
-                            canNoSelect: false,
-                            canEndNotMax: false,
-                            isShowOpponent: true,
-                            selectCardCoroutine: null,
-                            afterSelectCardCoroutine: null,
-                            mode: SelectHandEffect.Mode.Discard,
-                            cardEffect: activateClass);
+                    selectHandEffect.SetUp(
+                        selectPlayer: card.Owner,
+                        canTargetCondition: (cardSource) => true,
+                        canTargetCondition_ByPreSelecetedList: null,
+                        canEndSelectCondition: null,
+                        maxCount: discardCount,
+                        canNoSelect: false,
+                        canEndNotMax: false,
+                        isShowOpponent: true,
+                        selectCardCoroutine: null,
+                        afterSelectCardCoroutine: null,
+                        mode: SelectHandEffect.Mode.Discard,
+                        cardEffect: activateClass);
 
-                        yield return StartCoroutine(selectHandEffect.Activate());
-                    }
+                    await selectHandEffect.Activate();
                 }
             }
-            
-            #endregion
-
-            return cardEffects;
         }
+        #endregion
+
+        return cardEffects;
     }
-}
\ 파일 끝 개행 문자 없음
+}
