--- DCGO/Assets/Scripts/CardEffect/BT19/Black/BT19_061.cs	2026-05-18 20:03:14.812555600 +0900
+++ src/HeadlessDCGO.Engine/Assets/Scripts/CardEffect/BT19/Black/BT19_061.cs	2026-07-29 01:02:22.166849386 +0900
@@ -1,352 +1,352 @@
+namespace HeadlessDCGO.Engine.Assets.Scripts.CardEffect.BT19.Black;
+
 using System.Collections;
-using System.Collections.Generic;
+using System.Threading.Tasks;
+using HeadlessDCGO.Engine.Assets.Scripts.Script;
+using HeadlessDCGO.Engine.Assets.Scripts.Script.CardEffectCommons;
+using HeadlessDCGO.Engine.Assets.Scripts.Script.CardEffects;
+using HeadlessDCGO.Engine.Headless.Services;
 
-namespace DCGO.CardEffects.BT19
+public class BT19_061 : CEntity_Effect
 {
-    public class BT19_061 : CEntity_Effect
+    public override List<ICardEffect> CardEffects(EffectTiming timing, CardSource card)
     {
-        public override List<ICardEffect> CardEffects(EffectTiming timing, CardSource card)
-        {
-            List<ICardEffect> cardEffects = new List<ICardEffect>();
+        List<ICardEffect> cardEffects = new List<ICardEffect>();
 
-            #region Alternate Digivolution
+        #region Alternate Digivolution
 
-            if (timing == EffectTiming.None)
+        if (timing == EffectTiming.None)
+        {
+            bool PermanentCondition(Permanent targetPermanent)
             {
-                bool PermanentCondition(Permanent targetPermanent)
-                {
-                    return targetPermanent.TopCard.HasLevel && targetPermanent.TopCard.Level == 3 &&
-                           targetPermanent.TopCard.EqualsTraits("Xros Heart");
-                }
-
-                cardEffects.Add(CardEffectFactory.AddSelfDigivolutionRequirementStaticEffect(permanentCondition: PermanentCondition,
-                    digivolutionCost: 2, ignoreDigivolutionRequirement: false, card: card, condition: null));
+                return targetPermanent.TopCard.HasLevel && targetPermanent.TopCard.Level == 3 &&
+                       targetPermanent.TopCard.EqualsTraits("Xros Heart");
             }
 
-            #endregion
+            cardEffects.Add(CardEffectFactory.AddSelfDigivolutionRequirementStaticEffect(permanentCondition: PermanentCondition,
+                digivolutionCost: 2, ignoreDigivolutionRequirement: false, card: card, condition: null));
+        }
 
-            #region DigiXros name
+        #endregion
 
-            if (timing == EffectTiming.None)
-            {
-                ChangeCardNamesForDigiXrosClass changeCardNamesForDigiXrosClass = new ChangeCardNamesForDigiXrosClass();
-                changeCardNamesForDigiXrosClass.SetUpICardEffect("Also treated as [Sparrowmon] for a DigiXros", CanUseCondition, card);
-                changeCardNamesForDigiXrosClass.SetUpChangeCardNamesForDigiXrosClass(changeCardNames: ChangeCardNames);
+        #region DigiXros name
 
-                cardEffects.Add(changeCardNamesForDigiXrosClass);
+        if (timing == EffectTiming.None)
+        {
+            ChangeCardNamesForDigiXrosClass changeCardNamesForDigiXrosClass = new ChangeCardNamesForDigiXrosClass();
+            changeCardNamesForDigiXrosClass.SetUpICardEffect("Also treated as [Sparrowmon] for a DigiXros", CanUseCondition, card);
+            changeCardNamesForDigiXrosClass.SetUpChangeCardNamesForDigiXrosClass(changeCardNames: ChangeCardNames);
 
-                bool CanUseCondition(Hashtable hashtable)
-                {
-                    return true;
-                }
+            cardEffects.Add(changeCardNamesForDigiXrosClass);
 
-                List<string> ChangeCardNames(CardSource cardSource, List<string> cardNames)
-                {
-                    if (cardSource == card)
-                    {
-                        cardNames.Add("Sparrowmon");
-                    }
+            bool CanUseCondition(Hashtable hashtable)
+            {
+                return true;
+            }
 
-                    return cardNames;
+            List<string> ChangeCardNames(CardSource cardSource, List<string> cardNames)
+            {
+                if (cardSource == card)
+                {
+                    cardNames.Add("Sparrowmon");
                 }
+
+                return cardNames;
             }
+        }
+
+        #endregion
+
+        #region On Play/ When Digivolving Shared
+
+        bool CanSelectCardConditionShared(CardSource cardSource)
+        {
+            return cardSource.EqualsTraits("Xros Heart") || cardSource.EqualsTraits("Blue Flare");
+        }
+
+        bool CanActivateConditionShared(Hashtable hashtable)
+        {
+            return CardEffectCommons.IsExistOnBattleAreaDigimon(card) &&
+                   new Player(card.Context, card.Owner).LibraryCards.Count >= 1;
+        }
 
-            #endregion
+        #endregion
 
-            #region On Play/ When Digivolving Shared
+        #region On Play
 
-            bool CanSelectCardConditionShared(CardSource cardSource)
+        if (timing == EffectTiming.OnEnterFieldAnyone)
+        {
+            ActivateClass activateClass = new ActivateClass();
+            activateClass.SetUpICardEffect("Reveal the top 3 cards of deck", CanUseCondition, card);
+            activateClass.SetUpActivateClass(CanActivateConditionShared, ActivateCoroutine, -1, false, EffectDescription());
+            cardEffects.Add(activateClass);
+
+            string EffectDescription()
             {
-                return cardSource.EqualsTraits("Xros Heart") || cardSource.EqualsTraits("Blue Flare");
+                return
+                    "[On Play] Reveal the top 3 cards of your deck. Add 1 card with the [Xros Heart]/[Blue Flare] trait among them to the hand. Trash the rest.";
             }
 
-            bool CanActivateConditionShared(Hashtable hashtable)
+            bool CanUseCondition(Hashtable hashtable)
             {
-                return CardEffectCommons.IsExistOnBattleAreaDigimon(card) &&
-                       card.Owner.LibraryCards.Count >= 1;
+                return CardEffectCommons.CanTriggerOnPlay(hashtable, card);
             }
 
-            #endregion
+            async Task ActivateCoroutine(Hashtable hashtable)
+            {
+                await CardEffectCommons.SimplifiedRevealDeckTopCardsAndSelect(
+                        revealCount: 3,
+                        simplifiedSelectCardConditions:
+                        new SimplifiedSelectCardConditionClass[]
+                        {
+                            new(
+                                canTargetCondition: CanSelectCardConditionShared,
+                                message: "Select 1 Digimon card with the [Xros Heart]/[Blue Flare] in one of its traits.",
+                                mode: SelectCardEffect.Mode.AddHand,
+                                maxCount: 1,
+                                selectCardCoroutine: null)
+                        },
+                        remainingCardsPlace: RemainingCardsPlace.Trash,
+                        activateClass: activateClass
+                    );
+            }
+        }
 
-            #region On Play
+        #endregion
 
-            if (timing == EffectTiming.OnEnterFieldAnyone)
-            {
-                ActivateClass activateClass = new ActivateClass();
-                activateClass.SetUpICardEffect("Reveal the top 3 cards of deck", CanUseCondition, card);
-                activateClass.SetUpActivateClass(CanActivateConditionShared, ActivateCoroutine, -1, false, EffectDescription());
-                cardEffects.Add(activateClass);
+        #region When Digivolving
 
-                string EffectDescription()
-                {
-                    return
-                        "[On Play] Reveal the top 3 cards of your deck. Add 1 card with the [Xros Heart]/[Blue Flare] trait among them to the hand. Trash the rest.";
-                }
+        if (timing == EffectTiming.WhenDigivolving)
+        {
+            ActivateClass activateClass = new ActivateClass();
+            activateClass.SetUpICardEffect("Reveal the top 3 cards of deck", CanUseCondition, card);
+            activateClass.SetUpActivateClass(CanActivateConditionShared, ActivateCoroutine, -1, false, EffectDescription());
+            cardEffects.Add(activateClass);
 
-                bool CanUseCondition(Hashtable hashtable)
-                {
-                    return CardEffectCommons.CanTriggerOnPlay(hashtable, card);
-                }
+            string EffectDescription()
+            {
+                return
+                    "[When Digivolving] Reveal the top 3 cards of your deck. Add 1 card with the [Xros Heart]/[Blue Flare] trait among them to the hand. Trash the rest.";
+            }
 
-                IEnumerator ActivateCoroutine(Hashtable hashtable)
-                {
-                    yield return ContinuousController.instance.StartCoroutine(
-                        CardEffectCommons.SimplifiedRevealDeckTopCardsAndSelect(
-                            revealCount: 3,
-                            simplifiedSelectCardConditions:
-                            new SimplifiedSelectCardConditionClass[]
-                            {
-                                new(
-                                    canTargetCondition: CanSelectCardConditionShared,
-                                    message: "Select 1 Digimon card with the [Xros Heart]/[Blue Flare] in one of its traits.",
-                                    mode: SelectCardEffect.Mode.AddHand,
-                                    maxCount: 1,
-                                    selectCardCoroutine: null)
-                            },
-                            remainingCardsPlace: RemainingCardsPlace.Trash,
-                            activateClass: activateClass
-                        ));
-                }
+            bool CanUseCondition(Hashtable hashtable)
+            {
+                return CardEffectCommons.CanTriggerWhenDigivolving(hashtable, card);
             }
 
-            #endregion
+            async Task ActivateCoroutine(Hashtable hashtable)
+            {
+                await CardEffectCommons.SimplifiedRevealDeckTopCardsAndSelect(
+                        revealCount: 3,
+                        simplifiedSelectCardConditions:
+                        new SimplifiedSelectCardConditionClass[]
+                        {
+                            new(
+                                canTargetCondition: CanSelectCardConditionShared,
+                                message: "Select 1 Digimon card with the [Xros Heart]/[Blue Flare] in one of its traits.",
+                                mode: SelectCardEffect.Mode.AddHand,
+                                maxCount: 1,
+                                selectCardCoroutine: null)
+                        },
+                        remainingCardsPlace: RemainingCardsPlace.Trash,
+                        activateClass: activateClass
+                    );
+            }
+        }
 
-            #region When Digivolving
+        #endregion
 
-            if (timing == EffectTiming.OnEnterFieldAnyone)
-            {
-                ActivateClass activateClass = new ActivateClass();
-                activateClass.SetUpICardEffect("Reveal the top 3 cards of deck", CanUseCondition, card);
-                activateClass.SetUpActivateClass(CanActivateConditionShared, ActivateCoroutine, -1, false, EffectDescription());
-                cardEffects.Add(activateClass);
+        #region On Deletion
 
-                string EffectDescription()
-                {
-                    return
-                        "[When Digivolving] Reveal the top 3 cards of your deck. Add 1 card with the [Xros Heart]/[Blue Flare] trait among them to the hand. Trash the rest.";
-                }
+        if (timing == EffectTiming.OnDestroyedAnyone)
+        {
+            ActivateClass activateClass = new ActivateClass();
+            activateClass.SetUpICardEffect("Place 1 [Xros Heart]/[Blue Flare] card from trash under 1 of your Tamers, then <Save>",
+                CanUseCondition,
+                card);
+            activateClass.SetUpActivateClass(CanActivateCondition, ActivateCoroutine, -1, false, EffectDescription());
+            cardEffects.Add(activateClass);
 
-                bool CanUseCondition(Hashtable hashtable)
-                {
-                    return CardEffectCommons.CanTriggerWhenDigivolving(hashtable, card);
-                }
+            string EffectDescription()
+            {
+                return
+                    "[On Deletion] Place 1 Digimon card with the [Xros Heart]/[Blue Flare] trait from your hand or trash under your Tamers.";
+            }
 
-                IEnumerator ActivateCoroutine(Hashtable hashtable)
-                {
-                    yield return ContinuousController.instance.StartCoroutine(
-                        CardEffectCommons.SimplifiedRevealDeckTopCardsAndSelect(
-                            revealCount: 3,
-                            simplifiedSelectCardConditions:
-                            new SimplifiedSelectCardConditionClass[]
-                            {
-                                new(
-                                    canTargetCondition: CanSelectCardConditionShared,
-                                    message: "Select 1 Digimon card with the [Xros Heart]/[Blue Flare] in one of its traits.",
-                                    mode: SelectCardEffect.Mode.AddHand,
-                                    maxCount: 1,
-                                    selectCardCoroutine: null)
-                            },
-                            remainingCardsPlace: RemainingCardsPlace.Trash,
-                            activateClass: activateClass
-                        ));
-                }
+            bool CanUseCondition(Hashtable hashtable)
+            {
+                return CardEffectCommons.CanTriggerOnDeletion(hashtable, card);
             }
 
-            #endregion
+            bool IsOwnTamerCondition(Permanent permanent)
+            {
+                return CardEffectCommons.IsPermanentExistsOnOwnerBattleArea(permanent, card)
+                       && permanent.IsTamer;
+            }
 
-            #region On Deletion
+            bool HasTraitCondition(CardSource cardSource)
+            {
+                return cardSource.IsDigimon && (cardSource.EqualsTraits("Xros Heart") || cardSource.EqualsTraits("Blue Flare"));
+            }
 
-            if (timing == EffectTiming.OnDestroyedAnyone)
+            bool CanActivateCondition(Hashtable hashtable)
             {
-                ActivateClass activateClass = new ActivateClass();
-                activateClass.SetUpICardEffect("Place 1 [Xros Heart]/[Blue Flare] card from trash under 1 of your Tamers, then <Save>",
-                    CanUseCondition,
-                    card);
-                activateClass.SetUpActivateClass(CanActivateCondition, ActivateCoroutine, -1, false, EffectDescription());
-                cardEffects.Add(activateClass);
+                return CardEffectCommons.CanActivateOnDeletion(hashtable, card) &&
+                       CardEffectCommons.HasMatchConditionOwnersPermanent(card, IsOwnTamerCondition) &&
+                       (CardEffectCommons.HasMatchConditionOwnersHand(card, HasTraitCondition) ||
+                        CardEffectCommons.HasMatchConditionOwnersCardInTrash(card, HasTraitCondition));
+            }
 
-                string EffectDescription()
-                {
-                    return
-                        "[On Deletion] Place 1 Digimon card with the [Xros Heart]/[Blue Flare] trait from your hand or trash under your Tamers.";
-                }
+            async Task ActivateCoroutine(Hashtable hashtable)
+            {
+                bool canSelectHand = CardEffectCommons.HasMatchConditionOwnersHand(card, HasTraitCondition);
+                bool canSelectTrash = CardEffectCommons.HasMatchConditionOwnersCardInTrash(card, HasTraitCondition);
 
-                bool CanUseCondition(Hashtable hashtable)
+                if (canSelectHand || canSelectTrash)
                 {
-                    return CardEffectCommons.CanTriggerOnDeletion(hashtable, card);
-                }
+                    if (canSelectHand && canSelectTrash)
+                    {
+                        List<SelectionElement<bool>> selectionElements = new List<SelectionElement<bool>>
+                        {
+                            new(message: "From hand", value: true, spriteIndex: 0),
+                            new(message: "From trash", value: false, spriteIndex: 1),
+                        };
+
+                        string selectPlayerMessage = "From which area do you play a card?";
+                        string notSelectPlayerMessage = "The opponent is choosing from which area to play a card.";
+
+                        GManager.instance.userSelectionManager.SetBoolSelection(selectionElements: selectionElements,
+                            selectPlayer: card.Owner, selectPlayerMessage: selectPlayerMessage,
+                            notSelectPlayerMessage: notSelectPlayerMessage);
+                    }
 
-                bool IsOwnTamerCondition(Permanent permanent)
-                {
-                    return CardEffectCommons.IsPermanentExistsOnOwnerBattleArea(permanent, card)
-                           && permanent.IsTamer;
-                }
+                    else
+                    {
+                        GManager.instance.userSelectionManager.SetBool(canSelectHand);
+                    }
 
-                bool HasTraitCondition(CardSource cardSource)
-                {
-                    return cardSource.IsDigimon && (cardSource.EqualsTraits("Xros Heart") || cardSource.EqualsTraits("Blue Flare"));
-                }
+                    await GManager.instance.userSelectionManager
+                        .WaitForEndSelect();
 
-                bool CanActivateCondition(Hashtable hashtable)
-                {
-                    return CardEffectCommons.CanActivateOnDeletion(hashtable, card) &&
-                           CardEffectCommons.HasMatchConditionOwnersPermanent(card, IsOwnTamerCondition) &&
-                           (CardEffectCommons.HasMatchConditionOwnersHand(card, HasTraitCondition) ||
-                            CardEffectCommons.HasMatchConditionOwnersCardInTrash(card, HasTraitCondition));
-                }
+                    bool fromHand = GManager.instance.userSelectionManager.SelectedBoolValue;
 
-                IEnumerator ActivateCoroutine(Hashtable hashtable)
-                {
-                    bool canSelectHand = CardEffectCommons.HasMatchConditionOwnersHand(card, HasTraitCondition);
-                    bool canSelectTrash = CardEffectCommons.HasMatchConditionOwnersCardInTrash(card, HasTraitCondition);
+                    List<CardSource> selectedCards = new List<CardSource>();
 
-                    if (canSelectHand || canSelectTrash)
+                    Task SelectCardCoroutine(CardSource cardSource)
                     {
-                        if (canSelectHand && canSelectTrash)
-                        {
-                            List<SelectionElement<bool>> selectionElements = new List<SelectionElement<bool>>
-                            {
-                                new(message: "From hand", value: true, spriteIndex: 0),
-                                new(message: "From trash", value: false, spriteIndex: 1),
-                            };
-
-                            string selectPlayerMessage = "From which area do you play a card?";
-                            string notSelectPlayerMessage = "The opponent is choosing from which area to play a card.";
-
-                            GManager.instance.userSelectionManager.SetBoolSelection(selectionElements: selectionElements,
-                                selectPlayer: card.Owner, selectPlayerMessage: selectPlayerMessage,
-                                notSelectPlayerMessage: notSelectPlayerMessage);
-                        }
+                        selectedCards.Add(cardSource);
+                        return Task.CompletedTask;
+                    }
 
-                        else
-                        {
-                            GManager.instance.userSelectionManager.SetBool(canSelectHand);
-                        }
+                    if (fromHand)
+                    {
+                        SelectHandEffect selectHandEffect = GManager.instance.GetComponent<SelectHandEffect>();
 
-                        yield return ContinuousController.instance.StartCoroutine(GManager.instance.userSelectionManager
-                            .WaitForEndSelect());
+                        selectHandEffect.SetUp(
+                            selectPlayer: card.Owner,
+                            canTargetCondition: HasTraitCondition,
+                            canTargetCondition_ByPreSelecetedList: null,
+                            canEndSelectCondition: null,
+                            maxCount: 1,
+                            canNoSelect: false,
+                            canEndNotMax: false,
+                            isShowOpponent: true,
+                            selectCardCoroutine: SelectCardCoroutine,
+                            afterSelectCardCoroutine: null,
+                            mode: SelectHandEffect.Mode.Custom,
+                            cardEffect: activateClass);
+
+                        selectHandEffect.SetUpCustomMessage("Select 1 card to place under a tamer.",
+                            "The opponent is selecting 1 card to play.");
+                        selectHandEffect.SetUpCustomMessage_ShowCard("Placed card");
 
-                        bool fromHand = GManager.instance.userSelectionManager.SelectedBoolValue;
+                        await selectHandEffect.Activate();
+                    }
+                    else
+                    {
+                        SelectCardEffect selectCardEffect = GManager.instance.GetComponent<SelectCardEffect>();
 
-                        List<CardSource> selectedCards = new List<CardSource>();
+                        selectCardEffect.SetUp(
+                            canTargetCondition: HasTraitCondition,
+                            canTargetCondition_ByPreSelecetedList: null,
+                            canEndSelectCondition: null,
+                            canNoSelect: () => false,
+                            selectCardCoroutine: SelectCardCoroutine,
+                            afterSelectCardCoroutine: null,
+                            message: "Select 1 card to place under a tamer.",
+                            maxCount: 1,
+                            canEndNotMax: false,
+                            isShowOpponent: true,
+                            mode: SelectCardEffect.Mode.Custom,
+                            root: SelectCardEffect.Root.Trash,
+                            customRootCardList: null,
+                            canLookReverseCard: true,
+                            selectPlayer: card.Owner,
+                            cardEffect: activateClass);
 
-                        IEnumerator SelectCardCoroutine(CardSource cardSource)
-                        {
-                            selectedCards.Add(cardSource);
-                            yield return null;
-                        }
+                        selectCardEffect.SetUpCustomMessage("Select 1 card to place under a tamer.",
+                            "The opponent is selecting 1 card to place under a tamer.");
 
-                        if (fromHand)
-                        {
-                            SelectHandEffect selectHandEffect = GManager.instance.GetComponent<SelectHandEffect>();
+                        await selectCardEffect.Activate();
+                    }
 
-                            selectHandEffect.SetUp(
-                                selectPlayer: card.Owner,
-                                canTargetCondition: HasTraitCondition,
-                                canTargetCondition_ByPreSelecetedList: null,
-                                canEndSelectCondition: null,
-                                maxCount: 1,
-                                canNoSelect: false,
-                                canEndNotMax: false,
-                                isShowOpponent: true,
-                                selectCardCoroutine: SelectCardCoroutine,
-                                afterSelectCardCoroutine: null,
-                                mode: SelectHandEffect.Mode.Custom,
-                                cardEffect: activateClass);
-
-                            selectHandEffect.SetUpCustomMessage("Select 1 card to place under a tamer.",
-                                "The opponent is selecting 1 card to play.");
-                            selectHandEffect.SetUpCustomMessage_ShowCard("Placed card");
+                    if (selectedCards.Count >= 1)
+                    {
+                        Permanent? selectedPermanent = null;
 
-                            yield return StartCoroutine(selectHandEffect.Activate());
-                        }
-                        else
-                        {
-                            SelectCardEffect selectCardEffect = GManager.instance.GetComponent<SelectCardEffect>();
+                        SelectPermanentEffect selectPermanentEffect = GManager.instance.GetComponent<SelectPermanentEffect>();
 
-                            selectCardEffect.SetUp(
-                                canTargetCondition: HasTraitCondition,
-                                canTargetCondition_ByPreSelecetedList: null,
-                                canEndSelectCondition: null,
-                                canNoSelect: () => false,
-                                selectCardCoroutine: SelectCardCoroutine,
-                                afterSelectCardCoroutine: null,
-                                message: "Select 1 card to place under a tamer.",
-                                maxCount: 1,
-                                canEndNotMax: false,
-                                isShowOpponent: true,
-                                mode: SelectCardEffect.Mode.Custom,
-                                root: SelectCardEffect.Root.Trash,
-                                customRootCardList: null,
-                                canLookReverseCard: true,
-                                selectPlayer: card.Owner,
-                                cardEffect: activateClass);
+                        selectPermanentEffect.SetUp(
+                            selectPlayer: card.Owner,
+                            canTargetCondition: IsOwnTamerCondition,
+                            canTargetCondition_ByPreSelecetedList: null,
+                            canEndSelectCondition: null,
+                            maxCount: 1,
+                            canNoSelect: true,
+                            canEndNotMax: false,
+                            selectPermanentCoroutine: SelectPermanentCoroutine,
+                            afterSelectPermanentCoroutine: null,
+                            mode: SelectPermanentEffect.Mode.Custom,
+                            cardEffect: activateClass);
 
-                            selectCardEffect.SetUpCustomMessage("Select 1 card to place under a tamer.",
-                                "The opponent is selecting 1 card to place under a tamer.");
+                        selectPermanentEffect.SetUpCustomMessage("Select 1 Tamer to place the chosen card under.",
+                            "The opponent is selecting 1 Tamer to place the chosen card under.");
 
-                            yield return StartCoroutine(selectCardEffect.Activate());
-                        }
+                        await selectPermanentEffect.Activate();
 
-                        if (selectedCards.Count >= 1)
+                        Task SelectPermanentCoroutine(Permanent permanent)
                         {
-                            Permanent selectedPermanent = null;
+                            selectedPermanent = permanent;
 
-                            SelectPermanentEffect selectPermanentEffect = GManager.instance.GetComponent<SelectPermanentEffect>();
+                            return Task.CompletedTask;
+                        }
 
-                            selectPermanentEffect.SetUp(
-                                selectPlayer: card.Owner,
-                                canTargetCondition: IsOwnTamerCondition,
-                                canTargetCondition_ByPreSelecetedList: null,
-                                canEndSelectCondition: null,
-                                maxCount: 1,
-                                canNoSelect: true,
-                                canEndNotMax: false,
-                                selectPermanentCoroutine: SelectPermanentCoroutine,
-                                afterSelectPermanentCoroutine: null,
-                                mode: SelectPermanentEffect.Mode.Custom,
-                                cardEffect: activateClass);
-
-                            selectPermanentEffect.SetUpCustomMessage("Select 1 Tamer to place the chosen card under.",
-                                "The opponent is selecting 1 Tamer to place the chosen card under.");
-
-                            yield return ContinuousController.instance.StartCoroutine(selectPermanentEffect.Activate());
-
-                            IEnumerator SelectPermanentCoroutine(Permanent permanent)
-                            {
-                                selectedPermanent = permanent;
-
-                                yield return null;
-                            }
-
-                            if (selectedPermanent != null)
-                            {
-                                yield return ContinuousController.instance.StartCoroutine(
-                                    selectedPermanent.AddDigivolutionCardsBottom(selectedCards, activateClass));
-                            }
+                        if (selectedPermanent != null)
+                        {
+                            await selectedPermanent.AddDigivolutionCardsBottom(selectedCards, activateClass.EffectSourceCard?.InstanceId);
                         }
                     }
                 }
             }
+        }
 
-            #endregion
+        #endregion
 
-            #region Your Turn - ESS
+        #region Your Turn - ESS
 
-            if (timing == EffectTiming.OnCounterTiming)
+        if (timing == EffectTiming.OnCounterTiming)
+        {
+            bool Condition()
             {
-                bool Condition()
-                {
-                    return CardEffectCommons.IsExistOnBattleAreaDigimon(card) && CardEffectCommons.IsOwnerTurn(card) &&
-                           card.PermanentOfThisCard().TopCard.EqualsTraits("Xros Heart");
-                }
-
-                cardEffects.Add(CardEffectFactory.CollisionSelfStaticEffect(isInheritedEffect: true, card: card, condition: Condition));
+                return CardEffectCommons.IsExistOnBattleAreaDigimon(card) && CardEffectCommons.IsOwnerTurn(card) &&
+                       ICardEffect.ResolvePermanentOfThisCard(card).TopCard.EqualsTraits("Xros Heart");
             }
 
-            #endregion
-
-            return cardEffects;
+            cardEffects.Add(CardEffectFactory.CollisionSelfStaticEffect(isInheritedEffect: true, card: card, condition: Condition));
         }
+
+        #endregion
+
+        return cardEffects;
     }
-}
\ 파일 끝 개행 문자 없음
+}
