--- DCGO/Assets/Scripts/CardEffect/BT25/Black/BT25_072.cs	2026-05-18 20:03:15.018146300 +0900
+++ src/HeadlessDCGO.Engine/Assets/Scripts/CardEffect/BT25/Black/BT25_072.cs	2026-07-29 01:02:22.681900319 +0900
@@ -1,377 +1,380 @@
+namespace HeadlessDCGO.Engine.Assets.Scripts.CardEffect.BT25.Black;
+
 using System;
 using System.Collections;
 using System.Collections.Generic;
 using System.Linq;
+using System.Threading.Tasks;
+using HeadlessDCGO.Engine.Assets.Scripts.Script;
+using HeadlessDCGO.Engine.Assets.Scripts.Script.CardEffectCommons;
+using HeadlessDCGO.Engine.Assets.Scripts.Script.CardEffects;
+using HeadlessDCGO.Engine.Headless.Services;
 
-namespace DCGO.CardEffects.BT25
+public class BT25_072 : CEntity_Effect
 {
-    // Shutmon
-    public class BT25_072 : CEntity_Effect
+    public override List<ICardEffect> CardEffects(EffectTiming timing, CardSource card)
     {
-        public override List<ICardEffect> CardEffects(EffectTiming timing, CardSource card)
-        {
-            List<ICardEffect> cardEffects = new List<ICardEffect>();
+        List<ICardEffect> cardEffects = new List<ICardEffect>();
 
-            #region Alternative Digivolution Condition
-            if (timing == EffectTiming.None)
+        #region Alternative Digivolution Condition
+        if (timing == EffectTiming.None)
+        {
+            bool PermanentCondition(Permanent targetPermanent)
             {
-                bool PermanentCondition(Permanent targetPermanent)
-                {
-                    return targetPermanent.TopCard.HasSuperAppTraits;
-                }
-
-                cardEffects.Add(CardEffectFactory.AddSelfDigivolutionRequirementStaticEffect(permanentCondition: PermanentCondition, digivolutionCost: 3, ignoreDigivolutionRequirement: false, card: card, condition: null));
+                return targetPermanent.TopCard.EqualsTraits("Sup.");
             }
-            #endregion
 
-            #region Link Condition
+            cardEffects.Add(CardEffectFactory.AddSelfDigivolutionRequirementStaticEffect(permanentCondition: PermanentCondition, digivolutionCost: 3, ignoreDigivolutionRequirement: false, card: card, condition: null));
+        }
+        #endregion
 
-            if (timing == EffectTiming.None)
-            {
-                static bool PermanentCondition(Permanent targetPermanent)
-                {
-                    return targetPermanent.TopCard.HasAppmonTraits;
-                }
+        #region Link Condition
 
-                cardEffects.Add(CardEffectFactory.AddSelfLinkConditionStaticEffect(permanentCondition: PermanentCondition, linkCost: 3, card: card));
+        if (timing == EffectTiming.None)
+        {
+            static bool PermanentCondition(Permanent targetPermanent)
+            {
+                return targetPermanent.TopCard.EqualsTraits("Appmon");
             }
 
-            #endregion
+            cardEffects.Add(CardEffectFactory.AddSelfLinkConditionStaticEffect(permanentCondition: PermanentCondition, linkCost: 3, card: card));
+        }
 
-            #region App Fusion (Logamon & Timemon)
+        #endregion
 
-            if (timing == EffectTiming.None)
-            {
-                cardEffects.Add(CardEffectFactory.AddAppfuseMethodByName(new List<string>() { "Logamon", "Timemon" }, card));
+        #region App Fusion (Logamon & Timemon)
 
-            }
+        if (timing == EffectTiming.None)
+        {
+            cardEffects.Add(CardEffectFactory.AddAppfuseMethodByName(new List<string>() { "Logamon", "Timemon" }, card));
 
-            #endregion
+        }
 
-            #region Jamming
-            if (timing == EffectTiming.None)
-            {
-                cardEffects.Add(CardEffectFactory.JammingSelfStaticEffect(isInheritedEffect: false, card: card, condition: null));
-            }
-            #endregion
+        #endregion
+
+        #region Jamming
+        if (timing == EffectTiming.None)
+        {
+            cardEffects.Add(CardEffectFactory.JammingSelfStaticEffect(isInheritedEffect: false, card: card, condition: null));
+        }
+        #endregion
+
+        #region Shared OP / WD / WA
 
-            #region Shared OP / WD / WA
+        string SharedEffectName = "Link from hand or digivolution cards to this for -2";
 
-            string SharedEffectName = "Link from hand or digivolution cards to this for -2";
+        string SharedEffectDescription(string tag)
+            => $"[{tag}] If it's your turn, you may link 1 [Social], [Tool] or [Game] trait Digimon card from your trash or this Digimon's digivolution cards to this Digimon with the cost reduced by 2.";
 
-            string SharedEffectDescription(string tag)
-                => $"[{tag}] If it's your turn, you may link 1 [Social], [Tool] or [Game] trait Digimon card from your trash or this Digimon's digivolution cards to this Digimon with the cost reduced by 2.";
+        bool AdditionalActivateCondition(Hashtable hashtable, ActivateClass activateClass)
+        {
+            return CardEffectCommons.IsOwnerTurn(card)
+                && (CardEffectCommons.HasMatchConditionOwnersCardInTrash(card, CanLinkCardActivateCondition)
+                    || ICardEffect.ResolvePermanentOfThisCard(card).DigivolutionCards.Any(CanLinkCardActivateCondition));
+        }
+
+        bool CanLinkCardActivateCondition(CardSource cardSource) => CanLinkCardCondition(cardSource, false);
+
+        bool CanLinkCardEffectCondition(CardSource cardSource) => CanLinkCardCondition(cardSource, true);
+
+        bool CanLinkCardCondition(CardSource cardSource, bool payCost)
+        {
+            return cardSource.IsDigimon
+                && (cardSource.EqualsTraits("Social")
+                    || cardSource.EqualsTraits("Tool")
+                    || cardSource.EqualsTraits("Game"))
+                && cardSource.CanLinkToTargetPermanent(ICardEffect.ResolvePermanentOfThisCard(card), payCost);
+        }
 
-            bool AdditionalActivateCondition(Hashtable hashtable, ActivateClass activateClass)
+        CardEffectFactory.ActivateClassesForSharedEffects
+            (ref cardEffects, timing, card,
+                SharedEffectName,
+                SharedActivateCoroutine,
+                SharedEffectDescription,
+                additionalActivateCondition: AdditionalActivateCondition,
+                optional: false,
+                isSkippable: true,
+                onPlay: true,
+                whenDigivolving: true,
+                whenAttacking: true);
+
+        async Task SharedActivateCoroutine(Hashtable hashtable, ActivateClass activateClass)
+        {
+            #region Link Cost Reduction
+            ICardEffect GetCardEffect(EffectTiming _timing)
             {
-                return CardEffectCommons.IsOwnerTurn(card)
-                    && (CardEffectCommons.HasMatchConditionOwnersCardInTrash(card, CanLinkCardActivateCondition)
-                        || card.PermanentOfThisCard().DigivolutionCards.Any(CanLinkCardActivateCondition));
+                if (_timing == EffectTiming.None)
+                {
+                    return CardEffectFactory.GrantedReduceLinkCostClass(
+                        card: card,
+                        reducedCost: 2,
+                        cardSourceCondition: _ => true,
+                        permanentCondition: _ => true,
+                        rootCondition: _ => true
+                    );
+                }
+
+                return null;
             }
 
-            bool CanLinkCardActivateCondition(CardSource cardSource) => CanLinkCardCondition(cardSource, false);
+            new Player(card.Context, card.Owner).UntilCalculateFixedCostEffect.Add(GetCardEffect);
+            #endregion
 
-            bool CanLinkCardEffectCondition(CardSource cardSource) => CanLinkCardCondition(cardSource, true);
+            bool canSelectTrash = CardEffectCommons.HasMatchConditionOwnersCardInTrash(card, CanLinkCardEffectCondition);
+            bool canSelectSources = ICardEffect.ResolvePermanentOfThisCard(card).DigivolutionCards.Any(CanLinkCardEffectCondition);
 
-            bool CanLinkCardCondition(CardSource cardSource, bool payCost)
+            List<SelectionElement<int>> selectionElements = new List<SelectionElement<int>>();
+            if (canSelectTrash)
             {
-                return cardSource.IsDigimon
-                    && (cardSource.EqualsTraits("Social")
-                        || cardSource.EqualsTraits("Tool")
-                        || cardSource.EqualsTraits("Game"))
-                    && cardSource.CanLinkToTargetPermanent(card.PermanentOfThisCard(), payCost);
+                selectionElements.Add(new SelectionElement<int>(message: $"From trash", value : 1, spriteIndex: 0));
             }
+            if (canSelectSources)
+            {
+                selectionElements.Add(new SelectionElement<int>(message: $"From digivolution cards", value : 2, spriteIndex: 0));
+            }
+            selectionElements.Add(new SelectionElement<int>(message: $"Do not Link", value : 3, spriteIndex: 1));
 
-            CardEffectFactory.ActivateClassesForSharedEffects
-                (ref cardEffects, timing, card,
-                    SharedEffectName,
-                    SharedActivateCoroutine,
-                    SharedEffectDescription,
-                    additionalActivateCondition: AdditionalActivateCondition,
-                    optional: false,
-                    isSkippable: true,
-                    onPlay: true,
-                    whenDigivolving: true,
-                    whenAttacking: true);
+            string selectPlayerMessage = "From which area will you link a card?";
+            string notSelectPlayerMessage = "The opponent is choosing from which area to select a card.";
 
-            IEnumerator SharedActivateCoroutine(Hashtable hashtable, ActivateClass activateClass)
-            {
-                #region Link Cost Reduction
-                    ICardEffect GetCardEffect(EffectTiming _timing)
-                    {
-                        if (_timing == EffectTiming.None)
-                        {
-                            return CardEffectFactory.GrantedReduceLinkCostClass(
-                                card: card, 
-                                reducedCost: 2,
-                                cardSourceCondition: _ => true,
-                                permanentCondition: _ => true,
-                                rootCondition: _ => true
-                            );
-                        }
+            GManager.instance.userSelectionManager.SetIntSelection(selectionElements: selectionElements, selectPlayer: card.Owner, selectPlayerMessage: selectPlayerMessage, notSelectPlayerMessage: notSelectPlayerMessage);
 
-                        return null;
-                    }
+            await GManager.instance.userSelectionManager.WaitForEndSelect();
 
-                    card.Owner.UntilCalculateFixedCostEffect.Add(GetCardEffect);
-                    #endregion
+            bool doLink = GManager.instance.userSelectionManager.SelectedIntValue != 3;
+            bool fromTrash = GManager.instance.userSelectionManager.SelectedIntValue == 1;
+            if (doLink)
+            {
+                if (fromTrash)
+                {
+                    SelectCardEffect selectCardEffect = GManager.instance.GetComponent<SelectCardEffect>();
 
-                    bool canSelectTrash = CardEffectCommons.HasMatchConditionOwnersCardInTrash(card, CanLinkCardEffectCondition);
-                    bool canSelectSources = card.PermanentOfThisCard().DigivolutionCards.Any(CanLinkCardEffectCondition);
+                    selectCardEffect.SetUp(
+                        canTargetCondition: CanLinkCardEffectCondition,
+                        canTargetCondition_ByPreSelecetedList: null,
+                        canEndSelectCondition: null,
+                        canNoSelect: () => true,
+                        selectCardCoroutine: SelectCardCoroutine,
+                        afterSelectCardCoroutine: null,
+                        message: "Select 1 card to add as source.",
+                        maxCount: 1,
+                        canEndNotMax: false,
+                        isShowOpponent: true,
+                        mode: SelectCardEffect.Mode.Custom,
+                        root: SelectCardEffect.Root.Trash,
+                        customRootCardList: null,
+                        canLookReverseCard: true,
+                        selectPlayer: card.Owner,
+                        cardEffect: activateClass);
+
+                    selectCardEffect.SetUpCustomMessage("Select 1 card to link.", "The opponent is selecting 1 card to link.");
+                    selectCardEffect.SetUpCustomMessage_ShowCard("Selected Card");
+
+                    await selectCardEffect.Activate();
+                }
+                else
+                {
+                    SelectCardEffect selectCardEffect = GManager.instance.GetComponent<SelectCardEffect>();
+
+                    selectCardEffect.SetUp(
+                        canTargetCondition: CanLinkCardEffectCondition,
+                        canTargetCondition_ByPreSelecetedList: null,
+                        canEndSelectCondition: null,
+                        canNoSelect: () => true,
+                        selectCardCoroutine: SelectCardCoroutine,
+                        afterSelectCardCoroutine: null,
+                        message: "Select 1 card to add as source.",
+                        maxCount: 1,
+                        canEndNotMax: false,
+                        isShowOpponent: true,
+                        mode: SelectCardEffect.Mode.Custom,
+                        root: SelectCardEffect.Root.DigivolutionCards,
+                        customRootCardList: ICardEffect.ResolvePermanentOfThisCard(card).DigivolutionCards.ToList(),
+                        canLookReverseCard: true,
+                        selectPlayer: card.Owner,
+                        cardEffect: activateClass);
+
+                    selectCardEffect.SetUpCustomMessage("Select 1 card to link.", "The opponent is selecting 1 card to link.");
+                    selectCardEffect.SetUpCustomMessage_ShowCard("Selected Card");
 
-                    List<SelectionElement<int>> selectionElements = new List<SelectionElement<int>>();
-                    if (canSelectTrash)
-                    {
-                        selectionElements.Add(new SelectionElement<int>(message: $"From trash", value : 1, spriteIndex: 0));
-                    }
-                    if (canSelectSources)
-                    {
-                        selectionElements.Add(new SelectionElement<int>(message: $"From digivolution cards", value : 2, spriteIndex: 0));
-                    }
-                    selectionElements.Add(new SelectionElement<int>(message: $"Do not Link", value : 3, spriteIndex: 1));
+                    await selectCardEffect.Activate();
+                }
+
+                async Task SelectCardCoroutine(CardSource cardSource)
+                {
+                    await new ILinkCard(true, cardSource, ICardEffect.ResolvePermanentOfThisCard(card), activateClass).LinkCard();
+                }
+            }
 
-                    string selectPlayerMessage = "From which area will you link a card?";
-                    string notSelectPlayerMessage = "The opponent is choosing from which area to select a card.";
+            #region Remove Link Cost Reduction
+            new Player(card.Context, card.Owner).UntilCalculateFixedCostEffect.Remove(GetCardEffect);
+            #endregion
+        }
 
-                    GManager.instance.userSelectionManager.SetIntSelection(selectionElements: selectionElements, selectPlayer: card.Owner, selectPlayerMessage: selectPlayerMessage, notSelectPlayerMessage: notSelectPlayerMessage);
+        #endregion
 
-                    yield return ContinuousController.instance.StartCoroutine(GManager.instance.userSelectionManager.WaitForEndSelect());
+        #region All Turns
 
-                    bool doLink = GManager.instance.userSelectionManager.SelectedIntValue != 3;
-                    bool fromTrash = GManager.instance.userSelectionManager.SelectedIntValue == 1;
-                    if (doLink)
-                    {
-                        if (fromTrash)
-                        {
-                            SelectCardEffect selectCardEffect = GManager.instance.GetComponent<SelectCardEffect>();
-
-                            selectCardEffect.SetUp(
-                                canTargetCondition: CanLinkCardEffectCondition,
-                                canTargetCondition_ByPreSelecetedList: null,
-                                canEndSelectCondition: null,
-                                canNoSelect: () => true,
-                                selectCardCoroutine: SelectCardCoroutine,
-                                afterSelectCardCoroutine: null,
-                                message: "Select 1 card to add as source.",
-                                maxCount: 1,
-                                canEndNotMax: false,
-                                isShowOpponent: true,
-                                mode: SelectCardEffect.Mode.Custom,
-                                root: SelectCardEffect.Root.Trash,
-                                customRootCardList: null,
-                                canLookReverseCard: true,
-                                selectPlayer: card.Owner,
-                                cardEffect: activateClass);
-
-                            selectCardEffect.SetUpCustomMessage("Select 1 card to link.", "The opponent is selecting 1 card to link.");
-                            selectCardEffect.SetUpCustomMessage_ShowCard("Selected Card");
-
-                            yield return ContinuousController.instance.StartCoroutine(selectCardEffect.Activate());
-                        }
-                        else
-                        {
-                            SelectCardEffect selectCardEffect = GManager.instance.GetComponent<SelectCardEffect>();
-
-                            selectCardEffect.SetUp(
-                                canTargetCondition: CanLinkCardEffectCondition,
-                                canTargetCondition_ByPreSelecetedList: null,
-                                canEndSelectCondition: null,
-                                canNoSelect: () => true,
-                                selectCardCoroutine: SelectCardCoroutine,
-                                afterSelectCardCoroutine: null,
-                                message: "Select 1 card to add as source.",
-                                maxCount: 1,
-                                canEndNotMax: false,
-                                isShowOpponent: true,
-                                mode: SelectCardEffect.Mode.Custom,
-                                root: SelectCardEffect.Root.DigivolutionCards,
-                                customRootCardList: card.PermanentOfThisCard().DigivolutionCards,
-                                canLookReverseCard: true,
-                                selectPlayer: card.Owner,
-                                cardEffect: activateClass);
-
-                            selectCardEffect.SetUpCustomMessage("Select 1 card to link.", "The opponent is selecting 1 card to link.");
-                            selectCardEffect.SetUpCustomMessage_ShowCard("Selected Card");
-
-                            yield return ContinuousController.instance.StartCoroutine(selectCardEffect.Activate());
-                        }
-
-                        IEnumerator SelectCardCoroutine(CardSource cardSource)
-                        {
-                            yield return ContinuousController.instance.StartCoroutine(new ILinkCard(true, cardSource, card.PermanentOfThisCard(), activateClass).LinkCard());
-                        }
-                    }
+        if (timing == EffectTiming.WhenLinked)
+        {
+            ActivateClass activateClass = new ActivateClass();
+            activateClass.SetUpICardEffect("1 enemy digimon or Tamer cannot digivolve until their turn ends", CanUseCondition, card);
+            activateClass.SetUpActivateClass(CanActivateCondition, ActivateCoroutine, 1, false, EffectDescription());
+            activateClass.SetHashString("BT25_072_AT");
+            cardEffects.Add(activateClass);
 
-                    #region Remove Link Cost Reduction
-                    card.Owner.UntilCalculateFixedCostEffect.Remove(GetCardEffect);
-                    #endregion
+            string EffectDescription()
+            {
+                return "[All Turns] [Once Per Turn] When this Digimon gets linked, 1 of your opponent's Digimon or Tamers can't digivolve until their turn ends.";
             }
-            
-            #endregion
 
-            #region All Turns
+            bool PermanentCondition(Permanent permanent)
+            {
+                return permanent == ICardEffect.ResolvePermanentOfThisCard(card);
+            }
 
-            if (timing == EffectTiming.WhenLinked)
+            bool IsOpponentsPermanent(Permanent permanent)
             {
-                ActivateClass activateClass = new ActivateClass();
-                activateClass.SetUpICardEffect("1 enemy digimon or Tamer cannot digivolve until their turn ends", CanUseCondition, card);
-                activateClass.SetUpActivateClass(CanActivateCondition, ActivateCoroutine, 1, false, EffectDescription());
-                activateClass.SetHashString("BT25_072_AT");
-                cardEffects.Add(activateClass);
+                return CardEffectCommons.IsPermanentExistsOnOpponentBattleArea(permanent, card)
+                    && (permanent.IsDigimon
+                        || permanent.IsTamer);
+            }
 
-                string EffectDescription()
-                {
-                    return "[All Turns] [Once Per Turn] When this Digimon gets linked, 1 of your opponent's Digimon or Tamers can't digivolve until their turn ends.";
-                }
+            bool CanUseCondition(Hashtable hashtable)
+            {
+                return CardEffectCommons.IsExistOnBattleArea(card)
+                    && CardEffectCommons.CanTriggerWhenLinked(hashtable, PermanentCondition, null);
+            }
 
-                bool PermanentCondition(Permanent permanent)
-                {
-                    return permanent == card.PermanentOfThisCard();
-                }
+            bool CanActivateCondition(Hashtable hashtable)
+            {
+                return CardEffectCommons.IsExistOnBattleArea(card);
+            }
 
-                bool IsOpponentsPermanent(Permanent permanent)
+            async Task ActivateCoroutine(Hashtable hashtable)
+            {
+                if (CardEffectCommons.HasMatchConditionPermanent(card, IsOpponentsPermanent))
                 {
-                    return CardEffectCommons.IsPermanentExistsOnOpponentBattleArea(permanent, card)
-                        && (permanent.IsDigimon
-                            || permanent.IsTamer);
-                }
+                    SelectPermanentEffect selectPermanentEffect = GManager.instance.GetComponent<SelectPermanentEffect>();
 
-                bool CanUseCondition(Hashtable hashtable)
-                {
-                    return CardEffectCommons.IsExistOnBattleArea(card)
-                        && CardEffectCommons.CanTriggerWhenLinked(hashtable, PermanentCondition, null);
-                }
+                    selectPermanentEffect.SetUp(
+                        selectPlayer: card.Owner,
+                        canTargetCondition: IsOpponentsPermanent,
+                        canTargetCondition_ByPreSelecetedList: null,
+                        canEndSelectCondition: null,
+                        maxCount: 1,
+                        canNoSelect: false,
+                        canEndNotMax: false,
+                        selectPermanentCoroutine: SelectPermanentCoroutine,
+                        afterSelectPermanentCoroutine: null,
+                        mode: SelectPermanentEffect.Mode.Custom,
+                        cardEffect: activateClass);
 
-                bool CanActivateCondition(Hashtable hashtable)
-                {
-                    return CardEffectCommons.IsExistOnBattleArea(card);
-                }
+                    await selectPermanentEffect.Activate();
 
-                IEnumerator ActivateCoroutine(Hashtable hashtable)
-                {
-                    if (CardEffectCommons.HasMatchConditionPermanent(IsOpponentsPermanent))
+                    Task SelectPermanentCoroutine(Permanent permanent)
                     {
-                        SelectPermanentEffect selectPermanentEffect = GManager.instance.GetComponent<SelectPermanentEffect>();
+                        bool PermanentCondition(Permanent otherPermanent) => otherPermanent == permanent;
 
-                        selectPermanentEffect.SetUp(
-                            selectPlayer: card.Owner,
-                            canTargetCondition: IsOpponentsPermanent,
-                            canTargetCondition_ByPreSelecetedList: null,
-                            canEndSelectCondition: null,
-                            maxCount: 1,
-                            canNoSelect: false,
-                            canEndNotMax: false,
-                            selectPermanentCoroutine: SelectPermanentCoroutine,
-                            afterSelectPermanentCoroutine: null,
-                            mode: SelectPermanentEffect.Mode.Custom,
-                            cardEffect: activateClass);
-
-                        yield return ContinuousController.instance.StartCoroutine(selectPermanentEffect.Activate());
-
-                        IEnumerator SelectPermanentCoroutine(Permanent permanent)
-                        {
-                            bool PermanentCondition(Permanent otherPermanent) => otherPermanent == permanent;
-
-                            CanNotDigivolveClass canNotEvolveClass = CardEffectFactory.CanNotDigivolveStaticEffect(
-                                permanentCondition: PermanentCondition, 
-                                cardCondition: (cardSource) => true, 
-                                isInheritedEffect: false, 
-                                card: card, 
-                                condition: () => true, 
-                                effectName: "Can't digivolve");
-
-                            CardEffectCommons.AddEffectToPermanent(
-                                targetPermanent: permanent, 
-                                effectDuration: EffectDuration.UntilOpponentTurnEnd, 
-                                card: card, 
-                                cardEffect: canNotEvolveClass, 
-                                timing: EffectTiming.None);
+                        CanNotDigivolveClass canNotEvolveClass = CardEffectFactory.CanNotDigivolveStaticEffect(
+                            permanentCondition: PermanentCondition,
+                            cardCondition: (cardSource) => true,
+                            isInheritedEffect: false,
+                            card: card,
+                            condition: () => true,
+                            effectName: "Can't digivolve");
+
+                        CardEffectCommons.AddEffectToPermanent(
+                            targetPermanent: permanent,
+                            effectDuration: EffectDuration.UntilOpponentTurnEnd,
+                            card: card,
+                            cardEffect: canNotEvolveClass,
+                            timing: EffectTiming.None);
 
-                            yield return null;
-                        }
+                        return Task.CompletedTask;
                     }
                 }
             }
+        }
 
-            #endregion
+        #endregion
+
+        #region Link
+        if (timing == EffectTiming.OnDeclaration)
+        {
+            cardEffects.Add(CardEffectFactory.LinkEffect(card));
+        }
+        #endregion
+
+        #region When Linking
+
+        if (timing == EffectTiming.WhenLinked)
+        {
+            ActivateClass activateClass = new ActivateClass();
+            activateClass.SetUpICardEffect("2 enemy Digimon or Tamers can't unsuspend until their turn ends", CanUseCondition, card);
+            activateClass.SetUpActivateClass(CanActivateCondition, ActivateCoroutine, -1, false, EffectDescription());
+            activateClass.SetIsLinkedEffect(true);
+            cardEffects.Add(activateClass);
 
-            #region Link
-            if (timing == EffectTiming.OnDeclaration)
+            string EffectDescription()
             {
-                cardEffects.Add(CardEffectFactory.LinkEffect(card));
+                return "[When Linking] 2 of your opponent's Digimon or Tamers can't unsuspend until their turn ends.";
             }
-            #endregion
 
-            #region When Linking
+            bool CanSelectPermanentCondition(Permanent permanent)
+            {
+                return CardEffectCommons.IsPermanentExistsOnOpponentBattleArea(permanent, card)
+                    && (permanent.IsDigimon || permanent.IsTamer);
+            }
 
-            if (timing == EffectTiming.WhenLinked)
+            bool CanUseCondition(Hashtable hashtable)
             {
-                ActivateClass activateClass = new ActivateClass();
-                activateClass.SetUpICardEffect("2 enemy Digimon or Tamers can't unsuspend until their turn ends", CanUseCondition, card);
-                activateClass.SetUpActivateClass(CanActivateCondition, ActivateCoroutine, -1, false, EffectDescription());
-                activateClass.SetIsLinkedEffect(true);
-                cardEffects.Add(activateClass);
+                return CardEffectCommons.CanTriggerWhenLinking(hashtable, null, card);
+            }
 
-                string EffectDescription()
-                {
-                    return "[When Linking] 2 of your opponent's Digimon or Tamers can't unsuspend until their turn ends.";
-                }
+            bool CanActivateCondition(Hashtable hashtable)
+            {
+                return CardEffectCommons.IsExistOnBattleAreaDigimon(card);
+            }
 
-                bool CanSelectPermanentCondition(Permanent permanent)
+            async Task ActivateCoroutine(Hashtable hashtable)
+            {
+                if (CardEffectCommons.HasMatchConditionPermanent(card, CanSelectPermanentCondition))
                 {
-                    return CardEffectCommons.IsPermanentExistsOnOpponentBattleArea(permanent, card)
-                        && (permanent.IsDigimon || permanent.IsTamer);
-                }
+                    int maxCount = Math.Min(2, CardEffectCommons.MatchConditionPermanentCount(card, CanSelectPermanentCondition));
+                    SelectPermanentEffect selectPermanentEffect = GManager.instance.GetComponent<SelectPermanentEffect>();
 
-                bool CanUseCondition(Hashtable hashtable)
-                {
-                    return CardEffectCommons.CanTriggerWhenLinking(hashtable, null, card);
-                }
+                    selectPermanentEffect.SetUp(
+                        selectPlayer: card.Owner,
+                        canTargetCondition: CanSelectPermanentCondition,
+                        canTargetCondition_ByPreSelecetedList: null,
+                        canEndSelectCondition: null,
+                        maxCount: maxCount,
+                        canNoSelect: false,
+                        canEndNotMax: false,
+                        selectPermanentCoroutine: SelectPermanentCoroutine,
+                        afterSelectPermanentCoroutine: null,
+                        mode: SelectPermanentEffect.Mode.Custom,
+                        cardEffect: activateClass);
 
-                bool CanActivateCondition(Hashtable hashtable)
-                {
-                    return CardEffectCommons.IsExistOnBattleAreaDigimon(card);
-                }
+                    await selectPermanentEffect.Activate();
 
-                IEnumerator ActivateCoroutine(Hashtable hashtable)
-                {
-                    if (CardEffectCommons.HasMatchConditionPermanent(CanSelectPermanentCondition))
+                    async Task SelectPermanentCoroutine(Permanent permanent)
                     {
-                        int maxCount = Math.Min(2, CardEffectCommons.MatchConditionPermanentCount(CanSelectPermanentCondition));
-                        SelectPermanentEffect selectPermanentEffect = GManager.instance.GetComponent<SelectPermanentEffect>();
-
-                        selectPermanentEffect.SetUp(
-                            selectPlayer: card.Owner,
-                            canTargetCondition: CanSelectPermanentCondition,
-                            canTargetCondition_ByPreSelecetedList: null,
-                            canEndSelectCondition: null,
-                            maxCount: maxCount,
-                            canNoSelect: false,
-                            canEndNotMax: false,
-                            selectPermanentCoroutine: SelectPermanentCoroutine,
-                            afterSelectPermanentCoroutine: null,
-                            mode: SelectPermanentEffect.Mode.Custom,
-                            cardEffect: activateClass);
-
-                        yield return ContinuousController.instance.StartCoroutine(selectPermanentEffect.Activate());
-
-                        IEnumerator SelectPermanentCoroutine(Permanent permanent)
-                        {
-                            yield return ContinuousController.instance.StartCoroutine(CardEffectCommons.GainCanNotUnsuspend(
-                                targetPermanent: permanent,
-                                effectDuration: EffectDuration.UntilOpponentTurnEnd,
-                                activateClass: activateClass,
-                                condition: null,
-                                effectName: "Can't unsuspend"
-                            ));
-                        }
+                        await CardEffectCommons.GainCanNotUnsuspend(
+                            targetPermanent: permanent,
+                            effectDuration: EffectDuration.UntilOpponentTurnEnd,
+                            activateClass: activateClass,
+                            condition: null,
+                            effectName: "Can't unsuspend"
+                        );
                     }
                 }
             }
+        }
 
-            #endregion
+        #endregion
 
-            return cardEffects;
-        }
+        return cardEffects;
     }
 }
