--- DCGO/Assets/Scripts/CardEffect/AD1/Blue/AD1_012.cs	2026-05-18 20:03:14.493282200 +0900
+++ src/HeadlessDCGO.Engine/Assets/Scripts/CardEffect/AD1/Blue/AD1_012.cs	2026-07-29 01:02:21.322765814 +0900
@@ -1,353 +1,352 @@
-using System;
+namespace HeadlessDCGO.Engine.Assets.Scripts.CardEffect.AD1.Blue;
+
 using System.Collections;
-using System.Collections.Generic;
-using System.Linq;
+using System.Threading.Tasks;
+using HeadlessDCGO.Engine.Assets.Scripts.Script;
+using HeadlessDCGO.Engine.Assets.Scripts.Script.CardEffectCommons;
+using HeadlessDCGO.Engine.Assets.Scripts.Script.CardEffects;
+using HeadlessDCGO.Engine.Headless.Services;
 
-// CresGarurumon
-namespace DCGO.CardEffects.AD1
+public class AD1_012 : CEntity_Effect
 {
-    public class AD1_012 : CEntity_Effect
+    public override List<ICardEffect> CardEffects(EffectTiming timing, CardSource card)
     {
-        public override List<ICardEffect> CardEffects(EffectTiming timing, CardSource card)
-        {
-            List<ICardEffect> cardEffects = new List<ICardEffect>();
+        List<ICardEffect> cardEffects = new List<ICardEffect>();
 
-            #region Alternative Digivolution Condition
-            if (timing == EffectTiming.None)
+        #region Alternative Digivolution Condition
+        if (timing == EffectTiming.None)
+        {
+            bool PermanentCondition(Permanent targetPermanent)
             {
-                bool PermanentCondition(Permanent targetPermanent)
-                {
-                    return targetPermanent.TopCard.ContainsCardName("Garurumon")
-                        || targetPermanent.TopCard.EqualsTraits("ADVENTURE");
-                }
-
-                cardEffects.Add(CardEffectFactory.AddSelfDigivolutionRequirementStaticEffect(
-                    level: 5,
-                    permanentCondition: PermanentCondition,
-                    digivolutionCost: 3,
-                    ignoreDigivolutionRequirement: false,
-                    card: card,
-                    condition: null)
-                );
+                return targetPermanent.TopCard.ContainsCardName("Garurumon")
+                    || targetPermanent.TopCard.EqualsTraits("ADVENTURE");
             }
-            #endregion
 
-            #region Alliance and Evade
-            if (timing == EffectTiming.OnAllyAttack)
-            {
-                cardEffects.Add(CardEffectFactory.AllianceSelfEffect(isInheritedEffect: false, card: card, condition: null));
-            }
+            cardEffects.Add(CardEffectFactory.AddSelfDigivolutionRequirementStaticEffect(
+                level: 5,
+                permanentCondition: PermanentCondition,
+                digivolutionCost: 3,
+                ignoreDigivolutionRequirement: false,
+                card: card,
+                condition: null)
+            );
+        }
+        #endregion
 
-            if (timing == EffectTiming.WhenPermanentWouldBeDeleted)
-            {
-                cardEffects.Add(CardEffectFactory.EvadeSelfEffect(isInheritedEffect: false, card: card, condition: null));
-            }
-            #endregion
+        #region Alliance and Evade
+        if (timing == EffectTiming.OnAllyAttack)
+        {
+            cardEffects.Add(CardEffectFactory.AllianceSelfEffect(isInheritedEffect: false, card: card, condition: null));
+        }
+
+        if (timing == EffectTiming.WhenPermanentWouldBeDeleted)
+        {
+            cardEffects.Add(CardEffectFactory.EvadeSelfEffect(isInheritedEffect: false, card: card, condition: null));
+        }
+        #endregion
 
-            #region Shared OP / WD / WA
+        #region Shared OP / WD / WA
 
-            string SharedHashString = "AD1_012_OP_WD_WA";
+        string SharedHashString = "AD1_012_OP_WD_WA";
 
-            string SharedEffectName = "You may return 1 lowest level opponent to hand. This and 1 [Greymon] may unsuspend";
+        string SharedEffectName = "You may return 1 lowest level opponent to hand. This and 1 [Greymon] may unsuspend";
 
-            string SharedEffectDescription(string tag) 
-                => $"[{tag}] [Once Per Turn] You may return 1 of your opponent's lowest level Digimon to the hand. Then, this Digimon and 1 of your Digimon with [Greymon] in its name may unsuspend.";
+        string SharedEffectDescription(string tag)
+            => $"[{tag}] [Once Per Turn] You may return 1 of your opponent's lowest level Digimon to the hand. Then, this Digimon and 1 of your Digimon with [Greymon] in its name may unsuspend.";
 
-            bool SharedCanActivateCondition(Hashtable hashtable) => CardEffectCommons.IsExistOnBattleArea(card);
+        bool SharedCanActivateCondition(Hashtable hashtable) => CardEffectCommons.IsExistOnBattleArea(card);
 
-            bool IsOpponentsLowestLevel(Permanent permanent) => CardEffectCommons.IsMinLevel(permanent, card.Owner.Enemy);
+        bool IsOpponentsLowestLevel(Permanent permanent) => CardEffectCommons.IsMinLevel(permanent, CardEffectCommons.OpponentOf(card));
 
-            bool IsOwnGreymon(Permanent permanent)
+        bool IsOwnGreymon(Permanent permanent)
+        {
+            return CardEffectCommons.IsPermanentExistsOnOwnerBattleAreaDigimon(permanent, card)
+                && permanent.TopCard.ContainsCardName("Greymon")
+                && permanent.IsSuspended;
+        }
+
+        async Task SharedActivateCoroutine(Hashtable hashtable, ActivateClass activateClass)
+        {
+            if (CardEffectCommons.HasMatchConditionOpponentsPermanent(card, permanent => IsOpponentsLowestLevel(permanent)))
             {
-                return CardEffectCommons.IsPermanentExistsOnOwnerBattleAreaDigimon(permanent, card)
-                    && permanent.TopCard.ContainsCardName("Greymon")
-                    && permanent.IsSuspended;
+                SelectPermanentEffect selectPermanentEffect = GManager.instance.GetComponent<SelectPermanentEffect>();
+                selectPermanentEffect.SetUp(
+                    selectPlayer: card.Owner,
+                    canTargetCondition: IsOpponentsLowestLevel,
+                    canTargetCondition_ByPreSelecetedList: null,
+                    canEndSelectCondition: null,
+                    maxCount: 1,
+                    canNoSelect: true,
+                    canEndNotMax: false,
+                    selectPermanentCoroutine: null,
+                    afterSelectPermanentCoroutine: null,
+                    mode: SelectPermanentEffect.Mode.Bounce,
+                    cardEffect: activateClass);
+
+                selectPermanentEffect.SetUpCustomMessage("Select 1 opponent's Digimon to return to hand", "The opponent is selecting 1 Digimon to return to hand");
+                await selectPermanentEffect.Activate();
             }
 
-            IEnumerator SharedActivateCoroutine(Hashtable hashtable, ActivateClass activateClass)
+            string selectPlayerMessage = "Will you unsuspend your digimon?";
+            string notSelectPlayerMessage = "The opponent is choosing if they will unsuspend.";
+
+            List<SelectionElement<bool>> command_SelectCommands = new List<SelectionElement<bool>>()
             {
-                if (CardEffectCommons.HasMatchConditionOpponentsPermanent(card, IsOpponentsLowestLevel))
+                new SelectionElement<bool>(message: $"Yes", value: true, spriteIndex: 0),
+                new SelectionElement<bool>(message: $"No", value: false, spriteIndex: 1),
+            };
+
+            GManager.instance.userSelectionManager.SetBoolSelection(selectionElements: command_SelectCommands, selectPlayer: card.Owner, selectPlayerMessage: selectPlayerMessage, notSelectPlayerMessage: notSelectPlayerMessage);
+
+            await GManager.instance.userSelectionManager.WaitForEndSelect();
+
+            bool unsuspend = GManager.instance.userSelectionManager.SelectedBoolValue;
+
+            if (unsuspend)
+            {
+                List<Permanent> selectedPermanents = new List<Permanent>() { ICardEffect.ResolvePermanentOfThisCard(card) };
+
+                if (CardEffectCommons.HasMatchConditionOwnersPermanent(card, IsOwnGreymon))
                 {
-                    SelectPermanentEffect selectPermanentEffect = GManager.instance.GetComponent<SelectPermanentEffect>();
+                    SelectPermanentEffect selectPermanentEffect =
+                            GManager.instance.GetComponent<SelectPermanentEffect>();
+
                     selectPermanentEffect.SetUp(
                         selectPlayer: card.Owner,
-                        canTargetCondition: IsOpponentsLowestLevel,
+                        canTargetCondition: IsOwnGreymon,
                         canTargetCondition_ByPreSelecetedList: null,
                         canEndSelectCondition: null,
                         maxCount: 1,
-                        canNoSelect: true,
+                        canNoSelect: false,
                         canEndNotMax: false,
-                        selectPermanentCoroutine: null,
+                        selectPermanentCoroutine: SelectPermanentCoroutine,
                         afterSelectPermanentCoroutine: null,
-                        mode: SelectPermanentEffect.Mode.Bounce,
+                        mode: SelectPermanentEffect.Mode.Custom,
                         cardEffect: activateClass);
 
-                    selectPermanentEffect.SetUpCustomMessage("Select 1 opponent's Digimon to return to hand", "The opponent is selecting 1 Digimon to return to hand");
-                    yield return ContinuousController.instance.StartCoroutine(selectPermanentEffect.Activate());
-                }
+                    selectPermanentEffect.SetUpCustomMessage("Select 1 Greymon to unsuspend with this card.", "The opponent is selecting 1 Digimon to unsuspend.");
 
-                string selectPlayerMessage = "Will you unsuspend your digimon?";
-                string notSelectPlayerMessage = "The opponent is choosing if they will unsuspend.";
+                    await selectPermanentEffect.Activate();
 
-                List<SelectionElement<bool>> command_SelectCommands = new List<SelectionElement<bool>>()
-                {
-                    new SelectionElement<bool>(message: $"Yes", value: true, spriteIndex: 0),
-                    new SelectionElement<bool>(message: $"No", value: false, spriteIndex: 1),
-                };
+                    Task SelectPermanentCoroutine(Permanent permanent)
+                    {
+                        selectedPermanents.Add(permanent);
+                        return Task.CompletedTask;
+                    }
+                }
 
-                GManager.instance.userSelectionManager.SetBoolSelection(selectionElements: command_SelectCommands, selectPlayer: card.Owner, selectPlayerMessage: selectPlayerMessage, notSelectPlayerMessage: notSelectPlayerMessage);
+                await new IUnsuspendPermanents(selectedPermanents, activateClass).Unsuspend();
+            }
+        }
+        #endregion
 
-                yield return ContinuousController.instance.StartCoroutine(GManager.instance.userSelectionManager.WaitForEndSelect());
+        #region On Play
+        if (timing == EffectTiming.OnEnterFieldAnyone)
+        {
+            ActivateClass activateClass = new ActivateClass();
+            activateClass.SetUpICardEffect(SharedEffectName, CanUseCondition, card);
+            activateClass.SetUpActivateClass(SharedCanActivateCondition, (hash) => SharedActivateCoroutine(hash, activateClass), 1, true, SharedEffectDescription("On Play"));
+            activateClass.SetHashString(SharedHashString);
+            cardEffects.Add(activateClass);
 
-                bool unsuspend = GManager.instance.userSelectionManager.SelectedBoolValue;
+            bool CanUseCondition(Hashtable hashtable)
+            {
+                return CardEffectCommons.IsExistOnBattleArea(card)
+                    && CardEffectCommons.CanTriggerOnPlay(hashtable, card);
+            }
+        }
+        #endregion
 
-                if (unsuspend)
-                {
-                    List<Permanent> selectedPermanents = new List<Permanent>() { card.PermanentOfThisCard() };
+        #region When Digivolving
+        if (timing == EffectTiming.WhenDigivolving)
+        {
+            ActivateClass activateClass = new ActivateClass();
+            activateClass.SetUpICardEffect(SharedEffectName, CanUseCondition, card);
+            activateClass.SetUpActivateClass(SharedCanActivateCondition, (hash) => SharedActivateCoroutine(hash, activateClass), 1, true, SharedEffectDescription("When Digivolving"));
+            activateClass.SetHashString(SharedHashString);
+            cardEffects.Add(activateClass);
 
-                    if (CardEffectCommons.HasMatchConditionOwnersPermanent(card, IsOwnGreymon))
-                    {
-                        SelectPermanentEffect selectPermanentEffect =
-                                GManager.instance.GetComponent<SelectPermanentEffect>();
+            bool CanUseCondition(Hashtable hashtable)
+            {
+                return CardEffectCommons.IsExistOnBattleArea(card)
+                    && CardEffectCommons.CanTriggerWhenDigivolving(hashtable, card);
+            }
+        }
+        #endregion
 
-                        selectPermanentEffect.SetUp(
-                            selectPlayer: card.Owner,
-                            canTargetCondition: IsOwnGreymon,
-                            canTargetCondition_ByPreSelecetedList: null,
-                            canEndSelectCondition: null,
-                            maxCount: 1,
-                            canNoSelect: false,
-                            canEndNotMax: false,
-                            selectPermanentCoroutine: SelectPermanentCoroutine,
-                            afterSelectPermanentCoroutine: null,
-                            mode: SelectPermanentEffect.Mode.Custom,
-                            cardEffect: activateClass);
-
-                        selectPermanentEffect.SetUpCustomMessage("Select 1 Greymon to unsuspend with this card.", "The opponent is selecting 1 Digimon to unsuspend.");
-
-                        yield return ContinuousController.instance.StartCoroutine(selectPermanentEffect.Activate());
-
-                        IEnumerator SelectPermanentCoroutine(Permanent permanent)
-                        {
-                            selectedPermanents.Add(permanent);
-                            yield return null;
-                        }
-                    }
+        #region When Attacking
+        if (timing == EffectTiming.OnAllyAttack)
+        {
+            ActivateClass activateClass = new ActivateClass();
+            activateClass.SetUpICardEffect(SharedEffectName, CanUseCondition, card);
+            activateClass.SetUpActivateClass(SharedCanActivateCondition, (hash) => SharedActivateCoroutine(hash, activateClass), 1, true, SharedEffectDescription("When Attacking"));
+            activateClass.SetHashString(SharedHashString);
+            cardEffects.Add(activateClass);
 
-                    yield return ContinuousController.instance.StartCoroutine(new IUnsuspendPermanents(selectedPermanents, activateClass).Unsuspend());
-                }
+            bool CanUseCondition(Hashtable hashtable)
+            {
+                return CardEffectCommons.IsExistOnBattleArea(card)
+                    && CardEffectCommons.CanTriggerOnAttack(hashtable, card);
             }
-            #endregion
+        }
+        #endregion
 
-            #region On Play
-            if (timing == EffectTiming.OnEnterFieldAnyone)
-            {
-                ActivateClass activateClass = new ActivateClass();
-                activateClass.SetUpICardEffect(SharedEffectName, CanUseCondition, card);
-                activateClass.SetUpActivateClass(SharedCanActivateCondition,(hash) => SharedActivateCoroutine(hash, activateClass), 1, true, SharedEffectDescription("On Play"));
-                activateClass.SetHashString(SharedHashString);
-                cardEffects.Add(activateClass);
+        #region Opponent's Turn
+        if (timing == EffectTiming.OnAllyAttack)
+        {
+            ActivateClass activateClass = new ActivateClass();
+            activateClass.SetUpICardEffect("You may DNA digivolve. Then you may change the attack target to 1 Digimon.", CanUseCondition, card);
+            activateClass.SetUpActivateClass(CanActivateCondition, ActivateCoroutine, 1, true, EffectDescription());
+            activateClass.SetHashString("Redirect_AD1_012");
+            cardEffects.Add(activateClass);
 
-                bool CanUseCondition(Hashtable hashtable)
-                {
-                    return CardEffectCommons.IsExistOnBattleArea(card)
-                        && CardEffectCommons.CanTriggerOnPlay(hashtable, card);
-                }
-            }
-            #endregion
+            string EffectDescription()
+                => "[Opponent's Turn] [Once Per Turn] When one of your opponent's Digimon attacks, 2 of your Digimon may DNA digivolve into [Omnimon Alter-S] in the hand. Then, you may change the attack target to 1 of your Digimon.";
 
-            #region When Digivolving
-            if (timing == EffectTiming.OnEnterFieldAnyone)
-            {
-                ActivateClass activateClass = new ActivateClass();
-                activateClass.SetUpICardEffect(SharedEffectName, CanUseCondition, card);
-                activateClass.SetUpActivateClass(SharedCanActivateCondition,(hash) => SharedActivateCoroutine(hash, activateClass), 1, true, SharedEffectDescription("When Digivolving"));
-                activateClass.SetHashString(SharedHashString);
-                cardEffects.Add(activateClass);
+            bool IsOpponentDigimon(Permanent permanent) => CardEffectCommons.IsPermanentExistsOnOpponentBattleAreaDigimon(permanent, card);
 
-                bool CanUseCondition(Hashtable hashtable)
-                {
-                    return CardEffectCommons.IsExistOnBattleArea(card)
-                        && CardEffectCommons.CanTriggerWhenDigivolving(hashtable, card);
-                }
+            bool CanUseCondition(Hashtable hashtable)
+            {
+                return CardEffectCommons.IsExistOnBattleAreaDigimon(card)
+                    && CardEffectCommons.IsOpponentTurn(card)
+                    && CardEffectCommons.CanTriggerOnPermanentAttack(hashtable, IsOpponentDigimon);
             }
-            #endregion
 
-            #region When Attacking
-            if (timing == EffectTiming.OnAllyAttack)
-            {
-                ActivateClass activateClass = new ActivateClass();
-                activateClass.SetUpICardEffect(SharedEffectName, CanUseCondition, card);
-                activateClass.SetUpActivateClass(SharedCanActivateCondition,(hash) => SharedActivateCoroutine(hash, activateClass), 1, true, SharedEffectDescription("When Attacking"));
-                activateClass.SetHashString(SharedHashString);
-                cardEffects.Add(activateClass);
+            bool CanActivateCondition(Hashtable hashtable) => CardEffectCommons.IsExistOnBattleAreaDigimon(card);
 
-                bool CanUseCondition(Hashtable hashtable)
-                {
-                    return CardEffectCommons.IsExistOnBattleArea(card)
-                        && CardEffectCommons.CanTriggerOnAttack(hashtable, card);
-                }
+            bool CanSelectDNACardCondition(CardSource cardSource)
+            {
+                return cardSource.IsDigimon
+                    && cardSource.CanPlayJogress(true)
+                    && cardSource.EqualsCardName("Omnimon Alter-S");
             }
-            #endregion
 
-            #region Opponent's Turn
-            if (timing == EffectTiming.OnAllyAttack)
+            bool IsOwnDigimon(Permanent permanent) => CardEffectCommons.IsPermanentExistsOnOwnerBattleAreaDigimon(permanent, card);
+
+            async Task ActivateCoroutine(Hashtable hashtable)
             {
-                ActivateClass activateClass = new ActivateClass();
-                activateClass.SetUpICardEffect("You may DNA digivolve. Then you may change the attack target to 1 Digimon.", CanUseCondition, card);
-                activateClass.SetUpActivateClass(CanActivateCondition, ActivateCoroutine, 1, true, EffectDescription());
-                activateClass.SetHashString("Redirect_AD1_012");
-                cardEffects.Add(activateClass);
+                #region DNA digivolve
+                await CardEffectCommons.DNADigivolvePermanentsIntoHandOrTrashCard(
+                        CanSelectDNACardCondition,
+                        payCost: true,
+                        isHand: true,
+                        activateClass
+                    );
+                #endregion
+                #region redirect attack
+                Permanent? selectedPermanent = null;
 
-                string EffectDescription()
-                    => "[Opponent's Turn] [Once Per Turn] When one of your opponent's Digimon attacks, 2 of your Digimon may DNA digivolve into [Omnimon Alter-S] in the hand. Then, you may change the attack target to 1 of your Digimon.";
+                SelectPermanentEffect selectPermanentEffect =
+                            GManager.instance.GetComponent<SelectPermanentEffect>();
 
-                bool IsOpponentDigimon(Permanent permanent) => CardEffectCommons.IsPermanentExistsOnOpponentBattleAreaDigimon(permanent, card);
+                selectPermanentEffect.SetUp(
+                    selectPlayer: card.Owner,
+                    canTargetCondition: IsOwnDigimon,
+                    canTargetCondition_ByPreSelecetedList: null,
+                    canEndSelectCondition: null,
+                    maxCount: 1,
+                    canNoSelect: true,
+                    canEndNotMax: false,
+                    selectPermanentCoroutine: SelectPermanentCoroutine,
+                    afterSelectPermanentCoroutine: null,
+                    mode: SelectPermanentEffect.Mode.Custom,
+                    cardEffect: activateClass);
 
-                bool CanUseCondition(Hashtable hashtable)
-                {
-                    return CardEffectCommons.IsExistOnBattleAreaDigimon(card)
-                        && CardEffectCommons.IsOpponentTurn(card)
-                        && CardEffectCommons.CanTriggerOnPermanentAttack(hashtable, IsOpponentDigimon);
-                }
+                selectPermanentEffect.SetUpCustomMessage("Select 1 Digimon to change the attack target to.", "The opponent is selecting 1 Digimon to change the attack target to.");
 
-                bool CanActivateCondition(Hashtable hashtable) => CardEffectCommons.IsExistOnBattleAreaDigimon(card);
+                await selectPermanentEffect.Activate();
 
-                bool CanSelectDNACardCondition(CardSource cardSource)
+                Task SelectPermanentCoroutine(Permanent permanent)
                 {
-                    return cardSource.IsDigimon
-                        && cardSource.CanPlayJogress(true)
-                        && cardSource.EqualsCardName("Omnimon Alter-S");
+                    selectedPermanent = permanent;
+                    return Task.CompletedTask;
                 }
 
-                bool IsOwnDigimon(Permanent permanent) => CardEffectCommons.IsPermanentExistsOnOwnerBattleAreaDigimon(permanent, card);
-
-                IEnumerator ActivateCoroutine(Hashtable hashtable)
+                if (selectedPermanent != null)
                 {
-                    #region DNA digivolve
-                    yield return ContinuousController.instance.StartCoroutine(
-                        CardEffectCommons.DNADigivolvePermanentsIntoHandOrTrashCard(
-                            CanSelectDNACardCondition,
-                            payCost: true,
-                            isHand: true,
-                            activateClass
-                        ));
-                    #endregion
-                    #region redirect attack
-                    Permanent selectedPermanent = null;
-
-                    SelectPermanentEffect selectPermanentEffect =
-                                GManager.instance.GetComponent<SelectPermanentEffect>();
+                    await GManager.instance.attackProcess.SwitchDefender(
+                            activateClass.EffectSourceCard?.InstanceId,
+                            false,
+                            selectedPermanent.InstanceId);
+                }
+                #endregion
+            }
+        }
+        #endregion
 
-                    selectPermanentEffect.SetUp(
-                        selectPlayer: card.Owner,
-                        canTargetCondition: IsOwnDigimon,
-                        canTargetCondition_ByPreSelecetedList: null,
-                        canEndSelectCondition: null,
-                        maxCount: 1,
-                        canNoSelect: true,
-                        canEndNotMax: false,
-                        selectPermanentCoroutine: SelectPermanentCoroutine,
-                        afterSelectPermanentCoroutine: null,
-                        mode: SelectPermanentEffect.Mode.Custom,
-                        cardEffect: activateClass);
+        #region Assembly
+        if (timing == EffectTiming.None)
+        {
+            string cardName1 = "WereGarurumon: Sagittarius Mode";
+            string cardName2 = "Garurumon";
+            AddAssemblyConditionClass addAssemblyConditionClass = new AddAssemblyConditionClass();
+            addAssemblyConditionClass.SetUpICardEffect($"Assembly", CanUseCondition, card);
+            addAssemblyConditionClass.SetUpAddAssemblyConditionClass(getAssemblyCondition: GetAssembly);
+            addAssemblyConditionClass.SetNotShowUI(true);
+            cardEffects.Add(addAssemblyConditionClass);
 
-                    selectPermanentEffect.SetUpCustomMessage("Select 1 Digimon to change the attack target to.", "The opponent is selecting 1 Digimon to change the attack target to.");
+            bool CanUseCondition(Hashtable hashtable)
+            {
+                return true;
+            }
 
-                    yield return ContinuousController.instance.StartCoroutine(selectPermanentEffect.Activate());
+            AssemblyCondition? GetAssembly(CardSource cardSource)
+            {
+                if (cardSource == card)
+                {
+                    AssemblyConditionElement element1 = new AssemblyConditionElement(CanSelectCardCondition1, selectMessage: $"[{cardName1}]", elementCount: 1);
+                    AssemblyConditionElement element2 = new AssemblyConditionElement(CanSelectCardCondition2, selectMessage: $"[{cardName2}]", elementCount: 1);
 
-                    IEnumerator SelectPermanentCoroutine(Permanent permanent)
+                    bool CanSelectCardCondition1(CardSource cardSource)
                     {
-                        selectedPermanent = permanent;
-                        yield return null;
+                        return cardSource != null &&
+                            cardSource.Owner == card.Owner &&
+                            cardSource.IsDigimon &&
+                            cardSource.EqualsCardName(cardName1);
                     }
 
-                    if (selectedPermanent != null)
+                    bool CanSelectCardCondition2(CardSource cardSource)
                     {
-                        yield return ContinuousController.instance.StartCoroutine(GManager.instance.attackProcess.SwitchDefender(
-                                activateClass,
-                                false,
-                                selectedPermanent));
+                        return cardSource != null &&
+                            cardSource.Owner == card.Owner &&
+                            cardSource.IsDigimon &&
+                            cardSource.EqualsCardName(cardName2);
                     }
-                    #endregion
-                }
-            }
-            #endregion
 
-            #region Assembly
-            if (timing == EffectTiming.None)
-            {
-                string cardName1 = "WereGarurumon: Sagittarius Mode";
-                string cardName2 = "Garurumon";
-                AddAssemblyConditionClass addAssemblyConditionClass = new AddAssemblyConditionClass();
-                addAssemblyConditionClass.SetUpICardEffect($"Assembly", CanUseCondition, card);
-                addAssemblyConditionClass.SetUpAddAssemblyConditionClass(getAssemblyCondition: GetAssembly);
-                addAssemblyConditionClass.SetNotShowUI(true);
-                cardEffects.Add(addAssemblyConditionClass);
+                    AssemblyCondition assemblyCondition = new AssemblyCondition(
+                        elements: new List<AssemblyConditionElement>() { element1, element2 },
+                        reduceCost: 4);
 
-                bool CanUseCondition(Hashtable hashtable)
-                {
-                    return true;
+                    return assemblyCondition;
                 }
 
-                AssemblyCondition GetAssembly(CardSource cardSource)
-                {
-                    if (cardSource == card)
-                    {
-                        AssemblyConditionElement element1 = new AssemblyConditionElement(CanSelectCardCondition1, selectMessage: $"[{cardName1}]", elementCount: 1);
-                        AssemblyConditionElement element2 = new AssemblyConditionElement(CanSelectCardCondition2, selectMessage: $"[{cardName2}]", elementCount: 1);
-
-                        bool CanSelectCardCondition1(CardSource cardSource)
-                        {
-                            return cardSource != null && 
-                                cardSource.Owner == card.Owner && 
-                                cardSource.IsDigimon && 
-                                cardSource.EqualsCardName(cardName1);
-                        }
-
-                        bool CanSelectCardCondition2(CardSource cardSource)
-                        {
-                            return cardSource != null && 
-                                cardSource.Owner == card.Owner && 
-                                cardSource.IsDigimon && 
-                                cardSource.EqualsCardName(cardName2);
-                        }
-
-                        AssemblyCondition assemblyCondition = new AssemblyCondition(
-                            elements:new List<AssemblyConditionElement>() { element1, element2 },
-                            reduceCost: 4);
+                return null;
+            }
+        }
+        #endregion
 
-                        return assemblyCondition;
-                    }
+        #region Your Turn - ESS
+        if (timing == EffectTiming.None)
+        {
+            CanNotSwitchAttackTargetClass canNotSwitchAttackTargetClass = new CanNotSwitchAttackTargetClass();
+            canNotSwitchAttackTargetClass.SetUpICardEffect("This Digimon's attack target can't be changed.", CanUseCondition, card);
+            canNotSwitchAttackTargetClass.SetUpCanNotSwitchAttackTargetClass(PermanentCondition: PermanentCondition);
+            canNotSwitchAttackTargetClass.SetIsInheritedEffect(true);
+            cardEffects.Add(canNotSwitchAttackTargetClass);
 
-                    return null;
-                }
+            bool CanUseCondition(Hashtable hashtable)
+            {
+                return CardEffectCommons.IsExistOnBattleAreaDigimon(card) &&
+                       CardEffectCommons.IsOwnerTurn(card);
             }
-            #endregion
 
-            #region Your Turn - ESS
-            if (timing == EffectTiming.None)
+            bool PermanentCondition(Permanent permanent)
             {
-                CanNotSwitchAttackTargetClass canNotSwitchAttackTargetClass = new CanNotSwitchAttackTargetClass();
-                canNotSwitchAttackTargetClass.SetUpICardEffect("This Digimon's attack target can't be changed.", CanUseCondition, card);
-                canNotSwitchAttackTargetClass.SetUpCanNotSwitchAttackTargetClass(PermanentCondition: PermanentCondition);
-                canNotSwitchAttackTargetClass.SetIsInheritedEffect(true);
-                cardEffects.Add(canNotSwitchAttackTargetClass);
-
-                bool CanUseCondition(Hashtable hashtable)
-                {
-                    return CardEffectCommons.IsExistOnBattleAreaDigimon(card) &&
-                           CardEffectCommons.IsOwnerTurn(card);
-                }
-
-                bool PermanentCondition(Permanent permanent)
-                {
-                    return permanent != null && permanent.TopCard && permanent == card.PermanentOfThisCard();
-                }
+                return permanent != null && permanent.TopCard != null && permanent == ICardEffect.ResolvePermanentOfThisCard(card);
             }
-            #endregion
-
-            return cardEffects;
         }
+        #endregion
+
+        return cardEffects;
     }
 }
